// @tableau/taco-toolkit/dist/eps/modules/network-adapter.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createHttpsAgent = exports.NetworkAdapter = void 0;
var node_fs_1 = __importDefault(require("node:fs"));
var node_https_1 = __importDefault(require("node:https"));
var https_proxy_agent_1 = require("https-proxy-agent");
var node_fetch_1 = __importDefault(require("node-fetch"));
var eps_logger_1 = require("../logging/eps-logger");
var url_utils_1 = require("../utils/url-utils");
var NetworkAdapter = /** @class */ (function () {
    function NetworkAdapter(options) {
        this.httpsAgent = createHttpsAgent(options);
    }
    // The function to send the http request to external resources.
    // Return node-fetch response, so the caller may handle the result accordingly.
    NetworkAdapter.prototype.fetch = function (url, options) {
        return __awaiter(this, void 0, void 0, function () {
            var requestInit, e_1, message, masked;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        requestInit = __assign({}, options);
                        if (this.httpsAgent) {
                            requestInit.agent = this.httpsAgent;
                        }
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, (0, node_fetch_1.default)(url, requestInit)];
                    case 2: return [2 /*return*/, _a.sent()];
                    case 3:
                        e_1 = _a.sent();
                        message = e_1.message;
                        if (message && message.includes(url)) {
                            masked = maskURL(url);
                            e_1.message = message.replace(url, masked);
                        }
                        throw e_1;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    return NetworkAdapter;
}());
exports.NetworkAdapter = NetworkAdapter;
function maskURL(urlString) {
    try {
        var urlAndQueryParams = (0, url_utils_1.parseUrlAndQueryParams)(urlString);
        return JSON.stringify(urlAndQueryParams);
    }
    catch (e) {
        // if the url is invalid, keep it as-is.
        return urlString;
    }
}
function createHttpsAgent(options) {
    var httpProxy = options.httpProxy, caCertPath = options.caCertPath;
    var agentOptions = {};
    if (caCertPath) {
        agentOptions.ca = node_fs_1.default.readFileSync(caCertPath);
    }
    if (httpProxy) {
        eps_logger_1.Logger.info("Create NetworkAdapter with http proxy: ".concat(httpProxy));
        return new https_proxy_agent_1.HttpsProxyAgent(httpProxy, agentOptions);
    }
    if (Object.keys(agentOptions).length === 0) {
        return;
    }
    return new node_https_1.default.Agent(agentOptions);
}
exports.createHttpsAgent = createHttpsAgent;
