// @tableau/taco-toolkit/dist/eps/modules/temp-data-manager.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TempDataManager = void 0;
var fs_1 = __importDefault(require("fs"));
var path_1 = __importDefault(require("path"));
var eps_logger_1 = require("../logging/eps-logger");
var file_path_1 = require("../utils/file-path");
var TempDataManager = /** @class */ (function () {
    function TempDataManager(epsInstanceId, tempDirRootPath) {
        this.epsInstanceId = epsInstanceId;
        this.tempDirRootPath = path_1.default.resolve(tempDirRootPath, 'tempData');
        if (this.tempDirRootPath.length >= 136) {
            // Duckdb read APIs can only support filepaths less than 255 characters in length.
            // Fixed path after tempData for Windows is 119 characters, so tempDirRootPath length
            // has to be less than 136.
            throw new Error("tempDirRootPath has to be less than 136 characters in length. tempDirRootPath: ".concat(this.tempDirRootPath));
        }
    }
    // path example: <tempDirRootPath>/tempData/eps-489f4343-1e73-4a22-b047-f0063969c28a/ext-b8ced5be-a728-4a71-98b4-6674cb064498/f-098f6bcd4621d373cade4e832627b4f6
    TempDataManager.prototype.getTempDataFilePath = function (extractorId, filename) {
        var tempDataDir = getExtractorDirPath(this.tempDirRootPath, this.epsInstanceId, extractorId);
        var filepath = path_1.default.join(tempDataDir, "f-".concat((0, file_path_1.getMD5Hash)(filename)));
        // Duckdb read APIs can only support filepaths less than 255 characters in length.
        if (filepath.length >= 255) {
            throw new Error("Temporary data filepath has to be less than 255 characters in length. tempDataFilePath: ".concat(filepath));
        }
        return filepath;
    };
    /**
     * Retrieves the directory path containing partitioned data stored as multiple files
     * associated with a specific table. e.g. parquet data
     *
     * @returns The directory path for the partitioned data of the specified table and extractor.
     * example: <tempDirRootPath>/tempData/eps-489f4343-1e73-4a22-b047-f0063969c28a/ext-b8ced5be-a728-4a71-98b4-6674cb064498/f-098f6bcd4621d373cade4e832627b4f6/
     */
    TempDataManager.prototype.getTempDataDirPath = function (extractorId, dirName) {
        var tempDataDir = getExtractorDirPath(this.tempDirRootPath, this.epsInstanceId, extractorId);
        var dirPath = path_1.default.join(tempDataDir, "f-".concat((0, file_path_1.getMD5Hash)(dirName)));
        return dirPath;
    };
    TempDataManager.prototype.createTempDataDirIfNotExists = function (extractorId) {
        // Adding eps- and extractor- prefixes to make sure that
        // directory creation is successful on Windows OS.
        var tempDataDir = getExtractorDirPath(this.tempDirRootPath, this.epsInstanceId, extractorId);
        // create temp log dir if it does not exist
        if (!fs_1.default.existsSync(tempDataDir)) {
            fs_1.default.mkdirSync(tempDataDir, { recursive: true });
            eps_logger_1.Logger.info("Temporary data directory created: ".concat(tempDataDir));
        }
        return tempDataDir;
    };
    TempDataManager.prototype.deleteExtractorTempDataDir = function (extractorId) {
        var tempDataDir = getExtractorDirPath(this.tempDirRootPath, this.epsInstanceId, extractorId);
        eps_logger_1.Logger.info("Clean up Extractor temporary data directory if exists: ".concat(tempDataDir));
        removeDirIfExists(tempDataDir);
    };
    TempDataManager.prototype.deleteEpsTempDataDir = function () {
        var dirPath = getEPSInstanceDirPath(this.tempDirRootPath, this.epsInstanceId);
        eps_logger_1.Logger.info("Clean up EPS temporary data directory if exists: ".concat(dirPath));
        removeDirIfExists(dirPath);
    };
    return TempDataManager;
}());
exports.TempDataManager = TempDataManager;
/**
 * Delete the given directory if it exists.
 *
 * It was observed that ENOTEMPTY error is intermittently thrown on Windows,
 * so the function will do 3 retries for deletion.
 */
function removeDirIfExists(dirPath) {
    if (fs_1.default.existsSync(dirPath)) {
        fs_1.default.rmSync(dirPath, {
            recursive: true,
            force: true,
            maxRetries: 3,
            retryDelay: 500,
        });
        eps_logger_1.Logger.info("Removed: ".concat(dirPath));
    }
}
function getExtractorDirPath(tempDirRootPath, epsInstanceId, extractorId) {
    var epsInstanceDirPath = getEPSInstanceDirPath(tempDirRootPath, epsInstanceId);
    return path_1.default.join(epsInstanceDirPath, "ext-".concat(extractorId));
}
function getEPSInstanceDirPath(tempDirRootPath, epsInstanceId) {
    return path_1.default.join(tempDirRootPath, "eps-".concat(epsInstanceId));
}
