// @tableau/taco-toolkit/dist/eps/main/core/app-utils/app-init.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getServiceEnablements = exports.getReservedPorts = exports.getGrpcPort = exports.getStaticServerOptions = void 0;
var uuid_1 = require("uuid");
var app_global_1 = __importDefault(require("../app-global"));
var caller_type_1 = require("../../../../shared/enums/caller-type");
function getStaticServerOptions(epsAppOptions) {
    return __awaiter(this, void 0, void 0, function () {
        var staticServerPort, staticServerOptions, _a;
        var _b;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    staticServerPort = epsAppOptions.staticServerPort;
                    _b = {
                        staticServerId: (0, uuid_1.v4)()
                    };
                    if (!(staticServerPort !== null && staticServerPort !== void 0)) return [3 /*break*/, 1];
                    _a = staticServerPort;
                    return [3 /*break*/, 3];
                case 1: return [4 /*yield*/, app_global_1.default.portManager.reserveUnusedPort()];
                case 2:
                    _a = (_c.sent());
                    _c.label = 3;
                case 3:
                    staticServerOptions = (_b.staticServerPort = _a,
                        _b);
                    return [2 /*return*/, staticServerOptions];
            }
        });
    });
}
exports.getStaticServerOptions = getStaticServerOptions;
function getGrpcPort(epsAppOptions) {
    return __awaiter(this, void 0, void 0, function () {
        var grpcServerPort;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    grpcServerPort = epsAppOptions.grpcServerPort;
                    if (grpcServerPort) {
                        return [2 /*return*/, grpcServerPort];
                    }
                    return [4 /*yield*/, app_global_1.default.portManager.reserveUnusedPort()];
                case 1: return [2 /*return*/, _a.sent()];
            }
        });
    });
}
exports.getGrpcPort = getGrpcPort;
function getReservedPorts(optionValues) {
    var mainPort = optionValues.mainPort, staticServerPort = optionValues.staticServerPort, grpcServerPort = optionValues.grpcServerPort;
    return [
        mainPort,
        grpcServerPort,
        staticServerPort,
    ].filter(function (port) { return port !== undefined; });
}
exports.getReservedPorts = getReservedPorts;
function getServiceEnablements(epsAppOptions) {
    // by default, all optional services are enabled
    var enablements = {
        browserWatcher: true,
        staticServer: true,
        grpcServer: true,
    };
    var callerType = epsAppOptions.callerType, featureFlags = epsAppOptions.featureFlags;
    // Turn off grpc server if feature flag is off, while the caller may enforcely
    // turn it on/off on demand.
    if (!featureFlags.EPSSupportgRPC) {
        enablements.grpcServer = false;
    }
    switch (callerType) {
        case caller_type_1.CallerType.dcf: {
            Object.assign(enablements, {
                browserWatcher: false,
                staticServer: false,
                grpcServer: false,
            });
            break;
        }
        case caller_type_1.CallerType.vizQLServer: {
            Object.assign(enablements, {
                browserWatcher: false,
            });
        }
    }
    return enablements;
}
exports.getServiceEnablements = getServiceEnablements;
