// @tableau/taco-toolkit/dist/shared/enums/caller-type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CallerType = void 0;
var CallerType;
(function (CallerType) {
    CallerType["desktop"] = "desktop";
    CallerType["prep"] = "prep";
    CallerType["vizQLServer"] = "vizqlserver";
    CallerType["backgrounder"] = "backgrounder";
    CallerType["vizportal"] = "vizportal";
    CallerType["dataserver"] = "dataserver";
    CallerType["minerva"] = "minerva";
    // connector ui dev server
    CallerType["appdev"] = "appdev";
    // sfdc
    CallerType["dcf"] = "dcf";
})(CallerType = exports.CallerType || (exports.CallerType = {}));
//
// TString AppConfig::GetProcessName()
// {
//     switch (AppConfig::GetInstance().GetServerProcessType())
//     {
//         case AppConfigEnums::ServerProcessType::Backgrounder:
//             return TS("backgrounder");
//         case AppConfigEnums::ServerProcessType::DataServer:
//             return TS("dataserver");
//         case AppConfigEnums::ServerProcessType::VizPortal:
//             return TS("vizportal");
//         case AppConfigEnums::ServerProcessType::VizQLServer:
//             return TS("vizqlserver");
//         case AppConfigEnums::ServerProcessType::WGServer:
//             return TS("wgserver");
//         case AppConfigEnums::ServerProcessType::SiteMaintenance:
//             return TS("siteimportexport");
//         case AppConfigEnums::ServerProcessType::Minerva:
//             return TS("minerva");
//         default:
//             return TS("unknown");
//     }
// }
