// @tableau/taco-toolkit/dist/shared/enums/column-header-property.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnitsFormat = exports.NumberFormat = exports.GeographicRole = exports.ColumnType = exports.ColumnRole = exports.AggType = exports.DataType = void 0;
/**
 * The data type for a table column.
 */
var DataType;
(function (DataType) {
    DataType["Bool"] = "bool";
    DataType["Date"] = "date";
    DataType["Datetime"] = "datetime";
    DataType["Float"] = "float";
    DataType["Int"] = "int";
    DataType["String"] = "string";
    DataType["Geometry"] = "geometry";
    DataType["Unknown"] = "unknown";
})(DataType = exports.DataType || (exports.DataType = {}));
/**
 * The aggregation type for a table column. It can be assigned as metadata
 * for columns of the int and float dataType.
 * */
var AggType;
(function (AggType) {
    AggType["Avg"] = "avg";
    AggType["Count"] = "count";
    AggType["CountDist"] = "count_dist";
    AggType["Median"] = "median";
    AggType["Sum"] = "sum";
})(AggType = exports.AggType || (exports.AggType = {}));
/**
 * The role for a table column.
 */
var ColumnRole;
(function (ColumnRole) {
    ColumnRole["Dimension"] = "dimension";
    ColumnRole["Measure"] = "measure";
})(ColumnRole = exports.ColumnRole || (exports.ColumnRole = {}));
/**
 * The type for a table column.
 */
var ColumnType;
(function (ColumnType) {
    ColumnType["Continuous"] = "continuous";
    ColumnType["Discrete"] = "discrete";
})(ColumnType = exports.ColumnType || (exports.ColumnType = {}));
/**
 * The geographic role for a table column.
 */
var GeographicRole;
(function (GeographicRole) {
    GeographicRole["AreaCode"] = "area_code";
    /** core-based statistical area / metropolitan statistical area */
    GeographicRole["CbsaMsa"] = "cbsa_msa";
    GeographicRole["City"] = "city";
    GeographicRole["CongressionalDistrict"] = "congressional_district";
    GeographicRole["CountryRegion"] = "country_region";
    GeographicRole["County"] = "county";
    GeographicRole["StateProvince"] = "state_province";
    GeographicRole["ZipCodePostcode"] = "zip_code_postcode";
    GeographicRole["Latitude"] = "latitude";
    GeographicRole["Longitude"] = "longitude";
})(GeographicRole = exports.GeographicRole || (exports.GeographicRole = {}));
/**
 * The number format for a table column. It can be assigned as metadata
 * for columns of the int and float dataType.
 * */
var NumberFormat;
(function (NumberFormat) {
    NumberFormat["Number"] = "number";
    NumberFormat["Currency"] = "currency";
    NumberFormat["Scientific"] = "scientific";
    NumberFormat["Percentage"] = "percentage";
})(NumberFormat = exports.NumberFormat || (exports.NumberFormat = {}));
/**
 * The unit format for a table column. It can be assigned as metadata
 * for columns of the int and float dataType.
 * */
var UnitsFormat;
(function (UnitsFormat) {
    UnitsFormat["Thousands"] = "thousands";
    UnitsFormat["Millions"] = "millions";
    UnitsFormat["BillionsEnglish"] = "billions_english";
    UnitsFormat["BillionsStandard"] = "billions_standard";
})(UnitsFormat = exports.UnitsFormat || (exports.UnitsFormat = {}));
