// @tableau/taco-toolkit/dist/shared/enums/connector-launch-status.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectorLaunchStatus = void 0;
/**
 * Connector launch/extract status derived from ExtractorMessage status which is
 * the original message sent from extractor app.
 *
 * This status is used in the EPS /extract api responses.
 */
var ConnectorLaunchStatus;
(function (ConnectorLaunchStatus) {
    /**
     * Extractor process gets started, extractor API is ready, and the initial
     * data extraction completes.
     */
    ConnectorLaunchStatus["Extract"] = "extract";
    /**
     * Extractor process gets started, extractor API is ready, but the initial
     * data extraction fails.
     */
    ConnectorLaunchStatus["Error"] = "error";
    /**
     * App user explicitly cancels the connection, e.g. close the connector UI.
     */
    ConnectorLaunchStatus["Canceled"] = "canceled";
    /**
     * Extractor process has started, but the process crashes during data extraction.
     * Normally, the error is logged on the extractor side.
     */
    ConnectorLaunchStatus["Aborted"] = "aborted";
    /**
     * Seems it was designated for taco/connector related errors, it's currently
     * misused for all unknowns...
     */
    ConnectorLaunchStatus["BadTaco"] = "badTaco";
})(ConnectorLaunchStatus = exports.ConnectorLaunchStatus || (exports.ConnectorLaunchStatus = {}));
