// @tableau/taco-toolkit/dist/eps/main/components/connection-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectionManager = void 0;
var os_1 = require("os");
var eps_logger_1 = require("../../logging/eps-logger");
var ConnectionResource = /** @class */ (function () {
    function ConnectionResource() {
        this.browserProcess = null;
        this.extractorProcesses = [];
        this.extractorIdToPid = {};
        this.isSubmitted = false;
    }
    return ConnectionResource;
}());
/** Cache of resources and settings used for connections keyed by connection id. */
var ConnectionManager = /** @class */ (function () {
    function ConnectionManager(processManager, tempDataManager) {
        this.processManager = processManager;
        this.tempDataManager = tempDataManager;
        /** Resources by connectionId */
        this.resourceMap = new Map();
    }
    // TODO: this could potentially create bunch of unneeded resources to the map...
    // The caller must do the diligent to validate before calling this.
    ConnectionManager.prototype.getOrAddResources = function (connectionId) {
        var resource = this.resourceMap.get(connectionId);
        if (!resource) {
            resource = new ConnectionResource();
            this.resourceMap.set(connectionId, resource);
        }
        return resource;
    };
    ConnectionManager.prototype.setBrowserProcess = function (connectionId, procOrCloseFunc) {
        var resources = this.getOrAddResources(connectionId);
        resources.browserProcess = procOrCloseFunc;
    };
    ConnectionManager.prototype.addExtractorProcess = function (connectionId, proc, extractorId) {
        if (proc.pid === undefined) {
            eps_logger_1.Logger.warn("Extractor process was not started properly. ConnectionId: ".concat(connectionId, ", ExtactorId: ").concat(extractorId));
            return;
        }
        var _a = this.getOrAddResources(connectionId), extractorProcesses = _a.extractorProcesses, extractorIdToPid = _a.extractorIdToPid;
        if (extractorProcesses.length > 0) {
            eps_logger_1.Logger.warn("ConnectionId '".concat(connectionId, "' already has an extractorId '").concat(Object.keys(extractorIdToPid), "'. Adding a new extractorId '").concat(extractorId, "'. The connection may not be in a healthy state."));
        }
        extractorProcesses.push(proc);
        extractorIdToPid[extractorId] = proc.pid;
    };
    ConnectionManager.prototype.hasActiveConnection = function (connectionId) {
        var resource = this.resourceMap.get(connectionId);
        if (resource === undefined) {
            return false;
        }
        return resource.extractorProcesses.length > 0;
    };
    ConnectionManager.prototype.hasConnection = function (connectionId) {
        return this.resourceMap.has(connectionId);
    };
    /** Frees all resources allocated for this connection. */
    ConnectionManager.prototype.free = function (connectionId) {
        var resources = this.resourceMap.get(connectionId);
        if (!resources) {
            return;
        }
        if (resources.browserProcess) {
            this.freeBrowser(connectionId);
        }
        this.freeExtractors(connectionId);
        this.resourceMap.delete(connectionId);
    };
    /** Frees browser resources (process) allocated for this connection. */
    ConnectionManager.prototype.freeBrowser = function (connectionId) {
        var resources = this.resourceMap.get(connectionId);
        if (!resources || !resources.browserProcess) {
            return;
        }
        if (typeof resources.browserProcess === 'function') {
            resources.browserProcess();
        }
        else {
            resources.browserProcess.kill(os_1.constants.signals.SIGINT); // use SIGINT to interrupt/cancel
        }
        resources.browserProcess = null;
    };
    /**
     * @param extractorIds Kills and deregister the specified extractor processes. If omitted, it will remove all extractors.
     * @returns Array of the extractor IDs that are terminated.
     */
    ConnectionManager.prototype.freeExtractors = function (connectionId, extractorIds) {
        var _this = this;
        var resource = this.resourceMap.get(connectionId);
        if (resource === undefined) {
            eps_logger_1.Logger.warn("The connectionID ".concat(connectionId, " couldn't be found."));
            return [];
        }
        var freedExtractorIds = [];
        var processesIds;
        if (extractorIds === undefined) {
            eps_logger_1.Logger.info("Terminating all extractor processes for connectionId ".concat(connectionId));
            processesIds = resource.extractorProcesses
                .map(function (p) { return p.pid; })
                .filter(function (pid) {
                return typeof pid === 'number';
            });
        }
        else {
            eps_logger_1.Logger.info("About to terminate extractor processes with IDs ".concat(extractorIds, " for connectionId ").concat(connectionId));
            processesIds = Object.keys(resource.extractorIdToPid)
                .filter(function (eid) { return extractorIds.includes(eid); })
                .map(function (eid) { return resource.extractorIdToPid[eid]; });
        }
        processesIds.forEach(function (pid) {
            _this.processManager.stopChildProcess(pid);
            resource.extractorProcesses = resource.extractorProcesses.filter(function (p) { return p.pid !== pid; });
            var extractorId = Object.keys(resource.extractorIdToPid).find(function (eid) { return resource.extractorIdToPid[eid] === pid; });
            if (extractorId !== undefined) {
                delete resource.extractorIdToPid[extractorId];
                freedExtractorIds.push(extractorId);
            }
        });
        // Clean up Extractor Temporary Data Directory
        freedExtractorIds.forEach(function (id) {
            _this.tempDataManager.deleteExtractorTempDataDir(id);
        });
        return freedExtractorIds;
    };
    /** Gets if the connection has been submitted, that is, the interactive phase has completed with submit. */
    ConnectionManager.prototype.isSubmitted = function (connectionId) {
        var resource = this.resourceMap.get(connectionId);
        if (!resource) {
            return false;
        }
        return resource.isSubmitted;
    };
    /** Sets if the connection has been submitted, that is, the interactive phase has completed with submit. */
    ConnectionManager.prototype.setIsSubmitted = function (connectorId, isSubmitted) {
        var resources = this.getOrAddResources(connectorId);
        resources.isSubmitted = isSubmitted;
    };
    /** Gets the distinct, unique connector name used by this connection id. */
    ConnectionManager.prototype.distinctConnectorName = function (connectionId) {
        var resource = this.resourceMap.get(connectionId);
        return resource === null || resource === void 0 ? void 0 : resource.distinctConnectorName;
    };
    ConnectionManager.prototype.tacoPath = function (connectionId) {
        var resource = this.resourceMap.get(connectionId);
        return resource === null || resource === void 0 ? void 0 : resource.tacoPath;
    };
    /** Sets the distinct, unique connector name used by this connection id. */
    ConnectionManager.prototype.setDistinctConnectorNameAndPath = function (connectionId, distinctConnectorName, tacoPath) {
        var resources = this.getOrAddResources(connectionId);
        resources.distinctConnectorName = distinctConnectorName;
        resources.tacoPath = tacoPath;
    };
    return ConnectionManager;
}());
exports.ConnectionManager = ConnectionManager;
