// @tableau/taco-toolkit/dist/eps/main/components/render-ui-launcher.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectorAppLauncher = void 0;
var child_process_1 = require("child_process");
var fs_1 = require("fs");
var os_1 = __importStar(require("os"));
var path_1 = __importDefault(require("path"));
var eps_logger_1 = require("../../logging/eps-logger");
var api_utils_1 = require("../../utils/api-utils");
var app_global_1 = __importDefault(require("../core/app-global"));
var browser_detector_1 = require("./browser-detector");
var chromeUserDir = path_1.default.join(os_1.default.tmpdir(), 'tableau-temp', 'eps', 'eps-chrome-tmp-dir');
var STOP_SIGNALS = ['SIGTERM', 'SIGINT', 'close', 'exit'];
var ConnectorAppLauncher = /** @class */ (function () {
    function ConnectorAppLauncher(connectionId, url, position, size) {
        this.proc = null;
        this.connectionId = connectionId;
        this.url = url;
        this.position = position;
        this.size = size;
    }
    // TODO: this should be consolidated with browser detection
    ConnectorAppLauncher.prototype.getArgs = function (browserName) {
        if ((0, browser_detector_1.isChromium)(browserName)) {
            return [
                "--app=".concat(this.url),
                '--chrome-frame',
                '--site-per-process',
                '--no-first-run',
                '--new-window',
                "--window-size=".concat(this.size.width, ",").concat(this.size.height),
                "--window-position=".concat(this.position.left, ",").concat(this.position.top),
                // Note that the following 2 options are sort of correlated.
                // Using incognito mode for chrome to avoid Chrome using any user profile, e.g. password manager.
                // Currently, we are providing different paths with --user - data - dir per session, so in theory,
                // the users cannot re - user any saved information in the profile.
                '--incognito',
                // Although we use incognito mode, the --user-data-dir is still needed at this point. Without this,
                // the 'close' event will be emitted once the browser is launched (this might be only on Mac).
                // I guess this is because 'close' is based on the process stdio. If this option is absent, the stdio
                // will not stay alive after the browser launch.
                // TODO: hopefully, we can have a better solution to wait for the browser process.
                "--user-data-dir=".concat(chromeUserDir),
            ];
        }
        if ((0, browser_detector_1.isFirefox)(browserName)) {
            return []; //['-kiosk'] //-kiosk was the only arg worked by firefox
        }
        return []; //safari does not accept any args
    };
    ConnectorAppLauncher.prototype.start = function () {
        return __awaiter(this, void 0, void 0, function () {
            var browser, args;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.deleteUserDir();
                        eps_logger_1.Logger.info("Launching browser with URL: ".concat((0, api_utils_1.maskURL)(this.url)));
                        return [4 /*yield*/, (0, browser_detector_1.detectBrowser)()];
                    case 1:
                        browser = _a.sent();
                        if (!browser || !browser.name || !browser.path) {
                            throw new Error('No browser was detected!');
                        }
                        this.registerStopSignalsListener();
                        eps_logger_1.Logger.info("launching ".concat(browser.path, " version ").concat(browser.version));
                        args = this.getArgs(browser.name);
                        this.proc = (0, child_process_1.spawn)(browser.path, args, {
                            detached: true,
                            stdio: 'ignore',
                        });
                        app_global_1.default.connectionManager.setBrowserProcess(this.connectionId, this.proc);
                        return [4 /*yield*/, waitForBrowserClose(this.proc)];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    ConnectorAppLauncher.prototype.stop = function () {
        eps_logger_1.Logger.debug("Stopping ".concat(ConnectorAppLauncher.name));
        if (this.proc) {
            eps_logger_1.Logger.debug("Killing process ".concat(this.proc.pid));
            this.proc.kill(os_1.constants.signals.SIGTERM);
        }
        this.deleteUserDir();
    };
    /**
     * Remove chrome user directory. Precautionary check to make sure we aren't deleting a random folder
     */
    ConnectorAppLauncher.prototype.deleteUserDir = function () {
        eps_logger_1.Logger.info("Clean up user-dir for Chromium if exists: ".concat(chromeUserDir));
        if (chromeUserDir.indexOf('eps-chrome-tmp-dir') !== -1 && (0, fs_1.existsSync)(chromeUserDir)) {
            eps_logger_1.Logger.info("Found temp chromium directory. Removing: ".concat(chromeUserDir));
            (0, fs_1.rmSync)(chromeUserDir, { recursive: true });
        }
    };
    ConnectorAppLauncher.prototype.registerStopSignalsListener = function () {
        var _this = this;
        STOP_SIGNALS.forEach(function (signal) {
            process.on(signal, function () {
                eps_logger_1.Logger.info("Main process received signal '".concat(signal, "', stopping ").concat(ConnectorAppLauncher.name));
                _this.stop();
            });
        });
    };
    return ConnectorAppLauncher;
}());
exports.ConnectorAppLauncher = ConnectorAppLauncher;
function waitForBrowserClose(childProcess) {
    return new Promise(function (resolve) {
        childProcess.on('close', function (code) {
            if (code !== 0) {
                eps_logger_1.Logger.warn("Browser process was closed with code ".concat(code));
            }
            else {
                eps_logger_1.Logger.info('Browser process was closed');
            }
            resolve();
        });
    });
}
