// @tableau/taco-toolkit/dist/cli/taco-commands/build/build.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildTaco = void 0;
var fs_1 = __importDefault(require("fs"));
var path_1 = __importDefault(require("path"));
var error_codes_1 = require("../../../shared/errors/error-codes");
var error_source_1 = require("../../../shared/errors/error-source");
var status_code_1 = require("../../../shared/errors/status-code");
var clean_1 = require("../clean/clean");
var command_1 = require("../utils/command");
var connector_1 = require("../utils/connector");
var constants_1 = require("../utils/constants");
var logging_1 = require("../utils/logging");
var npm_command_1 = require("../utils/npm-command");
var install_1 = require("./install");
var build_1 = require("../../taco-scripts/build/build");
function buildTaco(config, state) {
    return __awaiter(this, void 0, void 0, function () {
        var tacoRootPath, doClean, params, result, error;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    tacoRootPath = config.tacoRootPath, doClean = config.doClean;
                    if (!doClean) return [3 /*break*/, 2];
                    return [4 /*yield*/, (0, clean_1.cleanTaco)({
                            cleanOptional: false,
                            cleanLog: false,
                            tacoRootPath: tacoRootPath,
                        }, state)];
                case 1:
                    _a.sent();
                    _a.label = 2;
                case 2:
                    // If npm dependency packages are removed, auto install them before build
                    if (!isDependenciesInstalled(tacoRootPath)) {
                        (0, install_1.installDependencies)(tacoRootPath, state);
                    }
                    // To avoid large radius impact on the existing code, link to the globally installed taco-toolkit
                    // when the dependency is missing in the package.json
                    if ((0, connector_1.isOldDependencySetup)(tacoRootPath)) {
                        (0, install_1.installTacoToolkit)(tacoRootPath);
                        (0, build_1.buildConnector)(tacoRootPath, state);
                        return [2 /*return*/];
                    }
                    params = getBuildScriptParams(state.verbose);
                    result = (0, npm_command_1.executeNpmCmd)(params, {
                        cwd: (0, connector_1.getConnectorFolderPath)(tacoRootPath),
                        stdio: 'inherit',
                    });
                    if ((0, command_1.isErrorResult)(result)) {
                        error = result.error.error;
                        (0, logging_1.logFatalError)({
                            statusCode: status_code_1.StatusCode.Cancelled,
                            errorSource: error_source_1.ErrorSources.System,
                            errorCode: error_codes_1.ErrorCodes.Z0TX352C,
                            message: 'Failed to build connector',
                            error: error,
                        });
                        process.exit(constants_1.TACO_COMMAND_ERROR_EXIT_CODE);
                    }
                    return [2 /*return*/];
            }
        });
    });
}
exports.buildTaco = buildTaco;
function isDependenciesInstalled(tacoRootPath) {
    var nodeModulesPath = path_1.default.join((0, connector_1.getConnectorFolderPath)(tacoRootPath), 'node_modules');
    return fs_1.default.existsSync(nodeModulesPath);
}
function getBuildScriptParams(verbose) {
    var params = ['run', 'build'];
    if (verbose) {
        params.push('--', '--verbose');
    }
    return params;
}
