// @tableau/taco-toolkit/dist/cli/taco-commands/build/install.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.installTacoToolkit = exports.installDependencies = void 0;
var command_1 = require("../utils/command");
var error_codes_1 = require("../../../shared/errors/error-codes");
var error_source_1 = require("../../../shared/errors/error-source");
var status_code_1 = require("../../../shared/errors/status-code");
var connector_1 = require("../utils/connector");
var logging_1 = require("../utils/logging");
var console_1 = require("../utils/console");
var constants_1 = require("../utils/constants");
var npm_command_1 = require("../utils/npm-command");
function installDependencies(tacoRootPath, state) {
    var spinner = state.spinner;
    var connectorFolderPath = (0, connector_1.getConnectorFolderPath)(tacoRootPath);
    spinner.start('Installing dependency packages. It may take a few minutes...');
    var result = (0, npm_command_1.executeNpmCmd)(['install', '--loglevel=error', '--no-fund'], {
        cwd: connectorFolderPath,
    });
    if ((0, command_1.isErrorResult)(result)) {
        var _a = result.error, stderr = _a.stderr, stdout = _a.stdout, error = _a.error;
        logging_1.logger.debug("Execution failure: failed to install npm dependencies.\nstdout: \"".concat(stdout, "\"\nstderr: \"").concat(stderr, "\""));
        spinner.fail('Failed to install npm dependencies');
        (0, logging_1.logFatalError)({
            statusCode: status_code_1.StatusCode.Cancelled,
            errorSource: error_source_1.ErrorSources.Client,
            errorCode: error_codes_1.ErrorCodes.WQX8WEDW,
            message: "Failed to install npm dependencies.\nMore information: ".concat(stderr),
            error: error,
        });
        process.exit(constants_1.TACO_COMMAND_ERROR_EXIT_CODE);
    }
    spinner.success('Successfully installed npm dependencies!');
}
exports.installDependencies = installDependencies;
// TODO: taco project should specify taco-toolkit as dependency.
// Once it's done, we should remove this and add a npm package script to create the link for dev.
function installTacoToolkit(tacoRootPath) {
    var connectorFolderPath = (0, connector_1.getConnectorFolderPath)(tacoRootPath);
    // TODO: warn here for the projects that don't declare taco-toolkit in package.json
    // Currently, we have too many existing useages. Let's add it after some cleanup.
    // As a fallback, link to the global taco-toolkit, if the dependency is not declared in package.json
    var result = (0, npm_command_1.executeNpmCmd)(['link', '@tableau/taco-toolkit', '--loglevel=error'], {
        cwd: connectorFolderPath,
    });
    if ((0, command_1.isErrorResult)(result)) {
        var _a = result.error, stderr = _a.stderr, stdout = _a.stdout, error = _a.error;
        logging_1.logger.debug("Execution failure: failed to link taco-toolkit.\nstdout: \"".concat(stdout, "\"\nstderr: \"").concat(stderr, "\""));
        (0, console_1.consoleError)('Failed to link taco-toolkit', error);
        process.exit(constants_1.TACO_COMMAND_ERROR_EXIT_CODE);
    }
}
exports.installTacoToolkit = installTacoToolkit;
