// @tableau/taco-toolkit/dist/eps/main/components/browser-detector/detect-win-browsers.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.detectBrowsers = exports.detectBrowser = exports.getInstallationPath = exports.findBrowserRegistries = exports.findVersion = void 0;
var child_process_1 = require("child_process");
var MAIN_BROWSER_REG = 'HKEY_LOCAL_MACHINE\\SOFTWARE\\WOW6432Node\\Clients\\StartMenuInternet';
var REG_TYPE = 'REG_SZ';
var extractName = function (str) {
    return str.split('\\').slice(-1)[0] || '';
};
var extractPathFromRegistry = function (str) {
    if (!str) {
        return '';
    }
    var path = str.trim().split(REG_TYPE)[1].replace(/\t+/g, '').split(',');
    return path[0].trim();
};
var getSubRegistries = function (data, parentRegistry) {
    if (!data || !parentRegistry) {
        return [];
    }
    return data
        .trim()
        .split('\r\n')
        .filter(function (s) { return s.length > parentRegistry.length; });
};
var findVersion = function (browserPath) {
    browserPath = browserPath.replace(/\\/g, '\\\\');
    var command = "wmic datafile where name=\"".concat(browserPath, "\" get Version /value");
    return new Promise(function (resolve, reject) {
        (0, child_process_1.exec)(command, function (err, stdout, stderr) {
            if (stderr || err) {
                reject(err);
            }
            else {
                if (stdout) {
                    var version = stdout.trim().replace('Version=', '');
                    resolve(version);
                }
                else {
                    resolve('');
                }
            }
        });
    });
};
exports.findVersion = findVersion;
var findBrowserRegistries = function (mainRegistryPath) {
    return new Promise(function (resolve, reject) {
        var command = "reg query ".concat(mainRegistryPath);
        (0, child_process_1.exec)(command, function (error, stdout, _stderr) {
            if (error) {
                reject(error);
            }
            if (stdout) {
                var subPaths = getSubRegistries(stdout, mainRegistryPath);
                resolve(subPaths);
            }
            else {
                reject();
            }
        });
    });
};
exports.findBrowserRegistries = findBrowserRegistries;
var getInstallationPath = function (browserRegistry) {
    return new Promise(function (resolve, reject) {
        var fullPath = "\"".concat(browserRegistry, "\\DefaultIcon\"");
        (0, child_process_1.exec)("reg query ".concat(fullPath), function (error, stdout, _stderr) {
            if (error) {
                reject(error);
            }
            else {
                var browser = {
                    name: extractName(browserRegistry),
                    path: extractPathFromRegistry(stdout),
                };
                if (browser.name && browser.path) {
                    resolve(browser);
                }
                else {
                    resolve(null);
                }
            }
        });
    });
};
exports.getInstallationPath = getInstallationPath;
//it turns out the registry name for some browsers are diffrent on windows e.g. 'IEXPLORE.EXE', 'Firefox-308046B0AF4A39CB', 'Brave'
var toConsistentName = function (registryName) {
    var name = registryName.toLocaleLowerCase();
    if (name.includes('brave')) {
        return 'Brave Browser';
    }
    if (name.includes('firefox')) {
        return 'Firefox';
    }
    if (name.includes('iexplore')) {
        return 'Internet Explorer';
    }
    return registryName;
};
var detectBrowser = function (browserRegistry) { return __awaiter(void 0, void 0, void 0, function () {
    var browser, version, name;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, (0, exports.getInstallationPath)(browserRegistry)];
            case 1:
                browser = _a.sent();
                if (browser == null || !browser.path) {
                    return [2 /*return*/, null];
                }
                return [4 /*yield*/, (0, exports.findVersion)(browser.path)];
            case 2:
                version = _a.sent();
                name = toConsistentName(browser.name);
                return [2 /*return*/, {
                        path: browser.path,
                        version: version,
                        name: name,
                    }];
        }
    });
}); };
exports.detectBrowser = detectBrowser;
var detectBrowsers = function (browserRegistries) { return __awaiter(void 0, void 0, void 0, function () {
    var promises, browserRegistries_1, browserRegistries_1_1, browserRegistry, browsers, detectedBrowsers, browsers_1, browsers_1_1, browser;
    var e_1, _a, e_2, _b;
    return __generator(this, function (_c) {
        switch (_c.label) {
            case 0:
                promises = [];
                try {
                    for (browserRegistries_1 = __values(browserRegistries), browserRegistries_1_1 = browserRegistries_1.next(); !browserRegistries_1_1.done; browserRegistries_1_1 = browserRegistries_1.next()) {
                        browserRegistry = browserRegistries_1_1.value;
                        promises.push((0, exports.detectBrowser)(browserRegistry));
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (browserRegistries_1_1 && !browserRegistries_1_1.done && (_a = browserRegistries_1.return)) _a.call(browserRegistries_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
                return [4 /*yield*/, Promise.allSettled(promises)];
            case 1:
                browsers = _c.sent();
                detectedBrowsers = [];
                try {
                    for (browsers_1 = __values(browsers), browsers_1_1 = browsers_1.next(); !browsers_1_1.done; browsers_1_1 = browsers_1.next()) {
                        browser = browsers_1_1.value;
                        if (browser.status === 'fulfilled' && browser.value != null) {
                            detectedBrowsers.push(browser.value);
                        }
                    }
                }
                catch (e_2_1) { e_2 = { error: e_2_1 }; }
                finally {
                    try {
                        if (browsers_1_1 && !browsers_1_1.done && (_b = browsers_1.return)) _b.call(browsers_1);
                    }
                    finally { if (e_2) throw e_2.error; }
                }
                return [2 /*return*/, detectedBrowsers];
        }
    });
}); };
exports.detectBrowsers = detectBrowsers;
function detectWinInstalledBrowsers(browserNames) {
    return __awaiter(this, void 0, void 0, function () {
        var browserRegistries, detectedBrowsers;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, (0, exports.findBrowserRegistries)(MAIN_BROWSER_REG)];
                case 1:
                    browserRegistries = _a.sent();
                    return [4 /*yield*/, (0, exports.detectBrowsers)(browserRegistries)
                        //return only defined browsers
                    ];
                case 2:
                    detectedBrowsers = _a.sent();
                    //return only defined browsers
                    return [2 /*return*/, detectedBrowsers.filter(function (b) { return browserNames.includes(b.name); })];
            }
        });
    });
}
exports.default = detectWinInstalledBrowsers;
