// @tableau/taco-toolkit/dist/eps/main/components/browser-detector/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.detectBrowser = exports.detectAndPrioritizeAllInstalledBrowsers = exports.prioritize = exports.detectInstalledBrowsers = exports.isWebkit = exports.isFirefox = exports.isChromium = exports.isVersionSupported = exports.isCurrentVersionHigher = void 0;
var os_1 = require("os");
var detect_mac_browsers_1 = __importDefault(require("./detect-mac-browsers"));
var detect_win_browsers_1 = __importDefault(require("./detect-win-browsers"));
var PLATFORM = {
    MAC: 'darwin',
    WINDOWS: 'win32',
};
//TODO: move to config file?
var minSupportedVersions = {
    'Google Chrome': '98.0',
    'Microsoft Edge': '98.0',
    'Brave Browser': '98.0',
    Chromium: '98.0.',
    Safari: '15.0',
    Firefox: '97.0',
};
var browserNamesByPriority = [
    'Google Chrome',
    'Microsoft Edge',
    'Brave Browser',
    'Chromium',
    'Safari',
    'Firefox',
];
var isCurrentVersionHigher = function (currentVersion, minReqVersion) {
    var currentParts = currentVersion.split('.');
    var minParts = minReqVersion.split('.');
    for (var i = 0; i < minParts.length; i++) {
        var cur = currentParts[i] || '0';
        var min = minParts[i] || '0';
        if (Number(cur) > Number(min)) {
            return true;
        }
        if (Number(cur) < Number(min)) {
            return false;
        }
    }
    return true;
};
exports.isCurrentVersionHigher = isCurrentVersionHigher;
var isVersionSupported = function (browser, minSupported) {
    if (minSupported === void 0) { minSupported = minSupportedVersions; }
    if (!browser.version) {
        return false;
    }
    var minReqVersion = minSupported[browser.name];
    if (!minReqVersion) {
        return false;
    }
    return (0, exports.isCurrentVersionHigher)(browser.version, minReqVersion);
};
exports.isVersionSupported = isVersionSupported;
var isChromium = function (browserName) {
    return ['Chromium', 'Google Chrome', 'Microsoft Edge', 'Brave Browser'].includes(browserName);
};
exports.isChromium = isChromium;
var isFirefox = function (browserName) { return ['Firefox'].includes(browserName); };
exports.isFirefox = isFirefox;
var isWebkit = function (browserName) { return ['Safari'].includes(browserName); };
exports.isWebkit = isWebkit;
var detectInstalledBrowsers = function (browserNames) { return __awaiter(void 0, void 0, void 0, function () {
    var os;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                os = (0, os_1.platform)();
                if (!(os === PLATFORM.MAC)) return [3 /*break*/, 2];
                return [4 /*yield*/, (0, detect_mac_browsers_1.default)(browserNames)];
            case 1: return [2 /*return*/, _a.sent()];
            case 2:
                if (!(os === PLATFORM.WINDOWS)) return [3 /*break*/, 4];
                return [4 /*yield*/, (0, detect_win_browsers_1.default)(browserNames)];
            case 3: return [2 /*return*/, _a.sent()];
            case 4: return [2 /*return*/, null];
        }
    });
}); };
exports.detectInstalledBrowsers = detectInstalledBrowsers;
var prioritize = function (browsers, browserNames) {
    var e_1, _a;
    var prioritizedBrowsers = [];
    var _loop_1 = function (browserName) {
        var browser = browsers.find(function (b) { return b.name === browserName; });
        if (browser && (0, exports.isVersionSupported)(browser)) {
            prioritizedBrowsers.push(browser);
        }
    };
    try {
        for (var browserNames_1 = __values(browserNames), browserNames_1_1 = browserNames_1.next(); !browserNames_1_1.done; browserNames_1_1 = browserNames_1.next()) {
            var browserName = browserNames_1_1.value;
            _loop_1(browserName);
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (browserNames_1_1 && !browserNames_1_1.done && (_a = browserNames_1.return)) _a.call(browserNames_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return prioritizedBrowsers;
};
exports.prioritize = prioritize;
var detectAndPrioritizeAllInstalledBrowsers = function (browserNames) {
    if (browserNames === void 0) { browserNames = browserNamesByPriority; }
    return __awaiter(void 0, void 0, void 0, function () {
        var detectedBrowsers;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, (0, exports.detectInstalledBrowsers)(browserNames)];
                case 1:
                    detectedBrowsers = _a.sent();
                    if (!detectedBrowsers) {
                        return [2 /*return*/, null];
                    }
                    return [2 /*return*/, (0, exports.prioritize)(detectedBrowsers, browserNames)];
            }
        });
    });
};
exports.detectAndPrioritizeAllInstalledBrowsers = detectAndPrioritizeAllInstalledBrowsers;
var detectBrowser = function () { return __awaiter(void 0, void 0, void 0, function () {
    var detectedBrowsers;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, (0, exports.detectAndPrioritizeAllInstalledBrowsers)(browserNamesByPriority)];
            case 1:
                detectedBrowsers = _a.sent();
                if (!detectedBrowsers) {
                    return [2 /*return*/, null];
                }
                return [2 /*return*/, detectedBrowsers[0]];
        }
    });
}); };
exports.detectBrowser = detectBrowser;
