// @tableau/taco-toolkit/dist/eps/main/core/app-global.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var eps_logger_1 = require("../../logging/eps-logger");
var port_manager_1 = require("../../modules/port-manager");
var temp_data_manager_1 = require("../../modules/temp-data-manager");
var connection_manager_1 = require("../components/connection-manager");
var process_manager_1 = require("../components/process-manager");
var taco_initializer_1 = require("../components/taco-initializer");
var AppGlobal = /** @class */ (function () {
    function AppGlobal(config) {
        var callerType = config.callerType, epsInstanceId = config.epsInstanceId, epsTempDirPath = config.epsTempDirPath, portRange = config.portRange, reservedPorts = config.reservedPorts, connectorAppDir = config.connectorAppDir, connectorHandlerDir = config.connectorHandlerDir, i18nConfig = config.i18nConfig;
        var tempDataManager = new temp_data_manager_1.TempDataManager(epsInstanceId, epsTempDirPath);
        this._processManager = new process_manager_1.EPSProcessManager(tempDataManager);
        this._portManager = new port_manager_1.PortManager(portRange[0], portRange[1], reservedPorts);
        this._connectionManager = new connection_manager_1.ConnectionManager(this._processManager, tempDataManager);
        this._tacoInitializer = new taco_initializer_1.TacoInitializer(this._connectionManager, {
            callerType: callerType,
            appDir: connectorAppDir,
            handlersDir: connectorHandlerDir,
            i18n: i18nConfig,
        });
    }
    AppGlobal.init = function (config) {
        if (this.instance) {
            eps_logger_1.Logger.warn('AppGlobal has already been initialized. Skipping...');
            return;
        }
        this.instance = new AppGlobal(config);
        // in monolith context, this is Tableau app process
        this.instance._processManager.addPID(config.parentPid, 'eps-parent', process_manager_1.ProcessType.parent);
        this.instance._processManager.start();
    };
    Object.defineProperty(AppGlobal, "processManager", {
        get: function () {
            assertInit(this.instance);
            return this.instance._processManager;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AppGlobal, "portManager", {
        get: function () {
            assertInit(this.instance);
            return this.instance._portManager;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AppGlobal, "connectionManager", {
        get: function () {
            assertInit(this.instance);
            return this.instance._connectionManager;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AppGlobal, "tacoInitializer", {
        get: function () {
            assertInit(this.instance);
            return this.instance._tacoInitializer;
        },
        enumerable: false,
        configurable: true
    });
    return AppGlobal;
}());
function assertInit(instance) {
    if (!instance) {
        throw new Error('AppGlobal must be initialized.');
    }
}
exports.default = AppGlobal;
