// @tableau/taco-toolkit/dist/eps/main/core/main.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var os_1 = __importDefault(require("os"));
var error_codes_1 = require("../../../shared/errors/error-codes");
var error_source_1 = require("../../../shared/errors/error-source");
var status_code_1 = require("../../../shared/errors/status-code");
var eps_logger_1 = require("../../logging/eps-logger");
var app_global_1 = __importDefault(require("./app-global"));
var app_init_1 = require("./app-utils/app-init");
var command_line_parser_1 = require("./app-utils/command-line-parser");
var env_config_1 = require("./app-utils/env-config");
var eps_app_1 = __importDefault(require("./eps-app"));
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore Ignore file is not under 'rootDir'
var package_json_1 = __importDefault(require("../../../../package.json"));
(function () { return __awaiter(void 0, void 0, void 0, function () {
    var envConfig, optionValues, reservedPorts, i18nConfig, networkOption, connectorAppDir, connectorHandlerDir, epsApp, error_1;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                _a.trys.push([0, 2, , 3]);
                (0, command_line_parser_1.handleLoggerSetup)(process.argv);
                eps_logger_1.Logger.info("Starting eps@".concat(package_json_1.default.tableau.epsVersion, " on ").concat(os_1.default.platform, " from directory '").concat(process.cwd(), "' with nodejs ").concat(process.versions.node, "."));
                envConfig = (0, env_config_1.getEnvConfig)();
                optionValues = (0, command_line_parser_1.parseArgs)(process.argv);
                reservedPorts = (0, app_init_1.getReservedPorts)(optionValues);
                i18nConfig = {
                    language: optionValues.language,
                    locale: optionValues.locale,
                };
                networkOption = {
                    httpProxy: envConfig.httpProxy,
                    caCertPath: envConfig.caCertPath,
                };
                connectorAppDir = optionValues.staticServerUiDir, connectorHandlerDir = optionValues.staticServerHandlersDir;
                // Set child app env configs
                (0, env_config_1.setChildAppEnvConfig)(optionValues, networkOption);
                // Intialize EPS app global managers
                app_global_1.default.init({
                    parentPid: optionValues.callerPid,
                    callerType: optionValues.callerType,
                    epsInstanceId: optionValues.epsInstanceId,
                    epsTempDirPath: optionValues.epsTempDirPath,
                    portRange: optionValues.portRange,
                    reservedPorts: reservedPorts,
                    connectorAppDir: connectorAppDir,
                    connectorHandlerDir: connectorHandlerDir,
                    i18nConfig: i18nConfig,
                });
                eps_logger_1.Logger.debug("Initiating ".concat(eps_app_1.default.name));
                epsApp = new eps_app_1.default({
                    callerId: optionValues.callerId,
                    callerType: optionValues.callerType,
                    epsInstanceId: optionValues.epsInstanceId,
                    staticServerPort: optionValues.staticServerPort,
                    mainApiPort: optionValues.mainPort,
                    grpcServerPort: optionValues.grpcServerPort,
                    portFilePath: optionValues.portFile,
                    connectorAppDir: connectorAppDir,
                    featureFlags: optionValues.featureFlags,
                    networkOption: networkOption,
                    i18n: i18nConfig,
                });
                // Start all EPS services
                return [4 /*yield*/, epsApp.startServices()];
            case 1:
                // Start all EPS services
                _a.sent();
                return [3 /*break*/, 3];
            case 2:
                error_1 = _a.sent();
                eps_logger_1.Logger.logEPSError(status_code_1.StatusCode.Unknown, error_source_1.ErrorSources.System, error_codes_1.ErrorCodes.T1GDF8UC, error_1);
                return [3 /*break*/, 3];
            case 3: return [2 /*return*/];
        }
    });
}); })();
