// @tableau/taco-toolkit/dist/eps/logging/eps-logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLoggerConfig = exports.Logger = exports.EPS_LOG_FILE_NAME = void 0;
var path_1 = __importDefault(require("path"));
var log_level_1 = require("../../shared/logger/log-level");
var logger_1 = require("../../shared/logger/logger");
var file_1 = require("../../shared/logger/writer/file");
var nodejs_console_1 = require("../../shared/logger/writer/nodejs-console");
var file_path_1 = require("../utils/file-path");
var eps_log_serializer_1 = require("./eps-log-serializer");
var app_log_option_1 = require("../types/app-log-option");
exports.EPS_LOG_FILE_NAME = 'eps';
var Logger = /** @class */ (function () {
    function Logger() {
    }
    Logger.init = function (epsInstanceId, config) {
        if (Logger.instance !== undefined) {
            Logger.warn('Logger has already been initialized! Skipping...');
            return;
        }
        var writers = getWriters(epsInstanceId, config);
        Logger.instance = new logger_1.Logger({
            writers: writers,
            serializer: eps_log_serializer_1.serialize,
        });
    };
    Logger.info = function (message, options) {
        if (Logger.instance) {
            Logger.instance.info(message, options);
        }
    };
    Logger.error = function (message, options) {
        if (Logger.instance) {
            Logger.instance.error(message, options);
        }
    };
    Logger.warn = function (message, options) {
        if (Logger.instance) {
            Logger.instance.warn(message, options);
        }
    };
    Logger.debug = function (message, options) {
        if (Logger.instance) {
            Logger.instance.debug(message, options);
        }
    };
    Logger.logEPSError = function (statusCode, errorSource, epsErrorCode, message, options) {
        var errorInfo = {
            'Status Code': statusCode,
            'Error Source': errorSource,
            'EPS Error Code': epsErrorCode,
        };
        if (message instanceof Error) {
            // if message is an error, log it here so that the original stack can be logged.
            Logger.error(message);
            errorInfo['Error Message'] = message.message;
        }
        else {
            errorInfo['Error Message'] = message;
        }
        Logger.error(errorInfo, options);
    };
    return Logger;
}());
exports.Logger = Logger;
function getEPSLogDir(rootDir, epsInstanceId) {
    // Currently, the log dir path may be sent via a request from monolith.
    // In some cases, the path may contain quotes...
    return path_1.default.join((0, file_path_1.trimQuotes)(rootDir), 'EPS', epsInstanceId);
}
function isConsoleOnlyLoggerConfig(config) {
    return config.consoleLogLevel !== undefined && !('fileLogDirPath' in config);
}
function getWriters(epsInstanceId, config) {
    var sourceName = config.sourceName, consoleLogLevel = config.consoleLogLevel;
    if (isConsoleOnlyLoggerConfig(config)) {
        return [
            new nodejs_console_1.NodeJsConsoleWriter({
                sourceName: sourceName,
                level: consoleLogLevel,
                includeTimeStamp: true,
            }),
        ];
    }
    var fileLogLevel = config.fileLogLevel, fileLogDirPath = config.fileLogDirPath, filename = config.filename;
    var writers = [
        new file_1.FileWriter({
            filename: filename,
            sourceName: sourceName,
            level: fileLogLevel ? fileLogLevel : log_level_1.LogLevel.DEBUG,
            dirPath: getEPSLogDir(fileLogDirPath, epsInstanceId),
        }),
    ];
    if (consoleLogLevel) {
        writers.push(new nodejs_console_1.NodeJsConsoleWriter({
            sourceName: sourceName,
            level: consoleLogLevel,
            includeTimeStamp: true,
        }));
    }
    return writers;
}
function createLoggerConfig(sourceName, appLogOption) {
    if ((0, app_log_option_1.isConsoleOnlyAppLogOption)(appLogOption)) {
        return {
            consoleLogLevel: appLogOption.consoleLogLevel,
            sourceName: sourceName,
        };
    }
    var logDir = appLogOption.logDir, logLevel = appLogOption.logLevel, consoleLogLevel = appLogOption.consoleLogLevel;
    var loggerConfig = {
        sourceName: sourceName,
        filename: exports.EPS_LOG_FILE_NAME,
        fileLogDirPath: logDir,
        fileLogLevel: logLevel,
    };
    if (consoleLogLevel) {
        loggerConfig.consoleLogLevel = consoleLogLevel;
    }
    return loggerConfig;
}
exports.createLoggerConfig = createLoggerConfig;
