// @tableau/taco-toolkit/dist/cli/taco-commands/utils/command.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isErrorResult = exports.isSuccessResult = exports.executeCmd = exports.exitIfAnyMissingPath = exports.exitIfPathExists = exports.exitIfNonExistPath = exports.resolvePath = void 0;
var child_process_1 = require("child_process");
var fs_1 = __importDefault(require("fs"));
var path_1 = __importDefault(require("path"));
var console_1 = require("./console");
var logging_1 = require("./logging");
/**
 * If the input path is a relative path, it resolves the path from
 * current working directory.
 *
 * @returns an absolute path
 */
function resolvePath(rootPath) {
    if (path_1.default.isAbsolute(rootPath)) {
        return rootPath;
    }
    return path_1.default.resolve(process.cwd(), rootPath);
}
exports.resolvePath = resolvePath;
/**
 * The function is designated to valid if a path from user input exists.
 * Note: this function only displays messages in console.
 * It does not log any info into file.
 *
 * @param filePath a relative path will be resolved relative to the current working directory
 * @param errorMessage error message to display before exit
 */
function exitIfNonExistPath(filePath, errorMessage) {
    if (!fs_1.default.existsSync(filePath)) {
        (0, console_1.consoleError)(errorMessage);
        process.exit(1);
    }
}
exports.exitIfNonExistPath = exitIfNonExistPath;
function exitIfPathExists(filePath, errorMessage) {
    if (fs_1.default.existsSync(filePath)) {
        (0, console_1.consoleError)(errorMessage);
        process.exit(1);
    }
}
exports.exitIfPathExists = exitIfPathExists;
function exitIfAnyMissingPath(paths, errorMessage) {
    var missingPaths = paths.filter(function (filePath) { return !fs_1.default.existsSync(filePath); });
    if (missingPaths.length > 0) {
        (0, console_1.consoleError)("".concat(errorMessage, ". Missing paths: ").concat(missingPaths));
        process.exit(1);
    }
}
exports.exitIfAnyMissingPath = exitIfAnyMissingPath;
/**
 * Invoking a global command with child_process.spawnSync
 */
function executeCmd(cmd, params, options) {
    var paramsStr = params ? params.join(' ') : '';
    var optionsStr = options ? JSON.stringify(options) : '';
    logging_1.logger.debug("Invoking spawnSync: ".concat(cmd, " ").concat(paramsStr, " ").concat(optionsStr));
    var _a = (0, child_process_1.spawnSync)(cmd, params, options), status = _a.status, error = _a.error, stderr = _a.stderr, stdout = _a.stdout;
    if (status !== 0 || error !== undefined) {
        return {
            state: 'error',
            error: {
                status: status,
                error: error,
                stderr: String(stderr),
                stdout: String(stdout),
            },
        };
    }
    return {
        state: 'success',
        data: String(stdout),
    };
}
exports.executeCmd = executeCmd;
function isSuccessResult(result) {
    return result.state === 'success';
}
exports.isSuccessResult = isSuccessResult;
function isErrorResult(result) {
    return result.state === 'error';
}
exports.isErrorResult = isErrorResult;
