// @tableau/taco-toolkit/dist/cli/taco-commands/utils/connector-json-schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.schema = void 0;
exports.schema = {
    type: 'object',
    required: ['name', 'version', 'tableau-version', 'vendor', 'permission', 'window', 'auth'],
    additionalProperties: false,
    properties: {
        name: {
            description: 'connector name',
            type: 'string',
            pattern: '^[\\w\\-_ ]+$',
            maxLength: 29,
        },
        displayName: {
            description: 'connector display name',
            type: 'string',
            minLength: 3,
            maxLength: 29,
            nullable: true, // ajv requires optional property to be nullable
        },
        version: {
            description: 'connector version',
            pattern: '^[\\w\\-_\\.\\+]+$',
            type: 'string',
            maxLength: 30,
        },
        'tableau-version': {
            type: 'object',
            properties: {
                min: {
                    description: 'minimum supported Tableau version',
                    type: 'string',
                    maxLength: 256,
                },
            },
            required: ['min'],
        },
        vendor: {
            type: 'object',
            properties: {
                name: {
                    description: 'creator name/title',
                    type: 'string',
                    maxLength: 256,
                },
                'support-link': {
                    description: 'support link for connector',
                    type: 'string',
                    maxLength: 256,
                },
                email: {
                    description: 'contact email for connector',
                    type: 'string',
                    maxLength: 256,
                },
            },
            required: ['name', 'support-link', 'email'],
        },
        auth: {
            type: 'object',
            discriminator: { propertyName: 'type' },
            required: ['type'],
            oneOf: [
                {
                    properties: {
                        type: { const: 'none' },
                    },
                },
                {
                    properties: {
                        type: { const: 'custom' },
                    },
                },
                {
                    // Refer to: https://tableau.github.io/connector-plugin-sdk/docs/oauth
                    required: ['oauth'],
                    properties: {
                        type: { const: 'oauth2' },
                        oauth: {
                            type: 'object',
                            properties: {
                                oauthConfigId: {
                                    description: 'id for the oauth config',
                                    type: 'string',
                                    maxLength: 256,
                                },
                                configLabel: {
                                    description: 'the label we display for this oauth config in the UI',
                                    type: 'string',
                                    maxLength: 256,
                                },
                                clientIdDesktop: {
                                    description: 'the client ID registered for Tableau Desktop',
                                    type: 'string',
                                    maxLength: 256,
                                },
                                clientSecretDesktop: {
                                    description: 'the client secret registered for Tableau Desktop',
                                    type: 'string',
                                    maxLength: 256,
                                },
                                redirectUrisDesktop: {
                                    description: 'redirect URIs for Desktop',
                                    type: 'array',
                                    items: {
                                        type: 'string',
                                    },
                                },
                                authUri: {
                                    description: 'authorization endpoint URI',
                                    type: 'string',
                                    maxLength: 256,
                                },
                                tokenUri: {
                                    description: 'token endpoint URI',
                                    type: 'string',
                                    maxLength: 256,
                                },
                                userInfoUri: {
                                    description: 'user Info endpoint URI',
                                    type: 'string',
                                    maxLength: 256,
                                },
                                instanceUrlSuffix: {
                                    description: 'the suffix for oauth endpoint(instanceurl)',
                                    type: 'string',
                                    maxLength: 256,
                                },
                                instanceUrlValidationRegex: {
                                    description: 'regex validation pattern for oauth endpoint(instanceurl)',
                                    type: 'string',
                                    maxLength: 256,
                                },
                                defaultInstanceUrl: {
                                    description: 'default oauth endpoint(instanceurl)',
                                    type: 'string',
                                    maxLength: 256,
                                },
                                scopes: {
                                    description: 'scopes',
                                    type: 'array',
                                    items: {
                                        type: 'string',
                                    },
                                },
                                capabilities: {
                                    description: 'defines how OAuth flow behaves',
                                    type: 'object',
                                    properties: {
                                        OAUTH_CAP_SUPPORTS_CUSTOM_DOMAIN: {
                                            type: 'boolean',
                                        },
                                        OAUTH_CAP_REQUIRE_PKCE: {
                                            type: 'boolean',
                                        },
                                        OAUTH_CAP_PKCE_REQUIRES_CODE_CHALLENGE_METHOD: {
                                            type: 'boolean',
                                        },
                                        OAUTH_CAP_SUPPORTS_STATE: {
                                            type: 'boolean',
                                        },
                                        OAUTH_CAP_REQUIRES_VERIFY_STATE: {
                                            type: 'boolean',
                                        },
                                        OAUTH_CAP_GET_USERNAME_USES_POST_REQUEST: {
                                            type: 'boolean',
                                        },
                                        OAUTH_CAP_CLIENT_SECRET_IN_URL_QUERY_PARAM: {
                                            type: 'boolean',
                                        },
                                        OAUTH_CAP_FIXED_PORT_IN_CALLBACK_URL: {
                                            type: 'boolean',
                                        },
                                        OAUTH_CAP_SUPPORTS_HTTP_SCHEME_LOOPBACK_REDIRECT_URLS: {
                                            type: 'boolean',
                                        },
                                        OAUTH_CAP_REQUIRES_PROMPT_CONSENT: {
                                            type: 'boolean',
                                        },
                                        OAUTH_CAP_REQUIRES_PROMPT_SELECT_ACCOUNT: {
                                            type: 'boolean',
                                        },
                                        OAUTH_CAP_SUPPORTS_GET_USERINFO_FROM_ID_TOKEN: {
                                            type: 'boolean',
                                        },
                                    },
                                },
                                accessTokenResponseMaps: {
                                    description: 'key value pair that maps an initial token request response attribute to Tableau recognized attribute',
                                    type: 'object',
                                    required: ['ACCESSTOKEN', 'REFRESHTOKEN', 'username'],
                                    additionalProperties: true,
                                    properties: {
                                        ACCESSTOKEN: {
                                            type: 'string',
                                            maxLength: 256,
                                        },
                                        REFRESHTOKEN: {
                                            type: 'string',
                                            maxLength: 256,
                                        },
                                        'access-token-issue-time': {
                                            type: 'string',
                                            maxLength: 256,
                                        },
                                        'access-token-expires-in': {
                                            type: 'string',
                                            maxLength: 256,
                                        },
                                        username: {
                                            type: 'string',
                                            maxLength: 256,
                                        },
                                    },
                                },
                                refreshTokenResponseMaps: {
                                    description: 'key value pair that maps a refresh token request response attribute to Tableau recognized attribute',
                                    type: 'object',
                                    required: ['ACCESSTOKEN'],
                                    additionalProperties: true,
                                    properties: {
                                        ACCESSTOKEN: {
                                            type: 'string',
                                            maxLength: 256,
                                        },
                                    },
                                },
                            },
                            required: ['authUri', 'tokenUri', 'scopes', 'accessTokenResponseMaps'],
                        },
                    },
                },
            ],
        },
        permission: {
            type: 'object',
            required: ['api'],
            properties: {
                api: {
                    type: 'object',
                    required: [],
                    minProperties: 1,
                    patternProperties: {
                        '.*': {
                            type: 'array',
                            uniqueItems: true,
                            items: {
                                type: 'string',
                                enum: ['GET', 'POST', 'OPTIONS', 'PATCH', 'PUT', 'DELETE', 'HEAD'],
                            },
                        },
                    },
                },
                disk: {
                    type: 'object',
                    nullable: true,
                    required: ['path', 'size'],
                    properties: {
                        path: {
                            type: 'array',
                            items: {
                                type: 'string',
                            },
                        },
                        size: {
                            type: 'string',
                            maxLength: 256,
                        },
                    },
                },
                libs: {
                    type: 'array',
                    nullable: true,
                    items: {
                        type: 'string',
                    },
                },
            },
        },
        window: {
            type: 'object',
            properties: {
                height: {
                    type: 'number',
                },
                width: {
                    type: 'number',
                },
            },
            required: ['height', 'width'],
        },
        config: {
            description: 'additional configurations from the users',
            type: 'object',
            additionalProperties: true,
            nullable: true, // ajv requires optional property to be nullable
        },
    },
};
