// @tableau/taco-toolkit/dist/cli/taco-commands/utils/connector.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isValidTacoFile = exports.isOldDependencySetup = exports.getConnectorHandlersPath = exports.getConnectorAppPath = exports.getConnectorJsonPath = exports.getConnectorFolderPath = exports.isTacoRootPath = exports.outputLocationErrorAndExit = exports.exitIfNotTacoRoot = exports.findTacoRootPath = exports.sourceTypes = void 0;
var fs_1 = __importDefault(require("fs"));
var fs_extra_1 = __importDefault(require("fs-extra"));
var path_1 = __importDefault(require("path"));
var console_1 = require("./console");
exports.sourceTypes = ['app', 'handlers'];
function findTacoRootPath() {
    var currentPath = process.cwd();
    var possiblePaths = [
        currentPath,
        path_1.default.dirname(currentPath), // parent path
    ];
    var tacoRootPath = possiblePaths.find(function (possiblePath) { return isTacoRootPath(possiblePath); });
    if (tacoRootPath === undefined) {
        outputLocationErrorAndExit(currentPath);
    }
    return tacoRootPath;
}
exports.findTacoRootPath = findTacoRootPath;
/**
 * Exit with code 1 from program if the given path is not a connector root path.
 */
function exitIfNotTacoRoot(rootPath) {
    if (!isTacoRootPath(rootPath)) {
        outputLocationErrorAndExit(rootPath);
    }
}
exports.exitIfNotTacoRoot = exitIfNotTacoRoot;
function outputLocationErrorAndExit(rootPath) {
    (0, console_1.consoleError)("Cannot locate taco connector.json file at current path, are you in the root directory of a taco connector?\nCurrent path: ".concat(rootPath));
    return process.exit(1);
}
exports.outputLocationErrorAndExit = outputLocationErrorAndExit;
function isTacoRootPath(path) {
    return fs_1.default.existsSync(getConnectorJsonPath(path));
}
exports.isTacoRootPath = isTacoRootPath;
function getConnectorFolderPath(tacoRootPath) {
    return path_1.default.join(tacoRootPath, 'connector');
}
exports.getConnectorFolderPath = getConnectorFolderPath;
function getConnectorJsonPath(tacoRootPath) {
    return path_1.default.join(tacoRootPath, 'connector', 'connector.json');
}
exports.getConnectorJsonPath = getConnectorJsonPath;
function getConnectorAppPath(tacoRootPath) {
    return path_1.default.join(tacoRootPath, 'connector', 'app');
}
exports.getConnectorAppPath = getConnectorAppPath;
function getConnectorHandlersPath(tacoRootPath) {
    return path_1.default.join(tacoRootPath, 'connector', 'handlers');
}
exports.getConnectorHandlersPath = getConnectorHandlersPath;
/**
 * Check if a taco project is using the old npm link setup.
 * Remove this when all existing projects define taco-toolkit as npm dependency.
 */
function isOldDependencySetup(tacoRootPath) {
    var packageJsonPath = path_1.default.join(tacoRootPath, 'connector', 'package.json');
    var packageJson = fs_extra_1.default.readJSONSync(packageJsonPath);
    if (packageJson.dependencies && '@tableau/taco-toolkit' in packageJson.dependencies) {
        return false;
    }
    return true;
}
exports.isOldDependencySetup = isOldDependencySetup;
function isValidTacoFile(tacoFilePath) {
    return path_1.default.extname(tacoFilePath).toLowerCase() === '.taco';
}
exports.isValidTacoFile = isValidTacoFile;
