// @tableau/taco-toolkit/dist/cli/taco-commands/utils/logging.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.logFatalError = exports.initLogger = exports.logger = exports.getLogPath = exports.cliLogLevel = exports.LOG_FILENAME = void 0;
var path_1 = __importDefault(require("path"));
var log_level_1 = require("../../../shared/logger/log-level");
var logger_1 = require("../../../shared/logger/logger");
var nodejs_serializer_1 = require("../../../shared/logger/serializer/nodejs-serializer");
var file_1 = require("../../../shared/logger/writer/file");
var nodejs_console_1 = require("../../../shared/logger/writer/nodejs-console");
exports.LOG_FILENAME = 'cli';
exports.cliLogLevel = {
    debug: log_level_1.LogLevel.DEBUG,
    info: log_level_1.LogLevel.INFO,
};
function getLogLevelFromString(str) {
    return exports.cliLogLevel[str];
}
function getLogPath(tacoRootPath) {
    return path_1.default.join(tacoRootPath, 'log');
}
exports.getLogPath = getLogPath;
exports.logger = new logger_1.Logger({
    writers: [],
    serializer: nodejs_serializer_1.serialize,
});
function initLogger(config) {
    var verbose = config.verbose;
    exports.logger.addWriter(new nodejs_console_1.NodeJsConsoleWriter({
        level: verbose ? log_level_1.LogLevel.DEBUG : log_level_1.LogLevel.INFO,
    }));
    if (isFileLoggerConfig(config)) {
        var fileLogLevel = config.fileLogLevel, fileLogDirPath = config.fileLogDirPath;
        exports.logger.addWriter(new file_1.FileWriter({
            filename: exports.LOG_FILENAME,
            sourceName: 'CLI',
            level: fileLogLevel ? getLogLevelFromString(fileLogLevel) : log_level_1.LogLevel.DEBUG,
            dirPath: fileLogDirPath,
        }));
    }
}
exports.initLogger = initLogger;
function logFatalError(fatalError) {
    var statusCode = fatalError.statusCode, errorCode = fatalError.errorCode, errorSource = fatalError.errorSource, message = fatalError.message, error = fatalError.error;
    var errorMsg = "\n[Status Code] ".concat(statusCode, "\n[Error Source] ").concat(errorSource, "\n[EPS Error Code] ").concat(errorCode, "\n[Error Message] ").concat(message, "\n");
    exports.logger.error(errorMsg);
    if (error instanceof Error) {
        exports.logger.error("Stacktrace: ".concat(error.stack));
    }
}
exports.logFatalError = logFatalError;
function isFileLoggerConfig(config) {
    return 'fileLogDirPath' in config;
}
