// @tableau/taco-toolkit/dist/cli/taco-commands/utils/spinner.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var ora_1 = __importDefault(require("ora"));
var logging_1 = require("./logging");
var Spinner = /** @class */ (function () {
    function Spinner(verbose) {
        this.spinner = (0, ora_1.default)();
        this.verbose = verbose;
    }
    Spinner.prototype.start = function (message) {
        if (!this.verbose) {
            this.spinner.start(message);
        }
        logging_1.logger.debug(message);
        return message;
    };
    Spinner.prototype.success = function (message) {
        if (!this.verbose) {
            this.spinner.succeed(message);
        }
        logging_1.logger.debug(message);
        return message;
    };
    Spinner.prototype.warn = function (message) {
        if (!this.verbose) {
            this.spinner.warn(message);
        }
        logging_1.logger.debug(message);
        return message;
    };
    Spinner.prototype.fail = function (message) {
        if (!this.verbose) {
            this.spinner.fail(message);
        }
        logging_1.logger.debug(message);
        return message;
    };
    Spinner.prototype.info = function (message) {
        if (!this.verbose) {
            this.spinner.info(message);
        }
        logging_1.logger.debug(message);
        return message;
    };
    return Spinner;
}());
exports.default = Spinner;
