// @tableau/taco-toolkit/dist/cli/taco-commands/utils/tableau-product.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTableauMajorVersion = exports.isSupportedPlatform = exports.getTableauExecutablePath = exports.PRODUCT_EXE_NAMES = exports.ProductName = void 0;
var ProductName;
(function (ProductName) {
    ProductName["Prep"] = "Prep";
    ProductName["Desktop"] = "Desktop";
})(ProductName = exports.ProductName || (exports.ProductName = {}));
exports.PRODUCT_EXE_NAMES = (_a = {},
    _a[ProductName.Prep] = 'Tableau Prep Builder',
    _a[ProductName.Desktop] = 'Tableau Desktop',
    _a);
function getTableauExecutablePath(platform, productName, tableauVersion) {
    if (platform === 'win32') {
        if (productName === ProductName.Prep) {
            // Prep: "C:\Program Files\Tableau\Tableau Prep Builder 2022.3"
            var executableName = exports.PRODUCT_EXE_NAMES[productName];
            return "C:\\Program Files\\Tableau\\".concat(executableName, " ").concat(tableauVersion, "\\").concat(executableName, ".exe");
        }
        // Desktop: "C:\Program Files\Tableau\Tableau 2022.3\bin\tableau.exe"
        return "C:\\Program Files\\Tableau\\Tableau ".concat(tableauVersion, "\\bin\\Tableau.exe");
    }
    return "/Applications/".concat(exports.PRODUCT_EXE_NAMES[productName], " ").concat(tableauVersion, ".app");
}
exports.getTableauExecutablePath = getTableauExecutablePath;
function isSupportedPlatform(platform) {
    return platform === 'win32' || platform === 'darwin';
}
exports.isSupportedPlatform = isSupportedPlatform;
/**
 * Get the tableau major version which only includes the first 2 numbers in the version
 */
function getTableauMajorVersion(version) {
    var numbers = version.split('.');
    return "".concat(numbers[0], ".").concat(numbers[1]);
}
exports.getTableauMajorVersion = getTableauMajorVersion;
