// @tableau/taco-toolkit/dist/cli/taco-commands/utils/toolkit-paths.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBoilerplatePath = exports.VALIDATION_DIR_PATH = exports.TEMPLATE_ROOT_PATH = exports.BOILERPLATE_PATH = exports.ASSETS_PATH = exports.EPS_MAIN_PATH = exports.EULA_FILE_PATH = exports.LICENSE_FILE_PATH = exports.TOOLKIT_ROOT_PATH = void 0;
var path_1 = __importDefault(require("path"));
exports.TOOLKIT_ROOT_PATH = path_1.default.resolve(__dirname, '..', '..', '..', '..');
exports.LICENSE_FILE_PATH = path_1.default.join(exports.TOOLKIT_ROOT_PATH, 'LICENSE.md');
exports.EULA_FILE_PATH = path_1.default.join(exports.TOOLKIT_ROOT_PATH, 'license-agreement.json');
exports.EPS_MAIN_PATH = path_1.default.join(__dirname, '..', '..', '..', 'eps', 'main', 'core', 'main.js');
exports.ASSETS_PATH = path_1.default.join(__dirname, '..', '..', 'assets');
exports.BOILERPLATE_PATH = path_1.default.join(exports.ASSETS_PATH, 'boilerplate');
exports.TEMPLATE_ROOT_PATH = path_1.default.join(exports.ASSETS_PATH, 'packager', 'connector_templates');
exports.VALIDATION_DIR_PATH = path_1.default.join(exports.ASSETS_PATH, 'packager', 'validation');
function getBoilerplatePath(boilerplate) {
    return path_1.default.join(exports.BOILERPLATE_PATH, boilerplate);
}
exports.getBoilerplatePath = getBoilerplatePath;
