// @tableau/taco-toolkit/dist/shared/logger/writer/http.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpWriter = exports.DEFAULT_BATCH_INTERVAL = void 0;
var ajax_1 = require("../../app-transport/ajax");
var writer_1 = require("./writer");
exports.DEFAULT_BATCH_INTERVAL = 3000;
var HttpWriter = /** @class */ (function (_super) {
    __extends(HttpWriter, _super);
    function HttpWriter(config) {
        var _this = this;
        var _a, _b, _c;
        _this = _super.call(this, config) || this;
        _this.logEntryQueue = [];
        _this.batch = (_a = config.batch) !== null && _a !== void 0 ? _a : false;
        _this.endpoint = config.endpoint;
        _this.headers = (_b = config.headers) !== null && _b !== void 0 ? _b : {};
        _this.batchInterval = (_c = config.batchInterval) !== null && _c !== void 0 ? _c : exports.DEFAULT_BATCH_INTERVAL;
        return _this;
    }
    HttpWriter.prototype.log = function (message, level) {
        if (!this.isLoggable(level)) {
            return;
        }
        var logEntry = {
            message: message,
            level: level,
            timestamp: Date.now(),
        };
        if (this.batch) {
            this.doBatch(logEntry);
        }
        else {
            this.emitLogs([logEntry]);
        }
    };
    HttpWriter.prototype.flush = function () {
        if (this.batchTimeoutId) {
            clearTimeout(this.batchTimeoutId);
            this.batchTimeoutId = undefined;
        }
        var entriesToFlush = this.logEntryQueue.slice();
        this.logEntryQueue = [];
        this.emitLogs(entriesToFlush);
    };
    HttpWriter.prototype.doBatch = function (logEntry) {
        var _this = this;
        this.logEntryQueue.push(logEntry);
        if (this.logEntryQueue.length === 1) {
            // starting timer when the first log is enqueued
            this.batchTimeoutId = setTimeout(function () {
                _this.flush();
            }, this.batchInterval);
        }
    };
    HttpWriter.prototype.emitLogs = function (logs) {
        (0, ajax_1.post)(this.endpoint, logs, {
            headers: this.headers,
        }).catch(function (error) {
            console.error('Failed to send logs to server', error);
        });
    };
    return HttpWriter;
}(writer_1.BatchableWriter));
exports.HttpWriter = HttpWriter;
