// @tableau/taco-toolkit/dist/shared/logger/writer/writer.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatchableWriter = exports.Writer = void 0;
var log_level_1 = require("../log-level");
var Writer = /** @class */ (function () {
    function Writer(config) {
        this.severity = getSeverityFromConfig(config);
    }
    Writer.prototype.setLoggerLevel = function (level) {
        this.loggerSeverity = (0, log_level_1.getLogSeverity)(level);
    };
    Writer.prototype.getMaxSeverity = function () {
        return this.severity || this.loggerSeverity;
    };
    Writer.prototype.isLoggable = function (level) {
        return (0, log_level_1.isLoggable)(this.getMaxSeverity(), level);
    };
    return Writer;
}());
exports.Writer = Writer;
var BatchableWriter = /** @class */ (function (_super) {
    __extends(BatchableWriter, _super);
    function BatchableWriter() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return BatchableWriter;
}(Writer));
exports.BatchableWriter = BatchableWriter;
function getSeverityFromConfig(config) {
    if (config.level) {
        return (0, log_level_1.getLogSeverity)(config.level);
    }
}
