// @tableau/taco-toolkit/dist/sdk/app/app-logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppLogger = void 0;
var logger_1 = require("../../shared/logger/logger");
var console_1 = require("../../shared/logger/writer/console");
var http_1 = require("../../shared/logger/writer/http");
/**
 * A logger class for logging connector app messages.
 *
 * - Connector code can utilize the static methods directly without requiring explicit initialization.
 * - All logged messages will be displayed in the browser console and written to the connector's log file.
 *
 * @class AppLogger
 */
var AppLogger = /** @class */ (function () {
    function AppLogger() {
    }
    /** @ignore */
    AppLogger.init = function (config) {
        if (AppLogger.instance !== undefined) {
            console.warn('Logger has already been initialized! Skipping...');
            return;
        }
        var level = config.level, headers = config.headers, endpoint = config.endpoint;
        AppLogger.instance = new logger_1.Logger({
            level: level,
            writers: [
                new console_1.ConsoleWriter({}),
                new http_1.HttpWriter({
                    endpoint: endpoint,
                    headers: headers,
                    batch: true,
                }),
            ],
        });
        console.info('App logger is initialized!');
    };
    /**
     * Logs an information message.
     *
     * @example
     * Logger.info('This is an information message.');
     */
    AppLogger.info = function (message) {
        if (AppLogger.instance) {
            AppLogger.instance.info(message);
        }
    };
    /**
     * Logs an error message.
     *
     * @example
     * Logger.error('This is an error message.');
     */
    AppLogger.error = function (message) {
        if (AppLogger.instance) {
            AppLogger.instance.error(message);
        }
    };
    /**
     * Logs a warning message.
     *
     * @example
     * Logger.warn('This is a warning message.');
     */
    AppLogger.warn = function (message) {
        if (AppLogger.instance) {
            AppLogger.instance.warn(message);
        }
    };
    /**
     * Logs a debug message.
     *
     * @example
     * Logger.debug('This is a debug message.');
     */
    AppLogger.debug = function (message) {
        if (AppLogger.instance) {
            AppLogger.instance.debug(message);
        }
    };
    /** @ignore */
    AppLogger.flush = function () {
        if (AppLogger.instance) {
            AppLogger.instance.flush();
        }
    };
    return AppLogger;
}());
exports.AppLogger = AppLogger;
