// @tableau/taco-toolkit/dist/shared/errors/error-codes.js
"use strict";
// Use this npm run gen-error-code
// or method to generate unique error code(ErrorCodes) of length - strlen
// generateUniqueCode(strlen: number): string {
//   return Math.round((Math.pow(36, strlen + 1) - Math.random() * Math.pow(36, strlen))).toString(36).toUpperCase().slice(1);
// }
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorCodes = void 0;
var ErrorCodes;
(function (ErrorCodes) {
    ErrorCodes["NAGGOW13"] = "NAGGOW13";
    ErrorCodes["QPFBBH4F"] = "QPFBBH4F";
    ErrorCodes["ZUWGQ7DL"] = "ZUWGQ7DL";
    ErrorCodes["RIPDN45L"] = "RIPDN45L";
    ErrorCodes["KJ2SJNYU"] = "KJ2SJNYU";
    ErrorCodes["QIC79P0A"] = "QIC79P0A";
    ErrorCodes["X8KTF8LS"] = "X8KTF8LS";
    ErrorCodes["KMSFUMU2"] = "KMSFUMU2";
    ErrorCodes["A1GHIXJ3"] = "A1GHIXJ3";
    ErrorCodes["WU2PPXTO"] = "WU2PPXTO";
    ErrorCodes["IAA7I66W"] = "IAA7I66W";
    ErrorCodes["BUY1GBYB"] = "BUY1GBYB";
    ErrorCodes["WMKEJO22"] = "WMKEJO22";
    ErrorCodes["T6MAR5LA"] = "T6MAR5LA";
    ErrorCodes["WQX8WEDW"] = "WQX8WEDW";
    ErrorCodes["IDBIRO82"] = "IDBIRO82";
    ErrorCodes["PGP0LOLF"] = "PGP0LOLF";
    ErrorCodes["BLLKEB1R"] = "BLLKEB1R";
    ErrorCodes["M7OX88FN"] = "M7OX88FN";
    ErrorCodes["IRX591ZY"] = "IRX591ZY";
    ErrorCodes["JWVEWN26"] = "JWVEWN26";
    ErrorCodes["Y6NPK392"] = "Y6NPK392";
    ErrorCodes["T7LNRZ5C"] = "T7LNRZ5C";
    ErrorCodes["LRYN2IMA"] = "LRYN2IMA";
    ErrorCodes["Y4GXMCZP"] = "Y4GXMCZP";
    ErrorCodes["EJB55C4V"] = "EJB55C4V";
    ErrorCodes["BCQYP73Y"] = "BCQYP73Y";
    ErrorCodes["G0LV97J0"] = "G0LV97J0";
    ErrorCodes["M3N2YI8H"] = "M3N2YI8H";
    ErrorCodes["YI89TM4Q"] = "YI89TM4Q";
    ErrorCodes["BUDOHYOY"] = "BUDOHYOY";
    ErrorCodes["LNQB4OXC"] = "LNQB4OXC";
    ErrorCodes["W5JZAKNZ"] = "W5JZAKNZ";
    ErrorCodes["LEASDMQQ"] = "LEASDMQQ";
    ErrorCodes["IRGS6QUI"] = "IRGS6QUI";
    ErrorCodes["MSAAL1S9"] = "MSAAL1S9";
    ErrorCodes["Q1XA447W"] = "Q1XA447W";
    ErrorCodes["OD3MMIGV"] = "OD3MMIGV";
    ErrorCodes["Z0TX352C"] = "Z0TX352C";
    ErrorCodes["ADS1HIGD"] = "ADS1HIGD";
    ErrorCodes["KY5XRFYV"] = "KY5XRFYV";
    ErrorCodes["GSIP0KQM"] = "GSIP0KQM";
    ErrorCodes["M54O33RX"] = "M54O33RX";
    ErrorCodes["P6WNIRTU"] = "P6WNIRTU";
    ErrorCodes["ZZD5FP0F"] = "ZZD5FP0F";
    ErrorCodes["QYM7P183"] = "QYM7P183";
    ErrorCodes["IIY6J1V4"] = "IIY6J1V4";
    ErrorCodes["PSZECVJ1"] = "PSZECVJ1";
    ErrorCodes["Y4LLMUJF"] = "Y4LLMUJF";
    ErrorCodes["XQJT9XXC"] = "XQJT9XXC";
    ErrorCodes["DXHBJZUO"] = "DXHBJZUO";
    ErrorCodes["BOAR6OM6"] = "BOAR6OM6";
    ErrorCodes["YIA0FDX5"] = "YIA0FDX5";
    ErrorCodes["WT2WHBA0"] = "WT2WHBA0";
    ErrorCodes["ENT97JSH"] = "ENT97JSH";
    ErrorCodes["MDDL9LNB"] = "MDDL9LNB";
    ErrorCodes["TR48H3Q3"] = "TR48H3Q3";
    ErrorCodes["HY40PKIN"] = "HY40PKIN";
    ErrorCodes["DNZ9QTGX"] = "DNZ9QTGX";
    ErrorCodes["S5B058E1"] = "S5B058E1";
    ErrorCodes["RXCCMIVQ"] = "RXCCMIVQ";
    ErrorCodes["UYJY51CE"] = "UYJY51CE";
    ErrorCodes["INEO9MTU"] = "INEO9MTU";
    ErrorCodes["ORP420NN"] = "ORP420NN";
    ErrorCodes["EZV90I5V"] = "EZV90I5V";
    ErrorCodes["RMT9JR0V"] = "RMT9JR0V";
    ErrorCodes["IYZVJGZ2"] = "IYZVJGZ2";
    ErrorCodes["A0WRFSSS"] = "A0WRFSSS";
    ErrorCodes["OHD0EC76"] = "OHD0EC76";
    ErrorCodes["H4XI3IY9"] = "H4XI3IY9";
    ErrorCodes["K036PO50"] = "K036PO50";
    ErrorCodes["T1GDF8UC"] = "T1GDF8UC";
    ErrorCodes["AYV9NKXK"] = "AYV9NKXK";
    ErrorCodes["CJ4MKY90"] = "CJ4MKY90";
    ErrorCodes["ZU2IBBK4"] = "ZU2IBBK4";
    ErrorCodes["NM8TMLBL"] = "NM8TMLBL";
    ErrorCodes["JPKID4LW"] = "JPKID4LW";
    ErrorCodes["ZUWAQ7DL"] = "ZUWAQ7DL";
    ErrorCodes["RWNO9V3N"] = "RWNO9V3N";
    ErrorCodes["M92GXFQP"] = "M92GXFQP";
    ErrorCodes["XNEXCOQ7"] = "XNEXCOQ7";
    ErrorCodes["RSH30AHC"] = "RSH30AHC";
    ErrorCodes["MHHG8I10"] = "MHHG8I10";
    ErrorCodes["HU3H87WB"] = "HU3H87WB";
})(ErrorCodes = exports.ErrorCodes || (exports.ErrorCodes = {}));
