// @tableau/taco-toolkit/dist/shared/errors/status-code.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StatusCode = void 0;
var StatusCode;
(function (StatusCode) {
    /** Process is launched but unhealthy, should be terminated by client */
    StatusCode["Unhealthy"] = "UNHEALTHY";
    StatusCode["Cancelled"] = "CANCELLED";
    StatusCode["Unknown"] = "UNKNOWN";
    StatusCode["InvalidArgument"] = "INVALID_ARGUMENT";
    StatusCode["Deadline_Exceeded"] = "DEADLINE_EXCEEDED";
    StatusCode["NotFound"] = "NOT_FOUND";
    StatusCode["AlreadyExists"] = "ALREADY_EXISTS";
    StatusCode["PermissionDenied"] = "PERMISSION_DENIED";
    StatusCode["ResourceExhausted"] = "RESOURCE_EXHAUSTED";
    StatusCode["FailedPrecondition"] = "FAILED_PRECONDITION";
    StatusCode["Aborted"] = "ABORTED";
    StatusCode["OutOfRange"] = "OUT_OF_RANGE";
    StatusCode["Unimplemented"] = "UNIMPLEMENTED";
    StatusCode["Internal"] = "INTERNAL";
    StatusCode["Unavailable"] = "UNAVAILABLE";
    StatusCode["DataLoss"] = "DATA_LOSS";
    StatusCode["Unauthenticated"] = "UNAUTHENTICATED";
})(StatusCode = exports.StatusCode || (exports.StatusCode = {}));
