// @tableau/taco-toolkit/dist/eps/sandbox/sdk/sandbox-plugin.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SandboxPlugin = void 0;
var SandboxPlugin = /** @class */ (function () {
    function SandboxPlugin(sandbox) {
        this.sandbox = sandbox;
    }
    /**
     * Creates a class in the sandbox with the given methods.
     * @param methods Methods to define in sandbox
     * @param className Class to define in sandbox.
     */
    SandboxPlugin.prototype.createClass = function (methods, className) {
        var _this = this;
        var sortedMethods = methods.map(function (m) { return (__assign(__assign({}, m), { sandboxName: className + '$' + m.externalName })); });
        sortedMethods.sort(function (a, b) {
            // sort statics in front, then alphabetically
            if (a.isStatic !== b.isStatic) {
                return a.isStatic ? -1 : 1;
            }
            return a.externalName.localeCompare(b.externalName);
        });
        sortedMethods.forEach(function (m) {
            _this.sandbox.addMethod(m.sandboxName, m.functionRef, {
                promise: true,
                reference: true,
            });
        });
        var code = "\n      class ".concat(className, " {\n        ").concat(sortedMethods === null || sortedMethods === void 0 ? void 0 : sortedMethods.map(this.createMethod).join(';'), "\n      }\n    ");
        this.sandbox.executeCodeSync(code);
    };
    SandboxPlugin.prototype.createMethod = function (m) {
        return "\n\n    ".concat(m.isStatic ? 'static' : '', " async ").concat(m.externalName, "(...args) {\n      const fn = await requireAsync('").concat(m.sandboxName, "')\n      return await fn(...args)\n    }\n      ");
    };
    return SandboxPlugin;
}());
exports.SandboxPlugin = SandboxPlugin;
