// @tableau/taco-toolkit/dist/shared/logger/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
var log_level_1 = require("./log-level");
var default_serializer_1 = require("./serializer/default-serializer");
var Logger = /** @class */ (function () {
    function Logger(config) {
        var _this = this;
        var _a, _b;
        this.writers = [];
        this.batchableWriters = [];
        this.level = (_a = config.level) !== null && _a !== void 0 ? _a : log_level_1.LogLevel.INFO;
        this.serializer = (_b = config.serializer) !== null && _b !== void 0 ? _b : default_serializer_1.serialize;
        config.writers.forEach(function (writer) {
            _this.addWriter(writer);
        });
    }
    Logger.prototype.addWriter = function (writer) {
        writer.setLoggerLevel(this.level);
        if (isBatchableWriter(writer)) {
            this.batchableWriters.push(writer);
        }
        this.writers.push(writer);
    };
    Logger.prototype.flush = function () {
        // force to flush batchable writers
        this.batchableWriters.forEach(function (writer) { return writer.flush(); });
    };
    Logger.prototype.debug = function (message, options) {
        this.log(message, log_level_1.LogLevel.DEBUG, options);
    };
    Logger.prototype.info = function (message, options) {
        this.log(message, log_level_1.LogLevel.INFO, options);
    };
    Logger.prototype.error = function (message, options) {
        this.log(message, log_level_1.LogLevel.ERROR, options);
    };
    Logger.prototype.warn = function (message, options) {
        this.log(message, log_level_1.LogLevel.WARNING, options);
    };
    Logger.prototype.log = function (message, level, options) {
        var msgString = this.serializer(message);
        this.writers.forEach(function (writer) {
            writer.log(msgString, level, options);
        });
    };
    return Logger;
}());
exports.Logger = Logger;
function isBatchableWriter(writer) {
    return 'flush' in writer;
}
