// @tableau/taco-toolkit/dist/eps/static-server/static-server.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var express_1 = __importDefault(require("express"));
var http_1 = __importDefault(require("http"));
var http_status_codes_1 = require("http-status-codes");
var api_utils_1 = require("../utils/api-utils");
var eps_logger_1 = require("../logging/eps-logger");
var server_utils_1 = require("../utils/server-utils");
var StaticServer = /** @class */ (function () {
    function StaticServer(staticServerId, staticServerAppDir) {
        this.staticServerId = staticServerId;
        this.staticServerAppDir = staticServerAppDir;
        this.app = (0, express_1.default)();
        this.interceptRequests();
        this.createServer();
        // register default error handlers
        this.app.use(api_utils_1.notFoundHandler);
        this.app.use(api_utils_1.errorHandler);
    }
    StaticServer.prototype.createServer = function () {
        this.app.use("/".concat(this.staticServerId), express_1.default.static(this.staticServerAppDir, {
            fallthrough: false,
        }));
    };
    StaticServer.prototype.interceptRequests = function () {
        var _this = this;
        this.app.use((0, api_utils_1.getRequestLogger)(this.constructor.name));
        this.app.use(function (req, res, next) {
            if (req.path.includes("/".concat(_this.staticServerId))) {
                return next();
            }
            return res.sendStatus(http_status_codes_1.StatusCodes.FORBIDDEN);
        });
    };
    StaticServer.prototype.start = function (port) {
        return __awaiter(this, void 0, void 0, function () {
            var error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        this.server = http_1.default.createServer(this.app);
                        return [4 /*yield*/, (0, server_utils_1.startHttpServer)(this.server, port)];
                    case 1:
                        _a.sent();
                        eps_logger_1.Logger.info("StaticServer is listening on port ".concat(port, "..."));
                        return [3 /*break*/, 3];
                    case 2:
                        error_1 = _a.sent();
                        error_1.message = "Failed to start StaticServer on port ".concat(port, ". ").concat(error_1.message);
                        throw error_1;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    StaticServer.prototype.close = function () {
        if (this.server === undefined) {
            return Promise.resolve();
        }
        return (0, server_utils_1.closeHttpServer)(this.server);
    };
    return StaticServer;
}());
exports.default = StaticServer;
