// @tableau/taco-toolkit/dist/eps/static-server/static-server-app-config.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStaticServerAppConfig = void 0;
var app_config_utils_1 = require("../utils/app-config-utils");
var appConfig;
/**
 * Compose Extractor app config from process args and environment variables.
 */
function getStaticServerAppConfig() {
    if (appConfig) {
        return appConfig;
    }
    /**** options from process arguments ****/
    var _a = __read(process.argv, 6), epsInstanceId = _a[2], connectorAppDir = _a[3], staticServerPort = _a[4], staticServerId = _a[5];
    /**** options from environment variables ****/
    var logOption = (0, app_config_utils_1.getRequiredEnvVariable)('EPS_APP_LOG_CONFIG');
    appConfig = {
        epsInstanceId: epsInstanceId,
        connectorAppDir: connectorAppDir,
        staticServerPort: parseInt(staticServerPort),
        staticServerId: staticServerId,
        logOption: (0, app_config_utils_1.parseAppLogOption)(logOption),
    };
    return appConfig;
}
exports.getStaticServerAppConfig = getStaticServerAppConfig;
