// @tableau/taco-toolkit/dist/eps/static-server/static-server-app.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var eps_logger_1 = require("../logging/eps-logger");
var eps_process_types_1 = require("../types/eps-process-types");
var static_server_1 = __importDefault(require("./static-server"));
var static_server_app_config_1 = require("./static-server-app-config");
var static_server_message_1 = require("./static-server-message");
var StaticServerApp = /** @class */ (function () {
    function StaticServerApp(staticServerId, connectorAppDir) {
        this.staticServer = new static_server_1.default(staticServerId, connectorAppDir);
    }
    StaticServerApp.prototype.start = function (port) {
        return this.staticServer.start(port);
    };
    StaticServerApp.prototype.close = function () {
        return this.staticServer.close();
    };
    return StaticServerApp;
}());
if (require.main === module) {
    process.title = eps_process_types_1.EPSProcessName.EPSStaticServer;
    var _a = (0, static_server_app_config_1.getStaticServerAppConfig)(), epsInstanceId = _a.epsInstanceId, logOption = _a.logOption, staticServerPort = _a.staticServerPort, staticServerId = _a.staticServerId, connectorAppDir = _a.connectorAppDir;
    var loggerConfig = (0, eps_logger_1.createLoggerConfig)('static-server', logOption);
    eps_logger_1.Logger.init(epsInstanceId, loggerConfig);
    eps_logger_1.Logger.debug("Initiating ".concat(StaticServerApp.name));
    var staticServerApp_1 = new StaticServerApp(staticServerId, connectorAppDir);
    process.on('SIGTERM', function () {
        eps_logger_1.Logger.info('SIGTERM signal received: closing Static Server');
        staticServerApp_1.close();
    });
    staticServerApp_1
        .start(staticServerPort)
        .then(function () {
        return (0, static_server_message_1.sendStaticServerMessage)({
            status: 'success',
        });
    })
        .catch(function (error) {
        eps_logger_1.Logger.error(error);
        (0, static_server_message_1.sendStaticServerMessage)({
            status: 'error',
            error: error,
        });
    });
}
