// @tableau/taco-toolkit/dist/eps/main/api/eps-api-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.requestLogReplacer = void 0;
var url_utils_1 = require("../../utils/url-utils");
function requestLogReplacer(request) {
    // root path endpoint is used for redirects from connectors,
    // which may contain user infos and credentials.
    if (request.path === '/') {
        var _a = request.query, url = _a.url, headers = _a.headers;
        return {
            requestPath: request.path,
            url: parseURLParam(url),
            headers: parseHeadersParam(headers),
        };
    }
}
exports.requestLogReplacer = requestLogReplacer;
function parseURLParam(urlStr) {
    if (urlStr === undefined) {
        return {};
    }
    if (typeof urlStr === 'string') {
        try {
            return (0, url_utils_1.parseUrlAndQueryParams)(urlStr);
        }
        catch (e) {
            // no-op, return message below
        }
    }
    return 'Invalid URL string';
}
function parseHeadersParam(headerStr) {
    if (headerStr === undefined) {
        return [];
    }
    if (typeof headerStr === 'string') {
        try {
            return Object.keys(JSON.parse(headerStr));
        }
        catch (e) {
            // no-op, return message below
        }
    }
    return 'Invalid header string';
}
