// @tableau/taco-toolkit/dist/cli/taco-commands/pack/prepare.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.preparePack = void 0;
var fs_1 = __importDefault(require("fs"));
var path_1 = __importDefault(require("path"));
var inquirer_1 = __importDefault(require("inquirer"));
var command_1 = require("../utils/command");
var connector_1 = require("../utils/connector");
var connector_json_1 = require("../utils/connector-json");
var constants_1 = require("../utils/constants");
var logging_1 = require("../utils/logging");
var taco_files_1 = require("./taco-files");
var console_1 = require("../utils/console");
var operationChoices = ['overwrite', 'skip', 'cancel'];
function preparePack(tacoRootPath, force, state) {
    return __awaiter(this, void 0, void 0, function () {
        var spinner, connectorJsonPath, connectorJson, templateAndDests, existedFiles, filenames, operation, createFilesResult;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    spinner = state.spinner;
                    spinner.info('Preparing the required files for taco...');
                    connectorJsonPath = (0, connector_1.getConnectorJsonPath)(tacoRootPath);
                    connectorJson = (0, connector_json_1.getConnectorJson)(connectorJsonPath);
                    templateAndDests = (0, taco_files_1.getTemplateAndDestPath)(tacoRootPath, connectorJson);
                    if (!!force) return [3 /*break*/, 2];
                    existedFiles = templateAndDests.filter(function (_a) {
                        var destFilePath = _a.destFilePath;
                        return fs_1.default.existsSync(destFilePath);
                    });
                    if (!(existedFiles.length > 0)) return [3 /*break*/, 2];
                    filenames = existedFiles.map(function (_a) {
                        var destFilePath = _a.destFilePath;
                        return path_1.default.basename(destFilePath);
                    });
                    return [4 /*yield*/, askForOperation(filenames)];
                case 1:
                    operation = _a.sent();
                    switch (operation) {
                        case 'skip': {
                            spinner.warn('Skipped creating the taco xml files. Packing connector with existing files.');
                            (0, console_1.consoleWarn)('Please make sure all information in connector.json are in sync with the existing taco xml file. Otherwise, it may cause runtime error when running the taco with Tableau products.');
                            return [2 /*return*/];
                        }
                        case 'cancel': {
                            spinner.warn('taco pack has been cancelled.');
                            process.exit(0);
                        }
                        // otherwise, continue to overwrite the files
                    }
                    _a.label = 2;
                case 2: return [4 /*yield*/, (0, taco_files_1.createFiles)(connectorJson, templateAndDests)];
                case 3:
                    createFilesResult = _a.sent();
                    if ((0, command_1.isErrorResult)(createFilesResult)) {
                        spinner.fail('Failed to create taco xml files');
                        (0, logging_1.logFatalError)(createFilesResult.error);
                        process.exit(constants_1.TACO_COMMAND_ERROR_EXIT_CODE);
                    }
                    spinner.success('Successfully generated all the required files for taco');
                    return [2 /*return*/];
            }
        });
    });
}
exports.preparePack = preparePack;
function askForOperation(existedFiles) {
    return __awaiter(this, void 0, void 0, function () {
        var operation;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, inquirer_1.default.prompt({
                        name: 'operation',
                        message: "The following files already exist, '".concat(existedFiles.join(', '), "'.\nPlease select one of the options:"),
                        type: 'rawlist',
                        choices: operationChoices,
                    })];
                case 1:
                    operation = (_a.sent()).operation;
                    return [2 /*return*/, operation];
            }
        });
    });
}
