// @tableau/taco-toolkit/dist/cli/taco-commands/pack/validation.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateDistConnectorJson = exports.validateXmlFiles = exports.validateTacoXmlFiles = exports.validateTaco = void 0;
var path_1 = __importDefault(require("path"));
var fs_1 = __importDefault(require("fs"));
var libxmljs2_1 = __importDefault(require("libxmljs2"));
var connector_1 = require("../utils/connector");
var connector_json_1 = require("../utils/connector-json");
var toolkit_paths_1 = require("../utils/toolkit-paths");
var logging_1 = require("../utils/logging");
var status_code_1 = require("../../../shared/errors/status-code");
var error_source_1 = require("../../../shared/errors/error-source");
var error_codes_1 = require("../../../shared/errors/error-codes");
var constants_1 = require("../utils/constants");
var COMMON_XML_XDS_MAP = {
    'manifest.xml': 'connector_plugin_manifest_latest.xsd',
    'connection-fields.xml': 'connection_fields.xsd',
    'connection-metadata.xml': 'connector_plugin_metadata.xsd',
};
function validateTaco(tacoRootPath, state) {
    var spinner = state.spinner;
    spinner.start("Validating taco project at ".concat(tacoRootPath));
    try {
        var connectorJsonPath = (0, connector_1.getConnectorJsonPath)(tacoRootPath);
        var connectorJson = (0, connector_json_1.getConnectorJson)(connectorJsonPath);
        validateDistConnectorJson(connectorJson, tacoRootPath);
        validateTacoXmlFiles(connectorJson, tacoRootPath);
    }
    catch (e) {
        spinner.fail('Validation failed!');
        (0, logging_1.logFatalError)({
            statusCode: status_code_1.StatusCode.Cancelled,
            errorSource: error_source_1.ErrorSources.Client,
            errorCode: error_codes_1.ErrorCodes.T7LNRZ5C,
            message: e.toString(),
            error: e,
        });
        process.exit(constants_1.TACO_COMMAND_ERROR_EXIT_CODE);
    }
    spinner.success('Validation succeeded! The project is ready to pack.');
}
exports.validateTaco = validateTaco;
function getValidationFileInfos(connectorJson, tacoRootPath) {
    var e_1, _a;
    var validationFileInfos = [];
    try {
        for (var _b = __values(Object.entries(COMMON_XML_XDS_MAP)), _c = _b.next(); !_c.done; _c = _b.next()) {
            var _d = __read(_c.value, 2), xmlFile = _d[0], xsdFile = _d[1];
            validationFileInfos.push({
                xmlFilePath: path_1.default.join(tacoRootPath, xmlFile),
                xsdFilePath: path_1.default.join(toolkit_paths_1.VALIDATION_DIR_PATH, xsdFile),
            });
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        }
        finally { if (e_1) throw e_1.error; }
    }
    if ((0, connector_json_1.isOauthConnector)(connectorJson)) {
        validationFileInfos.push({
            xmlFilePath: path_1.default.join(tacoRootPath, 'oauth-config.xml'),
            xsdFilePath: path_1.default.join(toolkit_paths_1.VALIDATION_DIR_PATH, 'oauth_config.xsd'),
        });
    }
    return validationFileInfos;
}
function validateTacoXmlFiles(connectorJson, tacoRootPath) {
    var validationFileInfos = getValidationFileInfos(connectorJson, tacoRootPath);
    // separate the impl, so that we can individually test the validation files
    validateXmlFiles(validationFileInfos);
}
exports.validateTacoXmlFiles = validateTacoXmlFiles;
function validateXmlFiles(validationFileInfos) {
    var e_2, _a;
    // validate required xml files are existed
    var missingFiles = validationFileInfos.filter(function (_a) {
        var xmlFilePath = _a.xmlFilePath;
        return !fs_1.default.existsSync(xmlFilePath);
    });
    if (missingFiles.length > 0) {
        throw new Error("Missing required taco files:\n".concat(missingFiles.map(function (_a) {
            var xmlFilePath = _a.xmlFilePath;
            return xmlFilePath;
        }).join('\n')));
    }
    try {
        // validate xml files content
        for (var validationFileInfos_1 = __values(validationFileInfos), validationFileInfos_1_1 = validationFileInfos_1.next(); !validationFileInfos_1_1.done; validationFileInfos_1_1 = validationFileInfos_1.next()) {
            var _b = validationFileInfos_1_1.value, xmlFilePath = _b.xmlFilePath, xsdFilePath = _b.xsdFilePath;
            logging_1.logger.debug("Validating ".concat(xmlFilePath, " with schema ").concat(xsdFilePath));
            var xmlDoc = libxmljs2_1.default.parseXml(fs_1.default.readFileSync(xmlFilePath).toString());
            var xsdDoc = libxmljs2_1.default.parseXml(fs_1.default.readFileSync(xsdFilePath).toString());
            if (!xmlDoc.validate(xsdDoc)) {
                throw new Error("Invalid taco XML file was found at: ".concat(xmlFilePath, ".\nErrors: ").concat(xmlDoc.validationErrors));
            }
        }
    }
    catch (e_2_1) { e_2 = { error: e_2_1 }; }
    finally {
        try {
            if (validationFileInfos_1_1 && !validationFileInfos_1_1.done && (_a = validationFileInfos_1.return)) _a.call(validationFileInfos_1);
        }
        finally { if (e_2) throw e_2.error; }
    }
}
exports.validateXmlFiles = validateXmlFiles;
function validateDistConnectorJson(sourceConnectorJson, tacoRootPath) {
    var e_3, _a;
    var sourceContent = JSON.stringify(sourceConnectorJson);
    try {
        for (var sourceTypes_1 = __values(connector_1.sourceTypes), sourceTypes_1_1 = sourceTypes_1.next(); !sourceTypes_1_1.done; sourceTypes_1_1 = sourceTypes_1.next()) {
            var target = sourceTypes_1_1.value;
            var connectorJsonPath = path_1.default.join(tacoRootPath, 'dist', target, 'connector.json');
            var connectJson = (0, connector_json_1.getConnectorJson)(connectorJsonPath);
            if (sourceContent !== JSON.stringify(connectJson)) {
                throw new Error("Forgot to build? connector.json at '".concat(connectorJsonPath, "' is inconsistent with its source at ").concat((0, connector_1.getConnectorJsonPath)(tacoRootPath)));
            }
        }
    }
    catch (e_3_1) { e_3 = { error: e_3_1 }; }
    finally {
        try {
            if (sourceTypes_1_1 && !sourceTypes_1_1.done && (_a = sourceTypes_1.return)) _a.call(sourceTypes_1);
        }
        finally { if (e_3) throw e_3.error; }
    }
}
exports.validateDistConnectorJson = validateDistConnectorJson;
