// @tableau/taco-toolkit/dist/cli/cli.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.run = void 0;
var chalk_1 = __importDefault(require("chalk"));
var figlet_1 = __importDefault(require("figlet"));
var commander_1 = require("commander");
var constants_1 = require("./taco-commands/utils/constants");
var package_json_1 = require("./taco-commands/utils/package-json");
var eula_1 = require("./taco-commands/license/eula");
var command_1 = require("./taco-commands/command");
var taco_1 = require("./taco-commands/taco");
var console_1 = require("./taco-commands/utils/console");
var USER_ERROR_CODES = {
    'commander.unknownOption': true,
    'commander.unknownCommand': true,
    'commander.help': true,
    'commander.optionMissingArgument': true,
    'commander.missingMandatoryOptionValue': true,
    'commander.missingArgument': true,
    'commander.invalidArgument': true,
};
function run(args) {
    var _a;
    return __awaiter(this, void 0, void 0, function () {
        var index, exitCode;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    index = args.indexOf(command_1.OPTION_ACCEPT_EULA);
                    if (!(index > -1)) return [3 /*break*/, 1];
                    // accept-eula is a global option to auto accept EULA without viewing the content,
                    // so we process it at the beginning and drop it
                    (0, eula_1.generateEula)();
                    args.splice(index, 1);
                    return [3 /*break*/, 3];
                case 1:
                    if (!!(0, eula_1.hasLicenseAgreement)()) return [3 /*break*/, 3];
                    return [4 /*yield*/, (0, eula_1.askForEula)()];
                case 2:
                    _b.sent();
                    _b.label = 3;
                case 3:
                    if (args.length < 3) {
                        showTacoBanner();
                        showToolkitVersion();
                        taco_1.program.outputHelp();
                        return [2 /*return*/];
                    }
                    try {
                        taco_1.program.parse(args);
                    }
                    catch (e) {
                        // with exitOverride commander.js throws on all exit, even for exit code 0
                        if (!(e instanceof commander_1.CommanderError) || (e.exitCode !== 0 && !isUserErrorCode(e))) {
                            (0, console_1.consoleError)('Unexpected error occurred from taco command.', e);
                            if (e instanceof commander_1.CommanderError) {
                                (0, console_1.consoleError)("CommanderError: ".concat(JSON.stringify(e)));
                            }
                            exitCode = (_a = e.exitCode) !== null && _a !== void 0 ? _a : constants_1.TACO_COMMAND_ERROR_EXIT_CODE;
                            process.exit(exitCode);
                        }
                    }
                    return [2 /*return*/];
            }
        });
    });
}
exports.run = run;
/**
 * The errors that are caused by users' invalid input and are handled by commander.js
 */
function isUserErrorCode(error) {
    return error.code in USER_ERROR_CODES;
}
function showTacoBanner() {
    console.log(chalk_1.default.keyword('orange')(figlet_1.default.textSync('TACO Toolkit', {
        font: 'Small',
        horizontalLayout: 'full',
    })));
}
function showToolkitVersion() {
    console.log("Running taco-toolkit version: ".concat(package_json_1.PACKAGE_INFO.version));
}
// for backward compatibility
// currently, latest npm is still consuming this patten
exports.default = {
    run: run,
};
