// @tableau/taco-toolkit/dist/cli/taco-commands/run/launch.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.launchTableau = void 0;
var os_1 = __importDefault(require("os"));
var fs_1 = __importDefault(require("fs"));
var error_codes_1 = require("../../../shared/errors/error-codes");
var error_source_1 = require("../../../shared/errors/error-source");
var status_code_1 = require("../../../shared/errors/status-code");
var command_1 = require("../utils/command");
var tableau_product_1 = require("../utils/tableau-product");
var constants_1 = require("../utils/constants");
var console_1 = require("../utils/console");
var logging_1 = require("../utils/logging");
var inquirer_1 = __importDefault(require("inquirer"));
function launchTableau(config, state) {
    return __awaiter(this, void 0, void 0, function () {
        var spinner, tacoFolderPath, tacoFiles, proceed, executableDisplay, platform, _a, cmd, params, result, _b, stderr, stdout, error;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    spinner = state.spinner;
                    tacoFolderPath = (0, command_1.resolvePath)(config.tacoFolderPath);
                    tacoFiles = getTacoFiles(tacoFolderPath);
                    if (!(tacoFiles.length === 0)) return [3 /*break*/, 2];
                    return [4 /*yield*/, askForContinue(tacoFolderPath)];
                case 1:
                    proceed = _c.sent();
                    if (!proceed) {
                        (0, console_1.consoleWarn)('taco run has been cancelled.');
                        process.exit(1);
                    }
                    _c.label = 2;
                case 2:
                    executableDisplay = getExecutableDisplayString(config);
                    platform = os_1.default.platform();
                    if (!(0, tableau_product_1.isSupportedPlatform)(platform)) {
                        (0, console_1.consoleError)("Unable to determine operating system type '".concat(platform, "' and corresponding executable '").concat(executableDisplay, "'."));
                        process.exit(1);
                    }
                    _a = getLaunchCmdAndParams(platform, tacoFolderPath, config), cmd = _a.cmd, params = _a.params;
                    spinner.start("Starting Tableau ".concat(executableDisplay));
                    logging_1.logger.debug("Launching Tableau ".concat(executableDisplay, " with taco connectors:\n").concat(tacoFiles.join('\n')));
                    result = (0, command_1.executeCmd)(cmd, params);
                    if ((0, command_1.isErrorResult)(result)) {
                        spinner.fail("Failed to launch Tableau ".concat(executableDisplay));
                        _b = result.error, stderr = _b.stderr, stdout = _b.stdout, error = _b.error;
                        logging_1.logger.debug("Execution failure: failed to install npm dependencies.\nstdout: \"".concat(stdout, "\"\nstderr: \"").concat(stderr, "\""));
                        (0, logging_1.logFatalError)({
                            statusCode: status_code_1.StatusCode.Cancelled,
                            errorSource: error_source_1.ErrorSources.Client,
                            errorCode: error_codes_1.ErrorCodes.X8KTF8LS,
                            message: "Failed to start Tableau '".concat(executableDisplay, "'.\nMore information: ").concat(stderr),
                            error: error,
                        });
                        process.exit(constants_1.TACO_COMMAND_ERROR_EXIT_CODE);
                    }
                    spinner.success("Successfully launched Tableau ".concat(executableDisplay));
                    return [2 /*return*/];
            }
        });
    });
}
exports.launchTableau = launchTableau;
function getLaunchCmdAndParams(platform, tacoFolderPath, launchConfig) {
    var epsPath = launchConfig.epsPath;
    var tableauExecutable = getExecutablePath(platform, launchConfig);
    var launchParams = [
        '--args',
        '-DLogLevel=debug',
        "-DConnectPluginsPath=".concat(tacoFolderPath),
        '-DDisableVerifyConnectorPluginSignature=true',
    ];
    if (epsPath) {
        launchParams.push("-DEPSPath=".concat(epsPath));
    }
    switch (platform) {
        case 'win32': {
            return {
                cmd: tableauExecutable,
                params: launchParams,
            };
        }
        case 'darwin': {
            launchParams.unshift('-a', tableauExecutable);
            return {
                cmd: 'open',
                params: launchParams,
            };
        }
    }
}
function isExecutableConfig(config) {
    return 'appPath' in config;
}
/**
 * Get an absolute path for a Tableau product
 */
function getExecutablePath(platform, config) {
    if (isExecutableConfig(config)) {
        return (0, command_1.resolvePath)(config.appPath);
    }
    var appName = config.appName, tableauVersion = config.appVersion;
    var executablePath = (0, tableau_product_1.getTableauExecutablePath)(platform, appName, tableauVersion);
    if (!fs_1.default.existsSync(executablePath)) {
        (0, logging_1.logFatalError)({
            statusCode: status_code_1.StatusCode.FailedPrecondition,
            errorSource: error_source_1.ErrorSources.Content,
            errorCode: error_codes_1.ErrorCodes.BLLKEB1R,
            message: "Unable to locate executable file for Tableau ".concat(appName, " ").concat(tableauVersion, " at ").concat(executablePath, " \nIf the Tableau app is installed at different location, you can use --app-path instead."),
        });
        process.exit(constants_1.TACO_COMMAND_ERROR_EXIT_CODE);
    }
    return executablePath;
}
function getExecutableDisplayString(config) {
    if (isExecutableConfig(config)) {
        return config.appPath;
    }
    return "".concat(config.appName, " ").concat(config.appVersion);
}
function askForContinue(tacoFolderPath) {
    return __awaiter(this, void 0, void 0, function () {
        var proceed;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, inquirer_1.default.prompt({
                        name: 'proceed',
                        message: "There isn't any taco file under '".concat(tacoFolderPath, "'. Do you want to continue launching Tableau app?"),
                        type: 'confirm',
                        default: false,
                    })];
                case 1:
                    proceed = (_a.sent()).proceed;
                    return [2 /*return*/, proceed];
            }
        });
    });
}
function getTacoFiles(tacoFolderPath) {
    return fs_1.default
        .readdirSync(tacoFolderPath, { withFileTypes: true })
        .filter(function (item) { return !item.isDirectory() && item.name.endsWith('.taco'); })
        .map(function (item) { return item.name; });
}
