// @tableau/taco-toolkit/dist/cli/taco-commands/license/eula.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasLicenseAgreement = exports.generateEula = exports.displayEulaView = exports.askForEula = exports.choices = void 0;
var fs_1 = __importDefault(require("fs"));
var fs_extra_1 = __importDefault(require("fs-extra"));
var inquirer_1 = __importDefault(require("inquirer"));
var open_1 = __importDefault(require("open"));
var toolkit_paths_1 = require("../utils/toolkit-paths");
var package_json_1 = require("../utils/package-json");
var constants_1 = require("../utils/constants");
var console_1 = require("../utils/console");
var LICENSE_PAGE = 'https://help.tableau.com/current/api/webdataconnector/en-us/docs/wdc_license.html';
exports.choices = {
    agree: 'Agree',
    decline: 'Decline',
    view: 'View license in browser',
};
function askForEula() {
    return __awaiter(this, void 0, void 0, function () {
        var decision, e_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, displayEulaView()];
                case 1:
                    decision = _a.sent();
                    if (decision === exports.choices.agree) {
                        generateEula();
                        (0, console_1.consoleInfo)("Generated the End User License Agreement (EULA) at: ".concat(toolkit_paths_1.EULA_FILE_PATH));
                    }
                    else {
                        (0, console_1.consoleWarn)('You have declined the terms and conditions.');
                        process.exit(0);
                    }
                    return [3 /*break*/, 3];
                case 2:
                    e_1 = _a.sent();
                    (0, console_1.consoleError)("Unable to process the taco toolkit End User License Agreement.", e_1);
                    process.exit(constants_1.TACO_COMMAND_ERROR_EXIT_CODE);
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/];
            }
        });
    });
}
exports.askForEula = askForEula;
function displayEulaView() {
    return __awaiter(this, void 0, void 0, function () {
        var termsAnswer, answers;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, inquirer_1.default.prompt([
                        {
                            type: 'rawlist',
                            name: 'termsAnswer',
                            message: "You need to agree to the terms and conditions before using taco-toolkit.\n  You can also find the terms and conditions at: \n".concat(toolkit_paths_1.LICENSE_FILE_PATH),
                            choices: Object.values(exports.choices),
                            default: exports.choices.view,
                        },
                    ])];
                case 1:
                    answers = _a.sent();
                    termsAnswer = answers.termsAnswer;
                    if (!(termsAnswer === exports.choices.view)) return [3 /*break*/, 3];
                    return [4 /*yield*/, viewInBrowser()];
                case 2:
                    _a.sent();
                    _a.label = 3;
                case 3:
                    if (termsAnswer === exports.choices.view) return [3 /*break*/, 0];
                    _a.label = 4;
                case 4: return [2 /*return*/, termsAnswer];
            }
        });
    });
}
exports.displayEulaView = displayEulaView;
function generateEula() {
    var eula = {
        time: new Date().toISOString(),
        'taco-toolkit-version': package_json_1.PACKAGE_INFO.version,
        'taco-toolkit-installation-path': toolkit_paths_1.TOOLKIT_ROOT_PATH,
    };
    try {
        fs_extra_1.default.writeJsonSync(toolkit_paths_1.EULA_FILE_PATH, eula, {
            spaces: 2,
            flag: 'w+',
        });
    }
    catch (error) {
        (0, console_1.consoleError)("Unable to generate the taco-toolkit user agreement at: ".concat(toolkit_paths_1.EULA_FILE_PATH), error);
        process.exit(constants_1.TACO_COMMAND_ERROR_EXIT_CODE);
    }
}
exports.generateEula = generateEula;
function hasLicenseAgreement() {
    if (!fs_1.default.existsSync(toolkit_paths_1.EULA_FILE_PATH)) {
        return false;
    }
    var licenseAgreement = fs_extra_1.default.readJSONSync(toolkit_paths_1.EULA_FILE_PATH, { throws: false });
    if (!isValidAgreement(licenseAgreement)) {
        (0, console_1.consoleWarn)("The existing terms and conditions agreement file is invalid. If the End User License Agreement does not automatically show up, please delete ".concat(toolkit_paths_1.EULA_FILE_PATH, " and try again."));
        return false;
    }
    // TODO add a logic to check the eula version, implement this when the eula changes.
    return true;
}
exports.hasLicenseAgreement = hasLicenseAgreement;
function isValidAgreement(value) {
    if (value === null || typeof value !== 'object') {
        return false;
    }
    return 'time' in value && 'taco-toolkit-version' in value && 'taco-toolkit-installation-path' in value;
}
function viewInBrowser() {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    (0, console_1.consoleInfo)("Navigating to ".concat(LICENSE_PAGE, "\nIf the webpage does not open, you can copy the above link and open it with any web browser."));
                    return [4 /*yield*/, (0, open_1.default)(LICENSE_PAGE)];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
