// @tableau/taco-toolkit/dist/eps/sandbox/sdk/parser-utils/csv-utils.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDataRows = exports.parse = void 0;
var papaparse_1 = __importDefault(require("papaparse"));
var sandbox_plugin_1 = require("../sandbox-plugin");
var parse_utils_1 = require("../../../utils/parse-utils");
var CsvUtilsPlugin = /** @class */ (function (_super) {
    __extends(CsvUtilsPlugin, _super);
    function CsvUtilsPlugin() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CsvUtilsPlugin.prototype.init = function () {
        var methods = [
            {
                isStatic: true,
                externalName: 'parse',
                functionRef: parse,
            },
            {
                isStatic: true,
                externalName: 'createDataRows',
                functionRef: createDataRows,
            },
        ];
        this.createClass(methods, 'CsvUtils');
    };
    return CsvUtilsPlugin;
}(sandbox_plugin_1.SandboxPlugin));
exports.default = CsvUtilsPlugin;
function parse(rawData, options) {
    return new Promise(function (resolve, _reject) {
        console.log('>CsvUtils.parse');
        var data;
        if (ArrayBuffer.isView(rawData)) {
            data = new TextDecoder('utf-8').decode(rawData);
        }
        else {
            data = rawData;
        }
        // TODO: figure out why don't we just use the header flag?
        var papaResult = papaparse_1.default.parse(data, { header: false });
        var rows;
        var headers;
        if (options.hasHeader) {
            headers = papaResult.data[0];
            rows = papaResult.data.slice(1);
        }
        else {
            rows = papaResult.data;
        }
        console.log("<CsvUtils.parse returning ".concat(rows.length, " rows."));
        resolve({ rows: rows, headers: headers });
    });
}
exports.parse = parse;
function createDataRows(rows, columnNames) {
    return new Promise(function (resolve, reject) {
        try {
            var dataRows = (0, parse_utils_1.composeDataRows)(rows, columnNames);
            resolve(dataRows);
        }
        catch (e) {
            e.message = "Failed to create DataRows. Caused by: ".concat(e.message);
            reject(e);
        }
    });
}
exports.createDataRows = createDataRows;
