// @tableau/taco-toolkit/dist/eps/sandbox/sdk/parser-utils/eps-utils.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.delay = void 0;
var sandbox_plugin_1 = require("../sandbox-plugin");
//This class will be used to inject some node modules that don't work in the sandbox
//This funationality will be for intenal use at this moment and won't be exposed in the SKD
var EpsUtilsPlugin = /** @class */ (function (_super) {
    __extends(EpsUtilsPlugin, _super);
    function EpsUtilsPlugin() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    EpsUtilsPlugin.prototype.init = function () {
        var methods = [
            {
                isStatic: true,
                externalName: 'delay',
                functionRef: delay,
            },
        ];
        this.createClass(methods, 'EpsUtils');
    };
    return EpsUtilsPlugin;
}(sandbox_plugin_1.SandboxPlugin));
exports.default = EpsUtilsPlugin;
function delay(timeMs) {
    return new Promise(function (resolve) {
        return setTimeout(function () {
            console.log('>EpsUtils.delay');
            resolve(true);
        }, timeMs);
    });
}
exports.delay = delay;
