// @tableau/taco-toolkit/dist/eps/sandbox/sdk/parser-utils/excel-utils.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parse = void 0;
var xlsx = __importStar(require("xlsx"));
var column_header_property_1 = require("../../../../shared/enums/column-header-property");
var data_table_1 = require("../../../../shared/types/data-table");
var sandbox_plugin_1 = require("../sandbox-plugin");
var ExcelUtilsPlugin = /** @class */ (function (_super) {
    __extends(ExcelUtilsPlugin, _super);
    function ExcelUtilsPlugin() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ExcelUtilsPlugin.prototype.init = function () {
        var methods = [
            {
                isStatic: true,
                externalName: 'parse',
                functionRef: parse,
            },
        ];
        this.createClass(methods, 'ExcelUtils');
    };
    return ExcelUtilsPlugin;
}(sandbox_plugin_1.SandboxPlugin));
exports.default = ExcelUtilsPlugin;
function parse(data) {
    console.log('In ExcelUtils parse');
    return new Promise(function (resolve, reject) {
        try {
            var workbook_1 = xlsx.read(Buffer.from(data));
            // Iterate through sheets and create tables for each
            var tables_1 = [];
            workbook_1.SheetNames.forEach(function (sheetName) {
                var _a;
                var table = (0, data_table_1.createDataTable)(sheetName);
                tables_1.push(table);
                var worksheet = workbook_1.Sheets[sheetName];
                if (!worksheet) {
                    throw new Error("Sheet '".concat(sheetName, "' is not parsed properly."));
                }
                var sheetRows = xlsx.utils.sheet_to_json(worksheet);
                if (sheetRows.length === 0) {
                    return;
                }
                // For each sheet, determine headers as the top row, and associate values accordingly
                var columns = Object.keys(sheetRows[0])
                    .filter(function (header) { return header.length > 0; })
                    .map(function (header) {
                    return { id: header, dataType: column_header_property_1.DataType.String };
                });
                (_a = table.columns).push.apply(_a, __spreadArray([], __read(columns), false));
                sheetRows.forEach(function (row) {
                    var dataRow = {};
                    columns.forEach(function (header) {
                        dataRow[header.id] = row[header.id];
                    });
                    table.rows.push(dataRow);
                });
            });
            resolve(tables_1);
        }
        catch (err) {
            reject(err);
        }
    });
}
exports.parse = parse;
