// @tableau/taco-toolkit/dist/eps/sandbox/sdk/parser-utils/parquet-utils.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parse = void 0;
var column_header_property_1 = require("../../../../shared/enums/column-header-property");
var data_table_1 = require("../../../../shared/types/data-table");
var sandbox_plugin_1 = require("../sandbox-plugin");
// This library does not have the ts support so need to have it in require
// eslint-disable-next-line @typescript-eslint/no-var-requires
var parquet = require('parquetjs-lite');
var ParquetUtilsPlugin = /** @class */ (function (_super) {
    __extends(ParquetUtilsPlugin, _super);
    function ParquetUtilsPlugin() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ParquetUtilsPlugin.prototype.init = function () {
        var methods = [
            {
                isStatic: true,
                externalName: 'parse',
                functionRef: parse,
            },
        ];
        this.createClass(methods, 'ParquetUtils');
    };
    return ParquetUtilsPlugin;
}(sandbox_plugin_1.SandboxPlugin));
exports.default = ParquetUtilsPlugin;
// TO DO - Add a unit test for parse function
function parse(data, tableName) {
    return __awaiter(this, void 0, void 0, function () {
        var reader, table, cursor, columns, record_1, _loop_1;
        var _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    console.log('In ParquetUtils parse');
                    return [4 /*yield*/, parquet.ParquetReader.openBuffer(Buffer.from(data))];
                case 1:
                    reader = _b.sent();
                    table = (0, data_table_1.createDataTable)(tableName);
                    _b.label = 2;
                case 2:
                    _b.trys.push([2, , 6, 8]);
                    cursor = reader.getCursor();
                    columns = cursor.schema.fieldList.map(function (h) {
                        return { id: h.name, dataType: column_header_property_1.DataType.String };
                    });
                    (_a = table.columns).push.apply(_a, __spreadArray([], __read(columns), false));
                    record_1 = null;
                    _loop_1 = function () {
                        var dataRow = {};
                        columns.forEach(function (header) {
                            dataRow[header.id] = record_1[header.id];
                        });
                        table.rows.push(dataRow);
                    };
                    _b.label = 3;
                case 3: return [4 /*yield*/, cursor.next()];
                case 4:
                    if (!(record_1 = _b.sent())) return [3 /*break*/, 5];
                    _loop_1();
                    return [3 /*break*/, 3];
                case 5: return [2 /*return*/, table];
                case 6: 
                // Close the Parquet reader
                return [4 /*yield*/, reader.close()];
                case 7:
                    // Close the Parquet reader
                    _b.sent();
                    return [7 /*endfinally*/];
                case 8: return [2 /*return*/];
            }
        });
    });
}
exports.parse = parse;
