// @tableau/taco-toolkit/dist/eps/main/api/grpc/proto/build/eps_grpc_pb.js
// GENERATED CODE -- DO NOT EDIT!

'use strict'
var grpc = require('@grpc/grpc-js')
var eps_pb = require('./eps_pb.js')

function serialize_tableau_eps_v1_AppUrlRequest(arg) {
  if (!(arg instanceof eps_pb.AppUrlRequest)) {
    throw new Error('Expected argument of type tableau.eps.v1.AppUrlRequest')
  }
  return Buffer.from(arg.serializeBinary())
}

function deserialize_tableau_eps_v1_AppUrlRequest(buffer_arg) {
  return eps_pb.AppUrlRequest.deserializeBinary(new Uint8Array(buffer_arg))
}

function serialize_tableau_eps_v1_AppUrlResponse(arg) {
  if (!(arg instanceof eps_pb.AppUrlResponse)) {
    throw new Error('Expected argument of type tableau.eps.v1.AppUrlResponse')
  }
  return Buffer.from(arg.serializeBinary())
}

function deserialize_tableau_eps_v1_AppUrlResponse(buffer_arg) {
  return eps_pb.AppUrlResponse.deserializeBinary(new Uint8Array(buffer_arg))
}

function serialize_tableau_eps_v1_ExtractRequest(arg) {
  if (!(arg instanceof eps_pb.ExtractRequest)) {
    throw new Error('Expected argument of type tableau.eps.v1.ExtractRequest')
  }
  return Buffer.from(arg.serializeBinary())
}

function deserialize_tableau_eps_v1_ExtractRequest(buffer_arg) {
  return eps_pb.ExtractRequest.deserializeBinary(new Uint8Array(buffer_arg))
}

function serialize_tableau_eps_v1_ExtractResponse(arg) {
  if (!(arg instanceof eps_pb.ExtractResponse)) {
    throw new Error('Expected argument of type tableau.eps.v1.ExtractResponse')
  }
  return Buffer.from(arg.serializeBinary())
}

function deserialize_tableau_eps_v1_ExtractResponse(buffer_arg) {
  return eps_pb.ExtractResponse.deserializeBinary(new Uint8Array(buffer_arg))
}

function serialize_tableau_eps_v1_LaunchRequest(arg) {
  if (!(arg instanceof eps_pb.LaunchRequest)) {
    throw new Error('Expected argument of type tableau.eps.v1.LaunchRequest')
  }
  return Buffer.from(arg.serializeBinary())
}

function deserialize_tableau_eps_v1_LaunchRequest(buffer_arg) {
  return eps_pb.LaunchRequest.deserializeBinary(new Uint8Array(buffer_arg))
}

function serialize_tableau_eps_v1_TerminateExtractorRequest(arg) {
  if (!(arg instanceof eps_pb.TerminateExtractorRequest)) {
    throw new Error('Expected argument of type tableau.eps.v1.TerminateExtractorRequest')
  }
  return Buffer.from(arg.serializeBinary())
}

function deserialize_tableau_eps_v1_TerminateExtractorRequest(buffer_arg) {
  return eps_pb.TerminateExtractorRequest.deserializeBinary(new Uint8Array(buffer_arg))
}

function serialize_tableau_eps_v1_TerminateExtractorResponse(arg) {
  if (!(arg instanceof eps_pb.TerminateExtractorResponse)) {
    throw new Error('Expected argument of type tableau.eps.v1.TerminateExtractorResponse')
  }
  return Buffer.from(arg.serializeBinary())
}

function deserialize_tableau_eps_v1_TerminateExtractorResponse(buffer_arg) {
  return eps_pb.TerminateExtractorResponse.deserializeBinary(new Uint8Array(buffer_arg))
}

var EpsService = (exports.EpsService = {
  launch: {
    path: '/tableau.eps.v1.Eps/Launch',
    requestStream: false,
    responseStream: false,
    requestType: eps_pb.LaunchRequest,
    responseType: eps_pb.ExtractResponse,
    requestSerialize: serialize_tableau_eps_v1_LaunchRequest,
    requestDeserialize: deserialize_tableau_eps_v1_LaunchRequest,
    responseSerialize: serialize_tableau_eps_v1_ExtractResponse,
    responseDeserialize: deserialize_tableau_eps_v1_ExtractResponse,
  },
  extract: {
    path: '/tableau.eps.v1.Eps/Extract',
    requestStream: false,
    responseStream: false,
    requestType: eps_pb.ExtractRequest,
    responseType: eps_pb.ExtractResponse,
    requestSerialize: serialize_tableau_eps_v1_ExtractRequest,
    requestDeserialize: deserialize_tableau_eps_v1_ExtractRequest,
    responseSerialize: serialize_tableau_eps_v1_ExtractResponse,
    responseDeserialize: deserialize_tableau_eps_v1_ExtractResponse,
  },
  getAppUrl: {
    path: '/tableau.eps.v1.Eps/GetAppUrl',
    requestStream: false,
    responseStream: false,
    requestType: eps_pb.AppUrlRequest,
    responseType: eps_pb.AppUrlResponse,
    requestSerialize: serialize_tableau_eps_v1_AppUrlRequest,
    requestDeserialize: deserialize_tableau_eps_v1_AppUrlRequest,
    responseSerialize: serialize_tableau_eps_v1_AppUrlResponse,
    responseDeserialize: deserialize_tableau_eps_v1_AppUrlResponse,
  },
  terminateExtractor: {
    path: '/tableau.eps.v1.Eps/TerminateExtractor',
    requestStream: false,
    responseStream: false,
    requestType: eps_pb.TerminateExtractorRequest,
    responseType: eps_pb.TerminateExtractorResponse,
    requestSerialize: serialize_tableau_eps_v1_TerminateExtractorRequest,
    requestDeserialize: deserialize_tableau_eps_v1_TerminateExtractorRequest,
    responseSerialize: serialize_tableau_eps_v1_TerminateExtractorResponse,
    responseDeserialize: deserialize_tableau_eps_v1_TerminateExtractorResponse,
  },
})

exports.EpsClient = grpc.makeGenericClientConstructor(EpsService)
