// @tableau/taco-toolkit/dist/shared/auth-credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getOAuthHeader = exports.getBasicAuthHeader = exports.getAuthHeader = exports.extractOAuthSecretsFromHeader = exports.OAuthCredentials = void 0;
var AUTH_TOKENS;
(function (AUTH_TOKENS) {
    AUTH_TOKENS["accessToken"] = "access-token";
})(AUTH_TOKENS || (AUTH_TOKENS = {}));
/**
 * Extracts OAuth secrets from the provided headers.
 *
 * @param headers - The headers containing OAuth secrets inclding access-token.
 * @returns The extracted OAuth secrets.
 */
function extractOAuthSecretsFromHeader(headers) {
    var _a;
    if (!headers) {
        return undefined;
    }
    var accessToken = headers[AUTH_TOKENS.accessToken];
    if (!accessToken) {
        return undefined;
    }
    return _a = {},
        _a[AUTH_TOKENS.accessToken] = accessToken,
        _a;
}
exports.extractOAuthSecretsFromHeader = extractOAuthSecretsFromHeader;
/**
 * Generates an authentication header using the provided token and optional prefix.
 *
 * @param token - The token used for authentication.
 * @param prefix - The prefix to be added before the token (default is 'Bearer').
 * @returns The authentication header.
 * @example
 * ```
 * //Generating a custom prefix token header
 * const token = 'your-custom-token';
 * const customPrefix = 'Custom';
 * const header = getAuthHeader(token, customPrefix);
 * // header will be: { Authorization: 'Custom your-custom-token' }
 *```
 */
function getAuthHeader(token, prefix) {
    if (prefix === void 0) { prefix = 'Bearer'; }
    return { Authorization: "".concat(prefix, " ").concat(token) };
}
exports.getAuthHeader = getAuthHeader;
/**
 * Generates a Basic authentication header using the provided username and password.
 *
 * @param username - The username for authentication.
 * @param password - The password for authentication.
 * @returns The Basic authentication header.
 *
 * @example
 * ```
 * // Generating a Basic authentication header
 * const username = 'yourUsername';
 * const password = 'yourPassword';
 * const basicAuthHeader = getBasicAuthHeader(username, password);
 * console.log(basicAuthHeader);
 * // Output will be: { Authorization: 'Basic base64EncodedString' }
 * ```
 */
function getBasicAuthHeader(username, password) {
    var token = Buffer.from("".concat(username, ":").concat(password)).toString('base64');
    return getAuthHeader(token, 'Basic');
}
exports.getBasicAuthHeader = getBasicAuthHeader;
/**
 * Generates an OAuth authentication header based on the provided secrets.
 * If no access token is provided in the secrets, an empty object is returned.
 *
 * @param headers - The headers containing access-token.
 * @returns The authentication header.
 */
function getOAuthHeader(headers) {
    if (!headers) {
        return {};
    }
    var oAuthSecrets = extractOAuthSecretsFromHeader(headers);
    var accessToken = oAuthSecrets === null || oAuthSecrets === void 0 ? void 0 : oAuthSecrets[AUTH_TOKENS.accessToken];
    if (accessToken) {
        return getAuthHeader(accessToken);
    }
    else {
        return {};
    }
}
exports.getOAuthHeader = getOAuthHeader;
/**
 * Represents OAuth credentials. includes only access token
 */
var OAuthCredentials = /** @class */ (function () {
    /**
     * Create a new instance of OAuthCredentials.
     *
     * @param headers - The headers containing OAuth secrets. Only `access-token` will be used from the provided headers
     */
    function OAuthCredentials(headers) {
        var oAuthSecrets = extractOAuthSecretsFromHeader(headers);
        this.accessToken = oAuthSecrets === null || oAuthSecrets === void 0 ? void 0 : oAuthSecrets[AUTH_TOKENS.accessToken];
    }
    Object.defineProperty(OAuthCredentials.prototype, "authHeader", {
        /**
         * Get the authentication header based on the access token.
         *
         * @returns The authentication header.
         */
        get: function () {
            if (this.accessToken) {
                return getAuthHeader(this.accessToken);
            }
            else {
                return {};
            }
        },
        enumerable: false,
        configurable: true
    });
    return OAuthCredentials;
}());
exports.OAuthCredentials = OAuthCredentials;
