// @tableau/taco-toolkit/dist/cli/taco-commands/command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addLoggingOptions = exports.createCommand = exports.OPTION_ACCEPT_EULA = void 0;
var commander_1 = require("commander");
var logging_1 = require("./utils/logging");
var toolkit_paths_1 = require("./utils/toolkit-paths");
exports.OPTION_ACCEPT_EULA = '--accept-eula';
function createCommand(name, description, config) {
    var program = new commander_1.Command(name)
        .description(description)
        .configureHelp({
        sortOptions: true,
    })
        .exitOverride();
    var loggingOptions = config.loggingOptions, withCommonOptions = config.withCommonOptions, handler = config.handler, setup = config.setup;
    // add the argument and options in the setup stage
    if (setup) {
        setup(program);
    }
    addCommonOptions(program, withCommonOptions);
    addLoggingOptions(program, loggingOptions);
    // register handlers
    program.action(function () {
        if (loggingOptions) {
            var loggerConfig = createLoggerConfig(program, loggingOptions);
            (0, logging_1.initLogger)(loggerConfig);
        }
        handler(program);
    });
    return program;
}
exports.createCommand = createCommand;
function createLoggerConfig(program, loggingOptions) {
    var options = program.optsWithGlobals();
    var verbose = options.verbose;
    if (loggingOptions.verbose) {
        return {
            verbose: verbose,
        };
    }
    return {};
}
function addLoggingOptions(program, loggingOptions) {
    if (loggingOptions === undefined) {
        return program;
    }
    var verbose = loggingOptions.verbose;
    if (verbose) {
        program.option('-v, --verbose', 'run this command in verbose mode', false);
    }
    return program;
}
exports.addLoggingOptions = addLoggingOptions;
function addCommonOptions(program, withCommonOptions) {
    if (withCommonOptions === false) {
        return program;
    }
    return program.option('--accept-eula', "accept the taco-toolkit end-user license agreement (EULA). The EULA is available in the following location: ".concat(toolkit_paths_1.LICENSE_FILE_PATH));
}
