// @tableau/taco-toolkit/dist/cli/taco-commands/run.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCommand = exports.program = void 0;
var commander_1 = require("commander");
var command_1 = require("./utils/command");
var package_json_1 = require("./utils/package-json");
var spinner_1 = __importDefault(require("./utils/spinner"));
var command_2 = require("./command");
var launch_1 = require("./run/launch");
var console_1 = require("./utils/console");
var tableau_product_1 = require("./utils/tableau-product");
var tableau_product_2 = require("./utils/tableau-product");
exports.program = createCommand();
function createCommand() {
    return (0, command_2.createCommand)('run', 'launch Tableau app with taco connectors', {
        setup: setup,
        handler: handler,
        withCommonOptions: true,
        loggingOptions: {
            verbose: true,
        },
    });
}
exports.createCommand = createCommand;
function setup(program) {
    return program
        .addArgument(new commander_1.Argument('[tableau-app-name]', 'specify the Tableau app to launch')
        .choices([tableau_product_1.ProductName.Desktop, tableau_product_1.ProductName.Prep])
        .default(tableau_product_1.ProductName.Desktop))
        .option('-V, --version <version>', 'specify the version number for Tableau app to be launched')
        .option('-ap, --app-path <tableau-app-path>', "specify the path to Tableau app's executable; when this option is present, [tableau-app-name] and --version will be ignored")
        .option('-td, --taco-dir <taco-dir-path>', 'specify the directory path that contains taco connector(s); the default value is current working directory')
        .addOption(new commander_1.Option('--eps-path <eps-path>', 'specific the path to a EPS executable').hideHelp());
}
function handler(program) {
    return __awaiter(this, void 0, void 0, function () {
        var options, epsPath, appPath, tacoDir, verbose, launchConfig;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    options = program.optsWithGlobals();
                    epsPath = options.epsPath, appPath = options.appPath, tacoDir = options.tacoDir, verbose = options.verbose;
                    if (epsPath !== undefined) {
                        (0, command_1.exitIfNonExistPath)(epsPath, 'Unable to find the path EPS executable path (--eps-path)');
                    }
                    if (appPath !== undefined) {
                        (0, command_1.exitIfNonExistPath)(appPath, 'Unable to find Tableau app executable (--app-path)');
                    }
                    if (tacoDir !== undefined) {
                        (0, command_1.exitIfNonExistPath)(tacoDir, 'Unable to find taco directory path (--taco-dir)');
                    }
                    launchConfig = getLaunchConfig(program);
                    return [4 /*yield*/, (0, launch_1.launchTableau)(launchConfig, {
                            spinner: new spinner_1.default(verbose),
                            verbose: verbose,
                        })];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
function getLaunchConfig(program) {
    var _a = program.opts(), version = _a.version, epsPath = _a.epsPath, appPath = _a.appPath, tacoDir = _a.tacoDir;
    var tacoFolderPath = tacoDir !== null && tacoDir !== void 0 ? tacoDir : process.cwd();
    if (appPath) {
        if (program.args[0] || version) {
            (0, console_1.consoleWarn)('Warning: --app-path is provided, ignoring [tableau-app-name] and --version.');
        }
        return {
            appPath: appPath,
            tacoFolderPath: tacoFolderPath,
            epsPath: epsPath,
        };
    }
    var _b = __read(program.processedArgs, 1), appName = _b[0];
    var tableauVersion = version !== null && version !== void 0 ? version : (0, tableau_product_2.getTableauMajorVersion)(package_json_1.PACKAGE_INFO.tableau.version);
    return {
        appName: appName,
        appVersion: tableauVersion,
        tacoFolderPath: tacoFolderPath,
        epsPath: epsPath,
    };
}
