// @tableau/taco-toolkit/dist/cli/taco-commands/show.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCommand = exports.program = void 0;
var adm_zip_1 = __importDefault(require("adm-zip"));
var command_1 = require("./command");
var connector_1 = require("./utils/connector");
var logging_1 = require("./utils/logging");
var console_1 = require("./utils/console");
var command_2 = require("./utils/command");
var process_1 = __importDefault(require("process"));
var chalk_1 = __importDefault(require("chalk"));
exports.program = createCommand();
function createCommand() {
    return (0, command_1.createCommand)('show', 'show the taco-toolkit related information', {
        withCommonOptions: false,
        setup: setup,
        handler: handler,
    });
}
exports.createCommand = createCommand;
function setup(program) {
    return program
        .option('--log-file', 'show the location of taco operation logs')
        .option('--taco-info <taco-file-path>', 'show the versioning information of taco connector');
}
function handler(module) {
    var options = module.optsWithGlobals();
    if (options.logFile) {
        showLogPath();
        return;
    }
    var tacoInfo = options.tacoInfo;
    if (tacoInfo) {
        showTacoInformation(tacoInfo);
        return;
    }
    module.help();
}
function showLogPath() {
    var tacoRootPath = (0, connector_1.findTacoRootPath)();
    (0, console_1.consoleInfo)((0, logging_1.getLogPath)(tacoRootPath));
}
function showTacoInformation(tacoFilePath) {
    var _a;
    if (!(0, connector_1.isValidTacoFile)(tacoFilePath)) {
        (0, console_1.consoleError)("The target file '".concat(tacoFilePath, "' is not a valid .taco file."));
        process_1.default.exit(1);
    }
    (0, command_2.exitIfNonExistPath)(tacoFilePath, "The taco file '".concat(tacoFilePath, "' could not be found at current path: '").concat(process_1.default.cwd(), "'"));
    var tacoInfo = {};
    var zip = new adm_zip_1.default(tacoFilePath);
    var connectorJsonEntry = zip.getEntry('connector/connector.json');
    if (!connectorJsonEntry) {
        throw Error("Could not locate the connector.json inside ".concat(tacoFilePath));
    }
    var connectorJson = JSON.parse(connectorJsonEntry.getData().toString('utf8'));
    var tableauVersion = connectorJson["tableau-version"], version = connectorJson.version;
    tacoInfo.tableauVersion = tableauVersion;
    tacoInfo.version = version;
    var packageJsonEntry = zip.getEntry('connector/package.json');
    if (!packageJsonEntry) {
        throw Error("Could not locate the package.json inside ".concat(tacoFilePath));
    }
    var packageJson = JSON.parse(packageJsonEntry.getData().toString('utf8'));
    var dependencies = packageJson.dependencies;
    tacoInfo.toolkitVersion = dependencies === null || dependencies === void 0 ? void 0 : dependencies['@tableau/taco-toolkit'];
    console.info("TACO connector \"".concat(chalk_1.default.green(tacoFilePath), "\" information:\n") +
        "  Connector version:       ".concat(chalk_1.default.green(tacoInfo.version), "\n") +
        "  Tableau minimum version: ".concat(chalk_1.default.green((_a = tacoInfo.tableauVersion) === null || _a === void 0 ? void 0 : _a.min), "\n") +
        "  Built with taco-toolkit: ".concat(chalk_1.default.green('@tableau/taco-toolkit@' + tacoInfo.toolkitVersion)));
}
