// @tableau/taco-toolkit/dist/cli/taco-commands/sign.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCommand = exports.program = void 0;
var process_1 = __importDefault(require("process"));
var child_process_1 = require("child_process");
var command_1 = require("./command");
var connector_1 = require("./utils/connector");
var console_1 = require("./utils/console");
var sign_1 = require("./sign/sign");
var command_2 = require("./utils/command");
var spinner_1 = __importDefault(require("./utils/spinner"));
exports.program = createCommand();
function createCommand() {
    return (0, command_1.createCommand)('sign', 'sign packaged connector', {
        setup: setup,
        handler: handler,
        withCommonOptions: true,
        loggingOptions: {
            verbose: true,
        },
    });
}
exports.createCommand = createCommand;
function setup(program) {
    return program
        .option('-ks, --keystore <path>', 'specify the keystore path')
        .option('-g, --gen-csr', 'generate a certificate signing request (CSR)');
}
function handler(program) {
    return __awaiter(this, void 0, void 0, function () {
        var _a, keystore, genCsr, verbose, currentPath, state;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _a = program.optsWithGlobals(), keystore = _a.keystore, genCsr = _a.genCsr, verbose = _a.verbose;
                    if (!keystore && !genCsr) {
                        (0, console_1.consoleError)('Path to keystore was not provided and the --gen-csr flag was not used.');
                        process_1.default.exit(1);
                    }
                    if (!isKeytoolInstalled() || !isJarsignerInstalled()) {
                        (0, console_1.consoleError)('Missing required JDK tools. Please download and install JDK, and add it to the PATH environment variable.');
                        process_1.default.exit(1);
                    }
                    currentPath = process_1.default.cwd();
                    if (!(0, connector_1.isTacoRootPath)(currentPath)) {
                        (0, connector_1.outputLocationErrorAndExit)(currentPath);
                    }
                    state = {
                        spinner: new spinner_1.default(verbose),
                        verbose: verbose,
                    };
                    if (!genCsr) return [3 /*break*/, 1];
                    if (keystore !== undefined) {
                        (0, console_1.consoleWarn)('Warning: --gen-csr is provided, ignoring --keystore.');
                    }
                    (0, sign_1.generateCsr)({
                        tacoRootPath: currentPath,
                    }, state);
                    return [3 /*break*/, 3];
                case 1:
                    (0, command_2.exitIfNonExistPath)(keystore, 'Keystore path does not exist (--keystore).');
                    return [4 /*yield*/, (0, sign_1.signTacoFile)({
                            currentPath: currentPath,
                            keystorePath: keystore,
                        }, state)];
                case 2:
                    _b.sent();
                    _b.label = 3;
                case 3: return [2 /*return*/];
            }
        });
    });
}
function isKeytoolInstalled() {
    var output = (0, child_process_1.spawnSync)('keytool').output;
    return output.toString().includes('Key and Certificate Management Tool');
}
function isJarsignerInstalled() {
    var output = (0, child_process_1.spawnSync)('jarsigner').output;
    return output.toString().includes('Usage: jarsigner');
}
