// @tableau/taco-toolkit/dist/cli/taco-commands/unpack.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCommand = exports.program = void 0;
var command_1 = require("./command");
var unpack_1 = require("./unpack/unpack");
var command_2 = require("./utils/command");
var console_1 = require("./utils/console");
var spinner_1 = __importDefault(require("./utils/spinner"));
var connector_1 = require("./utils/connector");
exports.program = createCommand();
function createCommand() {
    return (0, command_1.createCommand)('unpack', 'unpack the .taco file into the current directory', {
        setup: setup,
        handler: handler,
        withCommonOptions: true,
        loggingOptions: {
            verbose: true,
        },
    });
}
exports.createCommand = createCommand;
function setup(program) {
    return program
        .argument('<taco>', 'packed .taco filename')
        .option('-d, --dest <path>', 'unpack connector to a certain location');
}
function handler(program) {
    var _a = __read(program.processedArgs, 1), tacoName = _a[0];
    var _b = program.optsWithGlobals(), dest = _b.dest, verbose = _b.verbose;
    if (!(0, connector_1.isValidTacoFile)(tacoName)) {
        (0, console_1.consoleError)("The target file '".concat(tacoName, "' is not a valid .taco file."));
        process.exit(1);
    }
    (0, command_2.exitIfNonExistPath)(tacoName, "The taco file '".concat(tacoName, "' could not be found at current path '").concat(process.cwd(), "'"));
    if (dest) {
        (0, command_2.exitIfNonExistPath)(dest, "The taco unpack destination '".concat(dest, "' could not be found or inaccessible"));
    }
    (0, unpack_1.unpackTaco)({
        tacoName: tacoName,
        destPath: dest !== null && dest !== void 0 ? dest : process.cwd(),
    }, {
        spinner: new spinner_1.default(verbose),
        verbose: verbose,
    });
}
