// @tableau/taco-toolkit/dist/cli/taco-commands/create/git-setup.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupGitRepo = void 0;
var child_process_1 = require("child_process");
var fs_1 = __importDefault(require("fs"));
var path_1 = __importDefault(require("path"));
var command_1 = require("../utils/command");
var console_1 = require("../utils/console");
function setupGitRepo(tacoRootPath) {
    if (!isGitInstalled()) {
        (0, console_1.consoleError)('Unable to find Git. Please install Git first.');
        process.exit(1);
    }
    var result = (0, command_1.executeCmd)('git', ['init'], { cwd: tacoRootPath });
    if ((0, command_1.isErrorResult)(result)) {
        var _a = result.error, error = _a.error, stdout = _a.stdout;
        (0, console_1.consoleError)("Failed to initialize a Git repo. ".concat(stdout), error);
        process.exit(1);
    }
    try {
        createGitIgnoreFile(tacoRootPath);
        (0, child_process_1.execSync)('git add -A', {
            cwd: tacoRootPath,
            stdio: 'ignore',
        });
        (0, child_process_1.execSync)('git commit -m "init commit for taco project"', {
            cwd: tacoRootPath,
            stdio: 'ignore',
        });
    }
    catch (error) {
        (0, console_1.consoleError)('Failed to create the init commit.', error);
        process.exit(1);
    }
}
exports.setupGitRepo = setupGitRepo;
function isGitInstalled() {
    var output = (0, child_process_1.spawnSync)('git', ['--version']).output;
    return output.toString().includes('git version');
}
function createGitIgnoreFile(tacoRootPath) {
    var filePath = path_1.default.join(tacoRootPath, '.gitignore');
    var entries = [
        '# OS artifacts',
        '.DS_Store\n',
        '# Tools artifacts',
        'node_modules',
        '.parcel-cache\n',
        '# Project artifacts',
        'dist',
        'log',
        'tmp',
        '.eps',
    ].join('\n');
    fs_1.default.writeFileSync(filePath, entries, {
        flag: 'w',
    });
}
