// @tableau/taco-toolkit/dist/cli/taco-commands/create/json-files.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fillConnectorJson = exports.fillPackageJson = void 0;
var fs_extra_1 = __importDefault(require("fs-extra"));
var path_1 = __importDefault(require("path"));
var connector_1 = require("../utils/connector");
var package_json_1 = require("../utils/package-json");
var tableau_product_1 = require("../utils/tableau-product");
function fillPackageJson(tacoRootPath, projectName, skipToolkit) {
    var connectorFolder = (0, connector_1.getConnectorFolderPath)(tacoRootPath);
    var packageJsonPath = path_1.default.join(connectorFolder, 'package.json');
    var props = {
        name: projectName,
    };
    if (!skipToolkit) {
        Object.assign(props, {
            dependencies: {
                '@tableau/taco-toolkit': package_json_1.PACKAGE_INFO.version,
            },
            scripts: {
                build: 'taco-scripts build',
                clean: 'taco-scripts clean',
                start: 'taco-scripts start',
            },
        });
    }
    fillInfoToJson(packageJsonPath, props);
}
exports.fillPackageJson = fillPackageJson;
function fillConnectorJson(tacoRootPath, projectName) {
    var connectorJsonPath = (0, connector_1.getConnectorJsonPath)(tacoRootPath);
    // automatically put tableau min version into connector json
    fillInfoToJson(connectorJsonPath, {
        name: projectName,
        version: '1.0.0',
        'tableau-version': {
            min: (0, tableau_product_1.getTableauMajorVersion)(package_json_1.PACKAGE_INFO.tableau.version),
        },
    });
}
exports.fillConnectorJson = fillConnectorJson;
function fillInfoToJson(filePath, info) {
    var json = fs_extra_1.default.readJsonSync(filePath);
    mergeJsonObject(json, info);
    fs_extra_1.default.writeJSONSync(filePath, json, {
        spaces: 2,
        flag: 'w+',
    });
}
function mergeJsonObject(target, source) {
    var e_1, _a;
    try {
        for (var _b = __values(Object.entries(source)), _c = _b.next(); !_c.done; _c = _b.next()) {
            var _d = __read(_c.value, 2), key = _d[0], value = _d[1];
            if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
                if (!target[key]) {
                    target[key] = {};
                }
                mergeJsonObject(target[key], source[key]);
            }
            else {
                target[key] = source[key];
            }
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        }
        finally { if (e_1) throw e_1.error; }
    }
}
