// @tableau/taco-toolkit/dist/eps/sandbox/sdk/fetcher-utils/aws-utils.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
Object.defineProperty(exports, "__esModule", { value: true });
var client_s3_1 = require("@aws-sdk/client-s3");
var permission_1 = require("../../../../shared/types/permission");
var sandbox_plugin_1 = require("../sandbox-plugin");
var AwsUtils = /** @class */ (function (_super) {
    __extends(AwsUtils, _super);
    function AwsUtils() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AwsUtils.prototype.init = function () {
        var methods = [
            {
                isStatic: true,
                externalName: 'selectObjectContent',
                functionRef: createSelectObjectContent(this),
            },
        ];
        this.createClass(methods, 'AwsUtils');
    };
    return AwsUtils;
}(sandbox_plugin_1.SandboxPlugin));
exports.default = AwsUtils;
function checkPermission(thisPlugin, url, method) {
    if (thisPlugin.sandbox.getPermission()) {
        var permission = thisPlugin.sandbox.getPermission();
        if (!permission || !permission_1.PermissionChecker.checkApiPermission(permission.api, url, method)) {
            console.log('todo: perm error');
            // throw new Error(`No permission to make ${method} request URL ${url}`)
        }
    }
}
function createSelectObjectContent(thisPlugin) {
    var _this = this;
    return function (options) { return __awaiter(_this, void 0, void 0, function () {
        var region, accessKeyId, secretAccessKey, _a, bucket, key, method, url, s3Client, resp, contents, _b, _c, _d, event_1, recordsPayload, e_1_1;
        var _e, e_1, _f, _g;
        var _h;
        return __generator(this, function (_j) {
            switch (_j.label) {
                case 0:
                    region = options.region, accessKeyId = options.accessKeyId, secretAccessKey = options.secretAccessKey;
                    {
                        _a = options.commandInput, bucket = _a.Bucket, key = _a.Key;
                        method = 'POST';
                        url = "https://".concat(bucket, ".s3.").concat(region, ".amazonaws.com/").concat(key, "?");
                        checkPermission(thisPlugin, url, method);
                    }
                    s3Client = new client_s3_1.S3Client({ region: region, credentials: { accessKeyId: accessKeyId, secretAccessKey: secretAccessKey } });
                    return [4 /*yield*/, s3Client.send(new client_s3_1.SelectObjectContentCommand(options.commandInput))];
                case 1:
                    resp = _j.sent();
                    contents = [];
                    if (!(resp.Payload !== undefined)) return [3 /*break*/, 13];
                    _j.label = 2;
                case 2:
                    _j.trys.push([2, 7, 8, 13]);
                    _b = true, _c = __asyncValues(resp.Payload);
                    _j.label = 3;
                case 3: return [4 /*yield*/, _c.next()];
                case 4:
                    if (!(_d = _j.sent(), _e = _d.done, !_e)) return [3 /*break*/, 6];
                    _g = _d.value;
                    _b = false;
                    try {
                        event_1 = _g;
                        recordsPayload = (_h = event_1 === null || event_1 === void 0 ? void 0 : event_1.Records) === null || _h === void 0 ? void 0 : _h.Payload;
                        if (recordsPayload) {
                            contents.push(Buffer.from(recordsPayload).toString());
                        }
                    }
                    finally {
                        _b = true;
                    }
                    _j.label = 5;
                case 5: return [3 /*break*/, 3];
                case 6: return [3 /*break*/, 13];
                case 7:
                    e_1_1 = _j.sent();
                    e_1 = { error: e_1_1 };
                    return [3 /*break*/, 13];
                case 8:
                    _j.trys.push([8, , 11, 12]);
                    if (!(!_b && !_e && (_f = _c.return))) return [3 /*break*/, 10];
                    return [4 /*yield*/, _f.call(_c)];
                case 9:
                    _j.sent();
                    _j.label = 10;
                case 10: return [3 /*break*/, 12];
                case 11:
                    if (e_1) throw e_1.error;
                    return [7 /*endfinally*/];
                case 12: return [7 /*endfinally*/];
                case 13: return [2 /*return*/, contents.join('').trimEnd()];
            }
        });
    }); };
}
