// @tableau/taco-toolkit/dist/eps/main/api/grpc/client/eps-grpc-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EpsGrpcClient = exports.createEpsGrpcClient = void 0;
var grpc_js_1 = require("@grpc/grpc-js");
var eps_grpc_pb_1 = require("../proto/build/eps_grpc_pb");
var eps_pb_1 = require("../proto/build/eps_pb");
//TODO: change credential to be secure?
function createEpsGrpcClient(grpcPort) {
    return new eps_grpc_pb_1.EpsClient("localhost:".concat(grpcPort), grpc_js_1.credentials.createInsecure());
}
exports.createEpsGrpcClient = createEpsGrpcClient;
var EpsGrpcClient = /** @class */ (function () {
    function EpsGrpcClient(grpcPort) {
        this.client = createEpsGrpcClient(grpcPort);
    }
    EpsGrpcClient.prototype.getAppUrl = function (_a) {
        var _this = this;
        var connectorClass = _a.connectorClass, connectionId = _a.connectionId, callerType = _a.callerType, pluginPath = _a.pluginPath;
        return new Promise(function (resolve, reject) {
            var request = new eps_pb_1.AppUrlRequest();
            request.setConnectionId(connectionId);
            request.setCallerType(callerType);
            request.setConnectorClass(connectorClass);
            request.setPluginPath(pluginPath);
            _this.client.getAppUrl(request, function (error, response) {
                if (error) {
                    reject(error);
                }
                else {
                    resolve(response.toObject().connectorAppUrl);
                }
            });
        });
    };
    EpsGrpcClient.prototype.launch = function (_a) {
        var _this = this;
        var connectorClass = _a.connectorClass, connectionId = _a.connectionId, callerId = _a.callerId, pluginPath = _a.pluginPath, connectionData = _a.connectionData, epsInstanceId = _a.epsInstanceId, extractorMode = _a.extractorMode, language = _a.language, locale = _a.locale, accessToken = _a.accessToken, refreshToken = _a.refreshToken;
        return new Promise(function (resolve, reject) {
            var request = new eps_pb_1.LaunchRequest();
            var connData = JSON.stringify(connectionData);
            request.setConnectionId(connectionId);
            request.setCallerId(callerId);
            request.setConnectorClass(connectorClass);
            request.setPluginPath(pluginPath);
            request.setConnectionData(connData);
            request.setEpsInstanceId(epsInstanceId);
            request.setExtractorMode(extractorMode);
            request.setLanguage(language);
            request.setLocale(locale);
            if (accessToken) {
                request.setAccessToken(accessToken);
            }
            if (refreshToken) {
                request.setRefreshToken(refreshToken);
            }
            _this.client.launch(request, function (error, response) {
                if (error) {
                    reject(error);
                }
                else {
                    resolve(response);
                }
            });
        });
    };
    EpsGrpcClient.prototype.extract = function (_a) {
        var _this = this;
        var connectorClass = _a.connectorClass, connectionId = _a.connectionId, callerId = _a.callerId, pluginPath = _a.pluginPath, connectionData = _a.connectionData, epsInstanceId = _a.epsInstanceId, language = _a.language, locale = _a.locale, accessToken = _a.accessToken, refreshToken = _a.refreshToken;
        return new Promise(function (resolve, reject) {
            var request = new eps_pb_1.ExtractRequest();
            var connData = JSON.stringify(connectionData);
            request.setConnectionId(connectionId);
            request.setCallerId(callerId);
            request.setConnectorClass(connectorClass);
            request.setPluginPath(pluginPath);
            request.setConnectionData(connData);
            request.setEpsInstanceId(epsInstanceId);
            request.setLanguage(language);
            request.setLocale(locale);
            if (accessToken) {
                request.setAccessToken(accessToken);
            }
            if (refreshToken) {
                request.setRefreshToken(refreshToken);
            }
            _this.client.extract(request, function (error, response) {
                if (error) {
                    console.log(error);
                    reject(error);
                }
                else {
                    resolve(response);
                }
            });
        });
    };
    EpsGrpcClient.prototype.terminateExtractor = function (_a) {
        var _this = this;
        var connectionId = _a.connectionId, extractorIds = _a.extractorIds;
        return new Promise(function (resolve, reject) {
            var request = new eps_pb_1.TerminateExtractorRequest();
            request.setConnectionId(connectionId);
            request.setExtractorIdList(extractorIds);
            _this.client.terminateExtractor(request, function (error, response) {
                if (error) {
                    console.log('terminateExtractor error :>> ', error);
                    reject(error);
                }
                else {
                    resolve(response);
                }
            });
        });
    };
    return EpsGrpcClient;
}());
exports.EpsGrpcClient = EpsGrpcClient;
