// @tableau/taco-toolkit/dist/eps/utils/api-utils.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLocalizationHeader = exports.hasLocalizationHeader = exports.notFoundHandler = exports.errorHandler = exports.sendInternalServerError = exports.sendNotFound = exports.sendBadRequest = exports.getOptionalStringHeader = exports.getRequiredStringHeader = exports.getOptionalTacoPathHeader = exports.getRequiredTacoPathHeader = exports.getConnectionData = exports.getOptionalConnectionData = exports.getSingleOptionalHeader = exports.getSingleRequiredHeader = exports.maskURL = exports.getRequestLogger = void 0;
var http_status_codes_1 = require("http-status-codes");
var eps_log_serializer_1 = require("../logging/eps-log-serializer");
var eps_logger_1 = require("../logging/eps-logger");
var extract_1 = require("../main/api/eps-handlers/extract");
var LOG_TPYE = 'received request';
var TACO_PATH_HEADER = 'plugin-path';
function getRequestLogger(listeningServer, replacer) {
    return function (req, _res, next) {
        var method = req.method, url = req.url, headers = req.headers;
        if (replacer) {
            var info = replacer(req);
            // if replace returns undefined, use the default info
            if (info !== undefined) {
                eps_logger_1.Logger.info({
                    listeningServer: listeningServer,
                    method: method,
                    type: LOG_TPYE,
                    request: info,
                });
                next();
                return;
            }
        }
        eps_logger_1.Logger.info({
            listeningServer: listeningServer,
            method: method,
            type: LOG_TPYE,
            url: maskURL(url),
            headers: maskHeaders(headers),
        });
        next();
    };
}
exports.getRequestLogger = getRequestLogger;
var SECRETS_PARAM = 'secrets';
/**
 * Mask secret query params in the url for logging
 */
function maskURL(url) {
    if (!url.includes(SECRETS_PARAM)) {
        return url;
    }
    var params = new URLSearchParams(url);
    if (params.get('secrets')) {
        params.set('secrets', eps_log_serializer_1.MASK_VALUE);
    }
    return decodeURIComponent(params.toString());
}
exports.maskURL = maskURL;
/**
 * Mask secrets in headers for logging.
 * If you are adding new fields to this function, check the eps-log-serializer first.
 * eps-log-serializer will mask the entire sensitive field value, while this function
 * allows us to add logic to partially mask some infos.
 *
 * It returns the original object if there is sensitive fields exist, while
 * it returns a shallow copy if any value is masked.
 */
function maskHeaders(headers) {
    var referer = headers.referer, otherHeaders = __rest(headers, ["referer"]);
    if (!referer || !referer.includes(SECRETS_PARAM)) {
        return headers;
    }
    // if it has secrets, create a copy for masking fields
    var newHeaders = __assign({}, otherHeaders);
    if (referer) {
        newHeaders.referer = maskURL(referer);
    }
    return newHeaders;
}
/**
 * Ensures that there is exactly one of the header and returns its value.
 */
var getSingleRequiredHeader = function (name, req) {
    var val = (0, exports.getSingleOptionalHeader)(name, req);
    if (val === undefined) {
        var error = new Error("'".concat(name, "' header is undefined"));
        eps_logger_1.Logger.error(error);
        throw error;
    }
    return val;
};
exports.getSingleRequiredHeader = getSingleRequiredHeader;
/**
 * Ensures that there is at most one of the header and returns its value or undefined.
 * Throws if there are multiple entries.
 */
var getSingleOptionalHeader = function (name, req) {
    var val = req.header(name);
    if (val !== undefined) {
        if (val.indexOf(',') !== -1) {
            // Count even entries matching name. Evens are the header names; odds are the header values.
            var count = req.rawHeaders.filter(function (v, i) { return i % 2 === 0 && v === name; }).length;
            if (count !== 1) {
                // Duplicates were combined like plugin-path: "File1, File1", converted to array then joined to a string.
                var error = new Error("'".concat(name, "' header was provided more than once: got ").concat(count, " instances."));
                eps_logger_1.Logger.error(error);
                throw error;
            }
        }
    }
    return val;
};
exports.getSingleOptionalHeader = getSingleOptionalHeader;
function getOptionalConnectionData(req) {
    var body = req.body;
    if (!body ||
        // API client (e.g. superagent) may add empty object by default for POST
        (typeof body === 'object' && Object.keys(body).length === 0)) {
        return;
    }
    return getConnectionData(req);
}
exports.getOptionalConnectionData = getOptionalConnectionData;
function getConnectionData(req) {
    var body = req.body;
    (0, extract_1.validateConnectionData)(body);
    return body;
}
exports.getConnectionData = getConnectionData;
function getRequiredTacoPathHeader(req) {
    var value = getOptionalTacoPathHeader(req);
    if (value === undefined) {
        throw new Error("Header '".concat(TACO_PATH_HEADER, "' must be present."));
    }
    return value;
}
exports.getRequiredTacoPathHeader = getRequiredTacoPathHeader;
function getOptionalTacoPathHeader(req) {
    var value = getOptionalStringHeader(req, TACO_PATH_HEADER, {
        allowEmpty: false,
    });
    if (!value) {
        return value;
    }
    return decodeURIComponent(value);
}
exports.getOptionalTacoPathHeader = getOptionalTacoPathHeader;
/**
 * Retrieve a string value from a required request header.
 * It validates the value according to the options, and throws an Error when validation fails.
 */
function getRequiredStringHeader(req, headerName, options) {
    var value = getOptionalStringHeader(req, headerName, options);
    if (value === undefined) {
        throw new Error("Header '".concat(headerName, "' must be present."));
    }
    return value;
}
exports.getRequiredStringHeader = getRequiredStringHeader;
/**
 * Retrieve a string value from a required request header.
 * It validates the value according to the options, and throws an Error when validation fails.
 */
function getOptionalStringHeader(req, headerName, options) {
    var value = req.headers[headerName];
    if (value === undefined) {
        return value;
    }
    if (typeof value !== 'string') {
        throw new Error("Header '".concat(headerName, "' must be a string, but receieved ").concat(value, "."));
    }
    var allowEmpty = options.allowEmpty;
    if (allowEmpty === false && value.length === 0) {
        throw new Error("Header '".concat(headerName, "' must not be a empty string."));
    }
    return value;
}
exports.getOptionalStringHeader = getOptionalStringHeader;
function sendBadRequest(res, message) {
    eps_logger_1.Logger.debug(message);
    if (message instanceof Error) {
        message = message.toString();
    }
    return res.status(http_status_codes_1.StatusCodes.BAD_REQUEST).json({ error: { message: message } });
}
exports.sendBadRequest = sendBadRequest;
function sendNotFound(res, message) {
    eps_logger_1.Logger.debug(message);
    return res.status(http_status_codes_1.StatusCodes.NOT_FOUND).send({ message: message });
}
exports.sendNotFound = sendNotFound;
function sendInternalServerError(res, error) {
    eps_logger_1.Logger.error(error);
    if (error instanceof Error) {
        error = error.toString();
    }
    return res.status(http_status_codes_1.StatusCodes.INTERNAL_SERVER_ERROR).json({
        error: {
            message: "Internal Server Error: ".concat(error),
        },
    });
}
exports.sendInternalServerError = sendInternalServerError;
function errorHandler(err, _req, res) {
    sendInternalServerError(res, err);
}
exports.errorHandler = errorHandler;
function notFoundHandler(_req, res) {
    res.status(http_status_codes_1.StatusCodes.NOT_FOUND).send('Not Found');
}
exports.notFoundHandler = notFoundHandler;
/**
 * Check if request has certain localization header value.
 */
function hasLocalizationHeader(req) {
    return 'language' in req.headers || 'locale' in req.headers;
}
exports.hasLocalizationHeader = hasLocalizationHeader;
/**
 * Get i18n header value from the request headers. If the header is absent, it returns the given default value.
 */
function getLocalizationHeader(req, headerName, defaultValue) {
    if (!req.headers) {
        return defaultValue;
    }
    var value = req.headers[headerName];
    if (!value) {
        return defaultValue;
    }
    if (typeof value !== 'string') {
        eps_logger_1.Logger.error("[i18n] localization context header unexpected for ".concat(headerName, ": ").concat(value));
        return defaultValue;
    }
    return value;
}
exports.getLocalizationHeader = getLocalizationHeader;
