// @tableau/taco-toolkit/dist/eps/extractor/extractor-app-config.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getExtractorAppConfig = void 0;
var app_config_utils_1 = require("../utils/app-config-utils");
var appConfig;
/**
 * Compose Extractor app config from process args and environment variables.
 *
 * Wrap the logic into a function to trigger the value parsing on demand.
 */
function getExtractorAppConfig() {
    if (appConfig) {
        return appConfig;
    }
    /**** options from process arguments ****/
    var _a = __read(process.argv, 5), extractorId = _a[2], extractorOptions = _a[3], portConfig = _a[4];
    /**** options from environment variables ****/
    // EPS child app env configs
    var epsInstanceId = (0, app_config_utils_1.getRequiredEnvVariable)('EPS_INSTANCE_ID');
    var callerId = (0, app_config_utils_1.getRequiredEnvVariable)('EPS_CALLER_ID');
    var epsTempDirPath = (0, app_config_utils_1.getRequiredEnvVariable)('EPS_TEMP_DIR_PATH');
    var logOption = (0, app_config_utils_1.getRequiredEnvVariable)('EPS_APP_LOG_CONFIG');
    var networkOption = (0, app_config_utils_1.getRequiredEnvVariable)('EPS_NETWORK_CONFIG');
    appConfig = {
        epsInstanceId: epsInstanceId,
        callerId: callerId,
        extractorId: extractorId,
        epsTempDirPath: epsTempDirPath,
        networkAdapterOptions: JSON.parse(networkOption),
        extractorOptions: JSON.parse(extractorOptions),
        portConfig: JSON.parse(portConfig),
        logOption: (0, app_config_utils_1.parseAppLogOption)(logOption),
    };
    return appConfig;
}
exports.getExtractorAppConfig = getExtractorAppConfig;
