// @tableau/taco-toolkit/dist/eps/extractor/handler-resolver.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isHandlerConfig = exports.resolveHandlers = void 0;
var change_case_1 = require("change-case");
var node_fs_1 = __importDefault(require("node:fs"));
var node_path_1 = __importDefault(require("node:path"));
var taco_file_parser_1 = require("../../shared/enums/taco-file-parser");
var type_utils_1 = require("../../shared/type-utils");
var eps_logger_1 = require("../logging/eps-logger");
var handlersMetadata = {
    auth: {
        required: false,
    },
    fetcher: {
        required: true,
    },
    parser: {
        required: true,
    },
};
/**
 * Resolve the handlerInput from the connector app into the addressable information.
 *
 * @param handlersDirPath the connection handler directory path
 * @param handlerInput the handlerInput object from the connector app. This is an untrusted input from userland.
 * @returns an object that contains resolved information of the handlers
 */
function resolveHandlers(handlersDirPath, handlerInput) {
    var handlers = {
        // currently, everything is using memory store
        storagePath: 'memory-store',
    };
    (0, type_utils_1.getTypedKeys)(handlersMetadata).forEach(function (handlerType) {
        var required = handlersMetadata[handlerType].required;
        var handlerName = handlerInput[handlerType];
        if (!handlerName) {
            if (required) {
                throw new Error("Required handlerInput property '".concat(handlerType, "' is not specified."));
            }
            else {
                // skip optional properties if not present
                return;
            }
        }
        if (typeof handlerName !== 'string') {
            throw new Error("The value of handlerInput property '".concat(handlerType, "' must be string"));
        }
        if (handlerType === 'parser' && (0, taco_file_parser_1.isTacoFileParserName)(handlerName)) {
            // if filename is a built-in taco file parser, just keep the original name
            handlers[handlerType] = handlerName;
        }
        else {
            handlers[handlerType] = {
                path: resolveHandlerFile(handlerName, handlerType, handlersDirPath),
                className: (0, change_case_1.pascalCase)(handlerName),
            };
        }
    });
    // At this point, all validations are done, and all required infos are resolved.
    return handlers;
}
exports.resolveHandlers = resolveHandlers;
function isHandlerConfig(parser) {
    return typeof parser === 'object' && 'path' in parser && 'className' in parser;
}
exports.isHandlerConfig = isHandlerConfig;
/**
 * Resolve custom handler name in handlerInput into a full file path.
 *
 * @param handlerName handler file name without extension
 * @param handlerType the type of handler to resolve -- for logging, etc.
 * @param handlersDirPath connector's unpack handlers folder with full path
 * @returns the full path of the file that will be used by the sandbox
 */
function resolveHandlerFile(handlerName, handlerType, handlersDirPath) {
    var filename = "".concat(handlerName, ".js");
    var customHandlerPath = node_path_1.default.join(handlersDirPath, filename);
    if (!node_fs_1.default.existsSync(customHandlerPath)) {
        eps_logger_1.Logger.error("CustomHandlerPath does not exist for ".concat(handlerType, ": ").concat(customHandlerPath));
        throw new Error("Could not find handler file '".concat(filename, "' based on property '").concat(handlerType, "': '").concat(handlerName, "'"));
    }
    return customHandlerPath;
}
