// @tableau/taco-toolkit/dist/cli/taco-scripts/command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addLoggingOptions = exports.createScriptCommand = void 0;
var commander_1 = require("commander");
var connector_1 = require("../taco-commands/utils/connector");
var logging_1 = require("../taco-commands/utils/logging");
function createScriptCommand(name, description, config) {
    var program = new commander_1.Command(name).description(description).configureHelp({
        sortOptions: true,
    });
    var loggingOptions = config.loggingOptions, handler = config.handler, setup = config.setup;
    // add the argument and options in the setup stage
    if (setup) {
        setup(program);
    }
    addLoggingOptions(program, loggingOptions);
    // register handlers
    program.action(function () {
        if (loggingOptions) {
            var loggerConfig = createLoggerConfig(program, loggingOptions);
            (0, logging_1.initLogger)(loggerConfig);
        }
        handler(program);
    });
    return program;
}
exports.createScriptCommand = createScriptCommand;
function createLoggerConfig(program, loggingOptions) {
    var options = program.optsWithGlobals();
    var verbose = options.verbose;
    if (loggingOptions.logLevel) {
        var tacoRootPath = (0, connector_1.findTacoRootPath)();
        return {
            verbose: verbose,
            fileLogLevel: options.logLevel,
            fileLogDirPath: (0, logging_1.getLogPath)(tacoRootPath),
        };
    }
    return {
        verbose: verbose,
    };
}
function addLoggingOptions(program, loggingOptions) {
    if (loggingOptions === undefined) {
        return program;
    }
    var verbose = loggingOptions.verbose, logLevel = loggingOptions.logLevel;
    if (verbose) {
        program.option('-v, --verbose', 'run this command in verbose mode', false);
    }
    if (logLevel) {
        program.addOption(new commander_1.Option('--log-level <level>', 'set the logging level to one of the following choices').choices(Object.keys(logging_1.cliLogLevel)));
    }
    return program;
}
exports.addLoggingOptions = addLoggingOptions;
