// @tableau/taco-toolkit/dist/cli/taco-scripts/build/connector-handlers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateEntryFileNames = exports.postProcessHandlers = exports.preProcessHandlers = void 0;
var change_case_1 = require("change-case");
var fs_1 = __importDefault(require("fs"));
var fs_extra_1 = __importDefault(require("fs-extra"));
var path_1 = __importDefault(require("path"));
var error_codes_1 = require("../../../shared/errors/error-codes");
var error_source_1 = require("../../../shared/errors/error-source");
var status_code_1 = require("../../../shared/errors/status-code");
var connector_1 = require("../../taco-commands/utils/connector");
var parcel_error_1 = require("./parcel-error");
/*
   This function will remove all import statements from a handler and make a copy in the tmp folder
 */
function preProcessHandlers(tacoRootPath) {
    var tmpPath = getHandlersTmpPath(tacoRootPath);
    if (!fs_1.default.existsSync(tmpPath)) {
        fs_1.default.mkdirSync(tmpPath);
    }
    var handlersPath = (0, connector_1.getConnectorHandlersPath)(tacoRootPath);
    var entryFiles = [];
    fs_1.default.readdirSync(handlersPath, { withFileTypes: true }).forEach(function (dirent) {
        if (dirent.isFile()) {
            // TODO: this only handles the files on the root level. We should revamp this with a better solution.
            var filename = dirent.name;
            var content = fs_1.default.readFileSync(path_1.default.join(handlersPath, filename));
            var className = (0, change_case_1.pascalCase)(path_1.default.parse(filename).name);
            var processedContent = content.toString().replaceAll(/^\s*import[\s\S]*? from '@tableau\/taco-toolkit\/handlers'$/gm, '') +
                // const mapping is required below since parcel tries to obfuscate the name of the class with a GUID
                "globalRefs[\"".concat(className, "\"] = ").concat(className);
            var distPath = path_1.default.join(tmpPath, filename);
            fs_1.default.writeFileSync(distPath, processedContent);
            entryFiles.push(distPath);
        }
        else {
            // copy directories to tmp for compilation
            var dirname = dirent.name;
            var srcPath = path_1.default.join(handlersPath, dirname);
            var distPath = path_1.default.join(tmpPath, dirname);
            fs_extra_1.default.copySync(srcPath, distPath, { overwrite: true });
        }
    });
    return entryFiles;
}
exports.preProcessHandlers = preProcessHandlers;
function postProcessHandlers(tacoRootPath) {
    var tmpPath = getHandlersTmpPath(tacoRootPath);
    fs_1.default.rmSync(tmpPath, {
        recursive: true,
        force: true,
    });
}
exports.postProcessHandlers = postProcessHandlers;
function validateEntryFileNames(entries) {
    var duplicates = (0, parcel_error_1.findBundleNameDuplicates)(entries);
    if (duplicates.length > 0) {
        return {
            statusCode: status_code_1.StatusCode.Cancelled,
            errorSource: error_source_1.ErrorSources.Client,
            errorCode: error_codes_1.ErrorCodes.INEO9MTU,
            message: "handlers entry file names must be unique regardless of extension: ".concat(duplicates
                .map(function (names) { return "[".concat(names, "]"); })
                .join(', ')),
        };
    }
}
exports.validateEntryFileNames = validateEntryFileNames;
function getHandlersTmpPath(tacoRootPath) {
    return path_1.default.join(tacoRootPath, 'connector', 'tmp');
}
