// @tableau/taco-toolkit/dist/cli/taco-scripts/build/pre-build.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateConnector = void 0;
var validation_1 = require("./validation");
var package_json_1 = require("../../taco-commands/utils/package-json");
var console_1 = require("../../taco-commands/utils/console");
var tableau_product_1 = require("../../taco-commands/utils/tableau-product");
var connector_1 = require("../../taco-commands/utils/connector");
var fs_extra_1 = __importDefault(require("fs-extra"));
var inquirer_1 = __importDefault(require("inquirer"));
var connector_json_1 = require("../../taco-commands/utils/connector-json");
function validateConnector(tacoRootPath, options) {
    return __awaiter(this, void 0, void 0, function () {
        var connectorJsonPath, connectorJson;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    connectorJsonPath = (0, connector_1.getConnectorJsonPath)(tacoRootPath);
                    connectorJson = (0, connector_json_1.getConnectorJson)(connectorJsonPath);
                    if (!(!(0, connector_1.isOldDependencySetup)(tacoRootPath) && !options.skipVersionValidate)) return [3 /*break*/, 2];
                    return [4 /*yield*/, validateTableauMinVersion(connectorJson, tacoRootPath)];
                case 1:
                    _a.sent();
                    _a.label = 2;
                case 2: return [2 /*return*/];
            }
        });
    });
}
exports.validateConnector = validateConnector;
function validateTableauMinVersion(connectorJson, tacoRootPath) {
    return __awaiter(this, void 0, void 0, function () {
        var toolkitTableauVersion, connectorTableauVersion, majorVersion, decision, tacoJsonPath;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    toolkitTableauVersion = package_json_1.PACKAGE_INFO.tableau.version;
                    connectorTableauVersion = connectorJson['tableau-version'].min;
                    majorVersion = (0, tableau_product_1.getTableauMajorVersion)(toolkitTableauVersion);
                    if (!(0, validation_1.shouldBumpTableauMinVersion)(toolkitTableauVersion, connectorTableauVersion)) return [3 /*break*/, 2];
                    return [4 /*yield*/, askForVersionBump(majorVersion, connectorTableauVersion)];
                case 1:
                    decision = _a.sent();
                    if (decision === false) {
                        (0, console_1.consoleWarn)('Build aborted! The connector min Tableau version must be greater than or equals to the taco-toolkit Tableau major version. If the connector is built for earlier version of Tableau, please update the dependency @tableau/taco-toolkit package version accordingly.');
                        process.exit(0);
                    }
                    connectorJson['tableau-version'].min = majorVersion;
                    tacoJsonPath = (0, connector_1.getConnectorJsonPath)(tacoRootPath);
                    fs_extra_1.default.writeJsonSync(tacoJsonPath, connectorJson);
                    (0, console_1.consoleInfo)("The connector Tableau min version has been updated to '".concat(majorVersion, "'"));
                    _a.label = 2;
                case 2: return [2 /*return*/];
            }
        });
    });
}
function askForVersionBump(majorVersion, connectorTableauVersion) {
    return __awaiter(this, void 0, void 0, function () {
        var answers;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, inquirer_1.default.prompt([
                        {
                            type: 'confirm',
                            name: 'bumpAnswer',
                            message: "Your connector Tableau min version '".concat(connectorTableauVersion, "' is below the current taco-toolkit Tableau major version '").concat(majorVersion, "'. Continue to update your connector tableau min version to '").concat(majorVersion, "'?"),
                            default: true,
                        },
                    ])];
                case 1:
                    answers = _a.sent();
                    return [2 /*return*/, answers.bumpAnswer];
            }
        });
    });
}
