// @tableau/taco-toolkit/dist/eps/main/api/eps-handlers/app-url.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWebAuthoringAppUrl = void 0;
var http_status_codes_1 = require("http-status-codes");
var connector_launch_status_1 = require("../../../../shared/enums/connector-launch-status");
var eps_logger_1 = require("../../../logging/eps-logger");
var api_utils_1 = require("../../../utils/api-utils");
var app_global_1 = __importDefault(require("../../core/app-global"));
var handler_utils_1 = require("./handler-utils");
function getWebAuthoringAppUrl(_a) {
    var connectionId = _a.connectionId, connectorClass = _a.connectorClass, pluginPath = _a.pluginPath, staticServerOptions = _a.staticServerOptions, i18n = _a.i18n;
    return __awaiter(this, void 0, void 0, function () {
        var staticServerPort, staticServerId, distinctConnectorName, url, connectorAppUrl, error_1;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    staticServerPort = staticServerOptions.staticServerPort, staticServerId = staticServerOptions.staticServerId;
                    _b.label = 1;
                case 1:
                    _b.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, (0, handler_utils_1.getOrInitializeDistinctConnectorName)({
                            connectionId: connectionId,
                            connectorClass: connectorClass,
                            tacoPath: pluginPath,
                        })];
                case 2:
                    distinctConnectorName = _b.sent();
                    app_global_1.default.tacoInitializer.adaptLocalizedConnector(distinctConnectorName, i18n);
                    url = new URL("http://localhost:".concat(staticServerPort, "/").concat(staticServerId, "/").concat(distinctConnectorName, "/"));
                    connectorAppUrl = url.toString();
                    eps_logger_1.Logger.info("Web Authoring Static Server Url: ".concat((0, api_utils_1.maskURL)(connectorAppUrl)));
                    return [2 /*return*/, {
                            connectorAppUrl: connectorAppUrl,
                        }];
                case 3:
                    error_1 = _b.sent();
                    throw new handler_utils_1.EpsHandlerError("".concat(error_1), {
                        connectorClass: connectorClass,
                        connectorLaunchStatus: connector_launch_status_1.ConnectorLaunchStatus.Error,
                        httpStatusCode: http_status_codes_1.StatusCodes.INTERNAL_SERVER_ERROR,
                    });
                case 4: return [2 /*return*/];
            }
        });
    });
}
exports.getWebAuthoringAppUrl = getWebAuthoringAppUrl;
