// @tableau/taco-toolkit/dist/eps/main/api/eps-handlers/extract.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.waitForExtractorMessage = exports.handleExtract = exports.validateConnectionData = void 0;
var http_status_codes_1 = require("http-status-codes");
var node_path_1 = __importDefault(require("node:path"));
var uuid_1 = require("uuid");
var eps_error_1 = require("../../../types/eps-error");
var eps_error_info_1 = require("../../../types/eps-error-info");
var connector_launch_status_1 = require("../../../../shared/enums/connector-launch-status");
var extractor_type_1 = require("../../../../shared/enums/extractor-type");
var error_codes_1 = require("../../../../shared/errors/error-codes");
var error_source_1 = require("../../../../shared/errors/error-source");
var status_code_1 = require("../../../../shared/errors/status-code");
var extractor_status_1 = require("../../../types/extractor-status");
var eps_logger_1 = require("../../../logging/eps-logger");
var process_manager_1 = require("../../components/process-manager");
var app_global_1 = __importDefault(require("../../core/app-global"));
var eps_process_types_1 = require("../../../types/eps-process-types");
var extractor_message_1 = require("../../../types/extractor-message");
var handler_utils_1 = require("./handler-utils");
var HYPER_FETCHER = 'hyper-fetcher';
var EXTRACTOR_APP_PATH = node_path_1.default.resolve(__dirname, '../../../extractor/extractor-app.js');
var isHyperExtractor = function (handlerInputs) { return handlerInputs.some(function (x) { return x.fetcher === HYPER_FETCHER; }); };
var isNotHyperExtractor = function (handlerInputs) { return handlerInputs.some(function (x) { return x.fetcher !== HYPER_FETCHER; }); };
function validateConnectionData(body) {
    if (!body || typeof body !== 'object' || Object.keys(body).length === 0) {
        var message = 'Bad Request: Request body is empty or not JSON; request body must have ConnectionData.';
        eps_logger_1.Logger.logEPSError(status_code_1.StatusCode.FailedPrecondition, error_source_1.ErrorSources.Client, error_codes_1.ErrorCodes.M92GXFQP, message);
        throw Error(message);
    }
    // Check there are some handlerInputs
    var handlerInputs = body.handlerInputs;
    if (!handlerInputs || !Array.isArray(handlerInputs) || handlerInputs.length === 0) {
        var errorMessage = 'Bad Request: handlerInputs must be provided in request body.';
        eps_logger_1.Logger.logEPSError(status_code_1.StatusCode.FailedPrecondition, error_source_1.ErrorSources.Configuration, error_codes_1.ErrorCodes.ENT97JSH, errorMessage);
        throw new eps_error_1.EPSError(errorMessage, new eps_error_info_1.EPSErrorInfo(handler_utils_1.ERROR_CODE_BAD_TACO), connector_launch_status_1.ConnectorLaunchStatus.BadTaco);
    }
    // Check hyper fetcher and eps fetcher are not in same request, which is currently not supported
    if (isHyperExtractor(handlerInputs) && isNotHyperExtractor(handlerInputs)) {
        var error = new Error("You cannot mix ".concat(HYPER_FETCHER, " with non-").concat(HYPER_FETCHER, " in the same request."));
        eps_logger_1.Logger.error(error);
        throw error;
    }
}
exports.validateConnectionData = validateConnectionData;
// TODO: figure out if it is used. If not, remove this!
function handleHyperExtractor(extractorLaunchedEvent, extractHandlerRequest) {
    var extractorType = extractor_type_1.ExtractorType.hyper;
    var connectionId = extractHandlerRequest.connectionId, connectionData = extractHandlerRequest.connectionData, secrets = extractHandlerRequest.secrets;
    var extractorLaunchSuccessOptions = {
        type: 'success',
        extractorType: extractorType,
        connectionData: connectionData,
        secrets: secrets,
    };
    extractorLaunchedEvent.emit((0, handler_utils_1.getExtractorLaunchedConnectionId)(connectionId), extractorLaunchSuccessOptions);
    return {
        response: { status: connector_launch_status_1.ConnectorLaunchStatus.Extract, extractorType: extractorType, connectionData: connectionData, secrets: secrets },
        httpStatusCode: http_status_codes_1.StatusCodes.OK,
    };
}
function handleEpsExtractor(extractorLaunchedEvent, extractHandlerRequest) {
    return __awaiter(this, void 0, void 0, function () {
        var connectionId, secrets, connectionData, extractorMessage, extractorType, extractorLaunchSuccessOptions, extractApiResponse, errorMessage;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    connectionId = extractHandlerRequest.connectionId, secrets = extractHandlerRequest.secrets, connectionData = extractHandlerRequest.connectionData;
                    return [4 /*yield*/, launchEpsExtractor(extractHandlerRequest)];
                case 1:
                    extractorMessage = _a.sent();
                    extractorType = extractor_type_1.ExtractorType.eps;
                    extractorLaunchSuccessOptions = {
                        type: 'success',
                        extractorType: extractorType,
                        extractorMessage: extractorMessage,
                        connectionData: connectionData,
                        secrets: secrets,
                    };
                    extractorLaunchedEvent.emit((0, handler_utils_1.getExtractorLaunchedConnectionId)(connectionId), extractorLaunchSuccessOptions);
                    extractApiResponse = {
                        status: connector_launch_status_1.ConnectorLaunchStatus.Extract,
                        extractorType: extractorType,
                        port: extractorMessage.port,
                        extractorId: extractorMessage.extractorId,
                    };
                    errorMessage = extractorMessage.errorMessage;
                    if (errorMessage) {
                        extractApiResponse.status = connector_launch_status_1.ConnectorLaunchStatus.Error;
                        extractApiResponse.errorMessage = errorMessage;
                    }
                    return [2 /*return*/, {
                            httpStatusCode: http_status_codes_1.StatusCodes.OK,
                            response: extractApiResponse,
                        }];
            }
        });
    });
}
function handleExtract(extractorLaunchedEvent, extractHandlerRequest) {
    return __awaiter(this, void 0, void 0, function () {
        var connectionId, connectionData, error_1, options;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    connectionId = extractHandlerRequest.connectionId, connectionData = extractHandlerRequest.connectionData;
                    // Done with browser, mark completed to declare it was completed properly.
                    app_global_1.default.connectionManager.setIsSubmitted(connectionId, true);
                    if (isHyperExtractor(connectionData.handlerInputs)) {
                        return [2 /*return*/, handleHyperExtractor(extractorLaunchedEvent, extractHandlerRequest)];
                    }
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, handleEpsExtractor(extractorLaunchedEvent, extractHandlerRequest)];
                case 2: return [2 /*return*/, _a.sent()];
                case 3:
                    error_1 = _a.sent();
                    options = {
                        type: 'failure',
                        error: error_1,
                        connectorLaunchStatus: connector_launch_status_1.ConnectorLaunchStatus.BadTaco,
                    };
                    extractorLaunchedEvent.emit((0, handler_utils_1.getExtractorLaunchedConnectionId)(connectionId), options);
                    return [2 /*return*/, {
                            httpStatusCode: http_status_codes_1.StatusCodes.OK,
                            response: new eps_error_1.EPSError(error_1, new eps_error_info_1.EPSErrorInfo(handler_utils_1.ERROR_CODE_BAD_TACO), connector_launch_status_1.ConnectorLaunchStatus.BadTaco),
                        }];
                case 4: return [2 /*return*/];
            }
        });
    });
}
exports.handleExtract = handleExtract;
function launchEpsExtractor(extractHandlerRequest) {
    return __awaiter(this, void 0, void 0, function () {
        var tacoPath, connectionId, connectorClass, connectionData, secrets, distinctConnectorName, handlersDir, permission, extractorOptions, portConfig, extractorId, extractorProcess, extractorMessage;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    tacoPath = extractHandlerRequest.tacoPath, connectionId = extractHandlerRequest.connectionId, connectorClass = extractHandlerRequest.connectorClass, connectionData = extractHandlerRequest.connectionData, secrets = extractHandlerRequest.secrets;
                    return [4 /*yield*/, (0, handler_utils_1.getOrInitializeDistinctConnectorName)({
                            connectionId: connectionId,
                            connectorClass: connectorClass,
                            tacoPath: tacoPath,
                        })];
                case 1:
                    distinctConnectorName = _a.sent();
                    if (app_global_1.default.connectionManager.hasActiveConnection(connectionId)) {
                        // With ExtractorHandler, this should never be hit. Keep it here for a sanity check to avoid creating an orphan extractor process.
                        throw new Error("There is already a live extractor for ConnectionId '".concat(connectionId, "'."));
                    }
                    handlersDir = app_global_1.default.tacoInitializer.getConnectorHandlersPath(distinctConnectorName);
                    permission = app_global_1.default.tacoInitializer.getConnectorConfig(distinctConnectorName).permission;
                    extractorOptions = {
                        handlersDir: handlersDir,
                        connectorClass: connectorClass,
                        connectionId: connectionId,
                        handlerInputs: connectionData.handlerInputs,
                        secrets: secrets,
                        permission: permission,
                    };
                    portConfig = {
                        portRange: app_global_1.default.portManager.getPortRange(),
                        reservedPorts: app_global_1.default.portManager.getReservedPorts(),
                    };
                    extractorId = (0, uuid_1.v4)();
                    extractorProcess = app_global_1.default.processManager.fork(eps_process_types_1.EPSProcessName.EPSExtractor, process_manager_1.ProcessType.child, EXTRACTOR_APP_PATH, [
                        extractorId,
                        // TOOD: nodejs has size limit on param length. We need to handle it better.
                        JSON.stringify(extractorOptions),
                        JSON.stringify(portConfig),
                    ]);
                    return [4 /*yield*/, waitForExtractorMessage(extractorProcess)];
                case 2:
                    extractorMessage = _a.sent();
                    if ((0, extractor_message_1.isSchemaDoneExtractorMessage)(extractorMessage)) {
                        app_global_1.default.connectionManager.addExtractorProcess(connectionId, extractorProcess, extractorId);
                    }
                    return [2 /*return*/, extractorMessage];
            }
        });
    });
}
/**
 * Wait for the extractor process to send the message to EPS.
 * The function will settle the promise on the first (known) message sent from extractor.
 */
function waitForExtractorMessage(extractorProcess) {
    var schemaDone = false;
    return new Promise(function (resolve, reject) {
        var _a, _b;
        (_a = extractorProcess.stdout) === null || _a === void 0 ? void 0 : _a.on('data', function (data) {
            eps_logger_1.Logger.info("Extractor (pid:".concat(extractorProcess.pid, ") stdout: ").concat(data));
        });
        (_b = extractorProcess.stderr) === null || _b === void 0 ? void 0 : _b.on('data', function (data) {
            eps_logger_1.Logger.error("Extractor (pid:".concat(extractorProcess.pid, ") stderr: ").concat(data));
        });
        extractorProcess.on('message', function (message) {
            var status = message.status;
            if (status === extractor_status_1.ExtractorStatus.SchemaDone) {
                schemaDone = true;
                resolve(message);
            }
            else if (status === extractor_status_1.ExtractorStatus.ExtractorError) {
                var errMsg = "Extractor (pid:".concat(extractorProcess.pid, ") reported an error: ").concat(message.errorMessage);
                eps_logger_1.Logger.logEPSError(status_code_1.StatusCode.Cancelled, error_source_1.ErrorSources.Client, error_codes_1.ErrorCodes.HU3H87WB, errMsg);
                reject(errMsg);
            }
            else {
                // All messages from extractor should be sent by sendExtractorMessage(), so based on the code,
                // this line should never be hit. Just logging here in case anything is missed.
                eps_logger_1.Logger.debug(message);
            }
        });
        extractorProcess.on('error', function (err) {
            // ERR_IPC_CHANNEL_CLOSED is raised when parent is closing and is ignorable in that case.
            var errMsg = "Extractor (pid:".concat(extractorProcess.pid, ") emitted an error event: ").concat(err, ".");
            eps_logger_1.Logger.logEPSError(status_code_1.StatusCode.Cancelled, error_source_1.ErrorSources.Client, error_codes_1.ErrorCodes.PSZECVJ1, errMsg);
            reject(errMsg);
        });
        extractorProcess.on('disconnect', function () {
            if (!schemaDone) {
                var errMsg = "Extractor (pid:".concat(extractorProcess.pid, ") disconnected before extractor completes. Check logs for more details.");
                reject(errMsg);
            }
            // When schemaDone is true, the promise is already settled.
            // Log here for extractor process status.
            eps_logger_1.Logger.info("Extractor (pid:".concat(extractorProcess.pid, ") disconnected."));
        });
        extractorProcess.on('exit', function (code, signal) {
            if (!schemaDone) {
                var errMsg = "Extractor (pid:".concat(extractorProcess.pid, ") exited before extractor completes. Check logs for more details.");
                reject(errMsg);
            }
            // When schemaDone is true, the promise is already settled.
            // Log here for extractor process status.
            eps_logger_1.Logger.info("Extractor (pid:".concat(extractorProcess.pid, ") exited with code ").concat(code, " on signal '").concat(signal, "'."));
        });
        extractorProcess.on('close', function (code, signal) {
            if (!schemaDone) {
                var errMsg = "Extractor (pid:".concat(extractorProcess.pid, ") closed before extractor completes. Check logs for more details.");
                reject(errMsg);
            }
            // When schemaDone is true, the promise is already settled.
            // Log here for extractor process status.
            eps_logger_1.Logger.info("Extractor (pid:".concat(extractorProcess.pid, ") closed with code ").concat(code, " on signal '").concat(signal, "'."));
        });
    });
}
exports.waitForExtractorMessage = waitForExtractorMessage;
