// @tableau/taco-toolkit/dist/eps/main/api/eps-handlers/handler-utils.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validatePluginPath = exports.validateInstanceId = exports.validateCallerId = exports.getExtractorLaunchedConnectionId = exports.getOrInitializeDistinctConnectorName = exports.EpsHandlerError = exports.ERROR_CODE_GENERAL_EXTRACTOR = exports.ERROR_CODE_BAD_TACO = void 0;
var eps_logger_1 = require("../../../logging/eps-logger");
var app_global_1 = __importDefault(require("../../core/app-global"));
exports.ERROR_CODE_BAD_TACO = '0x78715AB7';
exports.ERROR_CODE_GENERAL_EXTRACTOR = '0x78715AB7'; // TODO use a different one
var EpsHandlerError = /** @class */ (function (_super) {
    __extends(EpsHandlerError, _super);
    function EpsHandlerError(error, errorInfo) {
        var _this = 
        // if (error instanceof Error) {
        //   error = error.message
        // }
        _super.call(this, error) || this;
        _this.error = error;
        _this.errorInfo = errorInfo;
        _this.name = 'EPSHandlerError';
        return _this;
    }
    Object.defineProperty(EpsHandlerError.prototype, "httpStatusCode", {
        get: function () {
            return this.errorInfo.httpStatusCode;
        },
        enumerable: false,
        configurable: true
    });
    return EpsHandlerError;
}(Error));
exports.EpsHandlerError = EpsHandlerError;
function getOrInitializeDistinctConnectorName(_a) {
    var _b;
    var connectionId = _a.connectionId, connectorClass = _a.connectorClass, tacoPath = _a.tacoPath;
    return __awaiter(this, void 0, void 0, function () {
        var error;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    // Look for tacoPath in header or else associated with connectionId (as required for vizQLServer)
                    if (!tacoPath) {
                        tacoPath = (_b = app_global_1.default.connectionManager.tacoPath(connectionId)) !== null && _b !== void 0 ? _b : undefined;
                    }
                    if (tacoPath === undefined) {
                        error = new Error("tacoPath was not provided by plugin-path header nor associated with connectionId '".concat(connectionId, "'."));
                        eps_logger_1.Logger.error(error);
                        throw error;
                    }
                    return [4 /*yield*/, app_global_1.default.tacoInitializer.getOrInitializeDistinctConnectorName({
                            connectionId: connectionId,
                            connectorClass: connectorClass,
                            tacoPath: tacoPath,
                        })];
                case 1: return [2 /*return*/, _c.sent()];
            }
        });
    });
}
exports.getOrInitializeDistinctConnectorName = getOrInitializeDistinctConnectorName;
function getExtractorLaunchedConnectionId(connectionId) {
    return "extractorLaunched-".concat(connectionId);
}
exports.getExtractorLaunchedConnectionId = getExtractorLaunchedConnectionId;
function validateCallerId(callerId, expectedCallerId) {
    if (!callerId) {
        throw new Error("callerId must be a string, but found ".concat(callerId));
    }
    if (callerId !== expectedCallerId && callerId !== '0') {
        throw new Error("callerId ".concat(callerId, " is not the same as the callerId for server initialization '").concat(expectedCallerId, "'."));
    }
}
exports.validateCallerId = validateCallerId;
function validateInstanceId(epsInstanceId, expectedInstanceId) {
    if (!epsInstanceId) {
        throw new Error("epsInstanceId must be a string, but found ".concat(epsInstanceId));
    }
    if (epsInstanceId !== '0' && epsInstanceId !== expectedInstanceId) {
        throw new Error("epsInstanceId ".concat(epsInstanceId, " is not the same as the epsInstanceId for server initialization '").concat(expectedInstanceId, "'."));
    }
}
exports.validateInstanceId = validateInstanceId;
function validatePluginPath(pluginPath) {
    if (typeof pluginPath !== 'string') {
        throw new Error("pluginPath must be a string, but found ".concat(pluginPath));
    }
    if (pluginPath.length === 0) {
        throw new Error('pluginPath must be a non-empty string');
    }
}
exports.validatePluginPath = validatePluginPath;
